{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# process the scRNA-seq data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# analysis script for analyzing human and mouse reads description\n",
    "suppressPackageStartupMessages({\n",
    "    library(monocle)\n",
    "    library(stringr)\n",
    "    library(Matrix.utils)\n",
    "    library(Matrix)\n",
    "    library(knitr)\n",
    "    library(tidyverse)\n",
    "    library(monocle)\n",
    "    library(RColorBrewer)\n",
    "    library(cowplot)\n",
    "    options(stringsAsFactors = FALSE)})\n",
    "\n",
    "#chosen_color = c('#e41a1c', '#377eb8', '#4daf4a', '#984ea3', '#ff7f00')\n",
    "chosen_color <- c(\"#999999\", \"#E69F00\", \"#56B4E9\", \"#009E73\", \"#F0E442\", \"#0072B2\", \"#D55E00\", \"#CC79A7\")\n",
    "my_palette = c(brewer.pal(5, \"Set1\")[c(1,2,4,5)], brewer.pal(5, \"Pastel1\")[c(2,5,1,4)])\n",
    "\n",
    "\n",
    "\n",
    "### read in the gene count matrix, gene annotat file, cell annotat file, and generate the sparse matrix\n",
    "### Here I write a function that accept a df_gene, df_cell, and a sparse matrix of the gene count, and then\n",
    "# count the gene count in human species ('_ambiguous' '_ambiguous_intron' 'ENSG' 'ENSMUSG' '_no_feature' '_unmapped')\n",
    "\n",
    "species_count = function(df_gene, df_cell, gene_matrix) {\n",
    "    df_gene_sep = df_gene %>% separate(gene_id, c(\"source\", \"gene_id\"), sep = \"0\")\n",
    "    unique_species = unique(df_gene_sep$source)\n",
    "    \n",
    "    # for each element in the unique species, filter out the genes for the gene_matrix and\n",
    "    # sum the count for each cell\n",
    "    # add a new column to the df_cell object\n",
    "    \n",
    "    for(i in 1:length(unique_species)) {\n",
    "        s_name = unique_species[i]\n",
    "        filtered_gene_matrix = gene_matrix[df_gene_sep$source == s_name, ]\n",
    "        s_count = colSums(filtered_gene_matrix)\n",
    "        df_cell[s_name] = s_count\n",
    "        g_name = paste(s_name, \"_gene\", sep=\"\")\n",
    "        g_count = colSums(filtered_gene_matrix > 0)\n",
    "        df_cell[g_name] = g_count\n",
    "    }\n",
    "    return(df_cell)\n",
    "}\n",
    "\n",
    "# Here I write a function, that accept a df_gene, and gene_count matrix, and then return a gene count matrix with\n",
    "# combined exonic and intronic count\n",
    "combine_exon_intron <- function(df_gene, gene_count) {\n",
    "    gene_count_exon = gene_count[df_gene$exon_intron == \"exon\", ]\n",
    "    gene_count_intron = gene_count[df_gene$exon_intron == \"intron\", ]\n",
    "    if(nrow(gene_count_exon) == nrow(gene_count_intron)) {\n",
    "        gene_count_combine = gene_count_exon  + gene_count_intron\n",
    "    }\n",
    "    else {\n",
    "        gene_count_combine = gene_count_exon[-nrow(gene_count_exon), ] + gene_count_intron\n",
    "        gene_count_combine = rbind(gene_count_combine, gene_count_exon[nrow(gene_count_exon), ])\n",
    "    }\n",
    "    \n",
    "    return(gene_count_combine)\n",
    "}\n",
    "\n",
    "# This function convert the count data to log transformed TPM data.\n",
    "count_to_logtmp <- function(counts)\n",
    "{\n",
    "  allsum = sum(counts)\n",
    "  result = counts / allsum * 1000000\n",
    "  result = log(result + 1)\n",
    "  return(result)\n",
    "}\n",
    "\n",
    "## Here I define a function that accept a gene folder and then return a object including a df_gene, df_cell and gene count\n",
    "# object\n",
    "\n",
    "sciRNAseq_gene_count_summary <- function(gene_count_folder)\n",
    "    {\n",
    "    # read the gene, cell and gene count matrix\n",
    "    gene_matrix = paste(gene_count_folder, \"/count.MM\", sep = \"\")\n",
    "    df_gene = paste(gene_count_folder, \"/gene_name_annotate.txt\", sep = \"\")\n",
    "    df_cell = paste(gene_count_folder, \"/cell_annotate.txt\", sep = \"\")\n",
    "    df_report = paste(gene_count_folder, \"/report.MM\", sep = \"\")\n",
    "    report_annotate = paste(gene_count_folder, \"/report_annotate.txt\", sep = \"\")\n",
    "\n",
    "    df_gene = read.csv(df_gene, header=F)\n",
    "    df_cell = read.csv(df_cell, header=F)\n",
    "    gene_matrix = read.csv(gene_matrix, header=F)\n",
    "\n",
    "    colnames(df_gene) = c(\"gene_id\", \"gene_type\", \"exon_intron\", \"gene_name\", \"index\")\n",
    "    colnames(df_cell) = c(\"sample\", \"index\")\n",
    "    rownames(df_gene) = df_gene$gene_id\n",
    "    rownames(df_cell) = df_cell$cell_name\n",
    "    \n",
    "    gene_count = sparseMatrix(i = gene_matrix$V1, j = gene_matrix$V2, x = gene_matrix$V3)\n",
    "    df_gene = df_gene[1:nrow(gene_count),]\n",
    "    rownames(gene_count) = df_gene$gene_id\n",
    "    colnames(gene_count) = df_cell$cell_name\n",
    "    \n",
    "    gene_count = combine_exon_intron(df_gene, gene_count)\n",
    "    df_gene = df_gene %>% filter(exon_intron == \"exon\")\n",
    "\n",
    "    df_cell = species_count(df_gene, df_cell, gene_count)\n",
    "\n",
    "    # Read in the report matrix (including the reads number mapping to exons and introns)\n",
    "    reportMM = read.csv(df_report, header=F)\n",
    "    df_report = sparseMatrix(i = reportMM$V1, j = reportMM$V2, x = reportMM$V3)\n",
    "    df_report = as.matrix(t(df_report))\n",
    "    df_report_annotate = read.csv(report_annotate, header=F)\n",
    "    colnames(df_report) = df_report_annotate$V2\n",
    "    df_report = data.frame(df_report)\n",
    "    df_report[\"index\"] = as.numeric(rownames(df_report))\n",
    "    df_cell_combine = inner_join(df_cell,df_report, by=\"index\")\n",
    "    df_cell_combine[\"all_exon\"] = df_cell_combine$X.Perfect.intersect.exon.match + df_cell_combine$X.Nearest.intersect.exon.match + df_cell_combine$X.Perfect.combine.exon.match + df_cell_combine$X.Nearest.combine.exon.match\n",
    "    df_cell_combine[\"all_intron\"] = df_cell_combine$X.Perfect.intersect.gene.match + df_cell_combine$X.Nearest.intersect.gene.match + df_cell_combine$X.Perfect.combine.gene.match + df_cell_combine$X.Nearest.combine.gene.match\n",
    "    df_cell_combine[\"all_reads\"] = df_cell_combine$all_exon + df_cell_combine$all_intron + df_cell_combine$X.No.match\n",
    "    df_cell_combine[\"unmatched_rate\"] = df_cell_combine$X.No.match / df_cell_combine$all_reads\n",
    "    df_cell_combine[\"exon_rate\"] = df_cell_combine$all_exon / df_cell_combine$all_reads\n",
    "    df_cell_combine[\"intron_rate\"] = df_cell_combine$all_intron / df_cell_combine$all_reads\n",
    "    rownames(df_cell_combine) = df_cell_combine$cell_name\n",
    "    # return the df_cell, df_gene_gene_count files\n",
    "    return(list(df_cell_combine, df_gene, gene_count))\n",
    "}\n",
    "\n",
    "# Here I define a function, that accept a report file and then generate the summary file including the df_gene, df_cell\n",
    "# that combine the gene count file and the human, mouse read count file, and the gene count file\n",
    "\n",
    "sciRNA_summary <- function(report_folder, purity_limit)\n",
    "    {\n",
    "    gene_count_folder = paste(report_folder, \"/human_mouse_gene_count\", sep = \"\")\n",
    "    sc_read_file = paste(report_folder, \"/read_human_mouse/human_mouse_fraction.txt\", sep = \"\")\n",
    "    sc_read = read.csv(sc_read_file, header=T)\n",
    "    sc_read[\"total\"] = sc_read$human_reads + sc_read$mouse_reads + sc_read$cele_reads\n",
    "    sc_read[\"human_ratio\"] = sc_read$human_reads / sc_read$total\n",
    "    sc_read[\"mouse_ratio\"] = sc_read$mouse_reads / sc_read$total\n",
    "    sc_read[\"cele_ratio\"] = sc_read$cele_reads / sc_read$total\n",
    "    sc_read[\"source\"] = \"Mixed\"\n",
    "    sc_read$source[sc_read$human_ratio > purity_limit] = \"Human\"\n",
    "    sc_read$source[sc_read$mouse_ratio > purity_limit] = \"Mouse\"\n",
    "    sc_read$source[sc_read$cele_ratio > purity_limit] = \"c.elegans\"\n",
    "    \n",
    "    gene_count_output = sciRNAseq_gene_count_summary(gene_count_folder)\n",
    "    df_cell = gene_count_output[[1]]\n",
    "    df_gene = gene_count_output[[2]]\n",
    "    gene_count = gene_count_output[[3]]\n",
    "    df_cell = full_join(df_cell, sc_read, by=\"sample\")\n",
    "    return(list(df_cell, df_gene, gene_count))\n",
    "}\n",
    "\n",
    "\n",
    "plot_human_mouse_summary <- function(df_cell)\n",
    "    {\n",
    "    # this function accept a df_cell object and then return the average human and mouse reads,\n",
    "    # average human and mouse genes, human and mouse purity, and also plot the box plot for\n",
    "    # all of these\n",
    "    \n",
    "    df_cell_human = df_cell %>% filter(source == \"Human\")\n",
    "    df_cell_mouse = df_cell %>% filter(source == \"Mouse\")\n",
    "    \n",
    "    human_reads = median(df_cell_human$human_reads)\n",
    "    mouse_reads = median(df_cell_mouse$mouse_reads)\n",
    "    \n",
    "    human_gene = median(df_cell_human$ENSG_gene)\n",
    "    mouse_gene = median(df_cell_mouse$ENSMUSG_gene)\n",
    "    \n",
    "    human_purity = median(df_cell_human$human_ratio)\n",
    "    mouse_purity = median(df_cell_mouse$mouse_ratio)\n",
    "    \n",
    "    doublet_rate = (nrow(df_cell) - nrow(df_cell_human) - nrow(df_cell_mouse)) / nrow(df_cell)\n",
    "    df_result = data.frame(\"Human\" = c(human_reads, human_gene, human_purity, doublet_rate), \n",
    "                           \"Mouse\" = c(mouse_reads, mouse_gene, mouse_purity, doublet_rate))\n",
    "    \n",
    "    rownames(df_result) = c(\"UMI\", \"Gene\", \"purity\", \"doublet_rate\")\n",
    "    g_reads = (ggplot() + geom_boxplot(aes(x=\"Human\", y = df_cell_human$human_reads), fill = chosen_color[7])\n",
    "              + geom_boxplot(aes(x=\"Mouse\", y=df_cell_mouse$mouse_reads), fill = chosen_color[6])\n",
    "              + labs(x=\"\", y=\"# of UMIs\"))\n",
    "    g_gene = (ggplot() + geom_boxplot(aes(x=\"Human\", y = df_cell_human$ENSG_gene), fill = chosen_color[7])\n",
    "              + geom_boxplot(aes(x=\"Mouse\", y=df_cell_mouse$ENSMUSG_gene), fill = chosen_color[6])\n",
    "              + labs(x=\"\", y=\"# of Genes\"))\n",
    "    \n",
    "    g_purity = (ggplot() + geom_boxplot(aes(x=\"Human\", y = df_cell_human$human_ratio), fill = chosen_color[7])\n",
    "              + geom_boxplot(aes(x=\"Mouse\", y=df_cell_mouse$mouse_ratio), fill = chosen_color[6])\n",
    "              + labs(x=\"\", y=\"Purity\"))\n",
    "    \n",
    "    g_distribution = (ggplot(data=df_cell, aes(x=human_reads, y=mouse_reads, colour=source))\n",
    "                        + geom_point()\n",
    "                        + labs(x=\"# of human UMIs\", y=\"# of mouse UMIs\")\n",
    "                        + scale_colour_manual(values = c(chosen_color[7], chosen_color[1], chosen_color[6]))\n",
    "                        + theme(legend.justification=c(1, 0), legend.position=c(1, 0.5)))\n",
    "    return(list(df_result, g_reads, g_gene, g_purity, g_distribution))\n",
    "}\n",
    "\n",
    "plot_UMI_gene <- function(df_cell)\n",
    "    {\n",
    "    # accept a data frame and plot the UMIs and genes boxplot for all the cells\n",
    "    g1_human = (ggplot()\n",
    "                + geom_boxplot(aes(x=\"UMIs\", y=df_cell$human_reads), fill = chosen_color[7])\n",
    "                + geom_boxplot(aes(x=\"Genes\", y=df_cell$ENSG_gene), fill = chosen_color[7])\n",
    "                + labs(x=\"\", y=\"Count\"))\n",
    "    g1_mouse = (ggplot()\n",
    "                + geom_boxplot(aes(x=\"UMIs\", y=df_cell$mouse_reads), fill = chosen_color[6])\n",
    "                + geom_boxplot(aes(x=\"Genes\", y=df_cell$ENSMUSG_gene), fill = chosen_color[6])\n",
    "                + labs(x=\"\", y=\"Count\"))\n",
    "    return(list(g1_human, g1_mouse))\n",
    "}\n",
    "\n",
    "construct_cds <- function(UMI, df_cell, df_gene)\n",
    "    {\n",
    "    # accept a UMI count matrix, a df_cell data frame, a df_gene data frame, and then construct\n",
    "    # a cds object for monocle downstream analysis\n",
    "    # and then estimate the dispersion and size\n",
    "    df_gene = df_gene %>% plyr::rename(c(\"gene_name\" = \"gene_short_name\"))\n",
    "    pd = new(\"AnnotatedDataFrame\", data = df_cell)\n",
    "    fd = new(\"AnnotatedDataFrame\", data = df_gene)\n",
    "    \n",
    "    colnames(UMI) = df_cell$sample\n",
    "    row.names(UMI) = df_gene$gene_id\n",
    "    row.names(pd) = colnames(UMI)\n",
    "    row.names(fd) = row.names(UMI) \n",
    "    \n",
    "    cds = newCellDataSet(UMI, phenoData = pd, featureData = fd, expressionFamily = negbinomial.size())\n",
    "    \n",
    "    cds = estimateSizeFactors(cds)\n",
    "    cds = estimateDispersions(cds)\n",
    "    cds = detectGenes(cds, 0.1)\n",
    "    return(cds)\n",
    "}\n",
    "\n",
    "\n",
    "# here I define a function that accept a cds, and filter the cells and genes\n",
    "# for cells, it will calculate the cells with outlier gene number (mean +- 3 SE)\n",
    "filter_cds <- function(cds) {\n",
    "    # filter cells with low read count and also high read count\n",
    "    pData(cds)$Total_mRNAs <- Matrix::colSums(exprs(cds))\n",
    "    cds <- cds[,pData(cds)$Total_mRNAs < 1e6]\n",
    "    upper_bound <- 10^(mean(log10(pData(cds)$Total_mRNAs)) + 2*sd(log10(pData(cds)$Total_mRNAs))) \n",
    "    lower_bound <- 10^(mean(log10(pData(cds)$Total_mRNAs)) - 2*sd(log10(pData(cds)$Total_mRNAs))) \n",
    "    \n",
    "    g1 = (qplot(Total_mRNAs, data=pData(cds), geom=\"density\", colour=cell_type) \n",
    "          + geom_vline(xintercept=lower_bound) + geom_vline(xintercept=upper_bound)\n",
    "          + monocle:::monocle_theme_opts())\n",
    "    g1\n",
    "\n",
    "    cat(\"Upper bound for cell filter: \", upper_bound)\n",
    "    cat(\"\\n\")\n",
    "    cat(\"Lower bound for cell filter:\", lower_bound)\n",
    "    \n",
    "    # filter the cells\n",
    "    cds <- cds[,pData(cds)$Total_mRNAs > lower_bound &\n",
    "             pData(cds)$Total_mRNAs < upper_bound]\n",
    "\n",
    "    fData(cds)$prop.expr = apply(exprs(cds), 1, function(x) sum(x > 0) / length(x))\n",
    "    expressed.genes = rownames(subset(fData(cds), prop.expr >= 0.005))\n",
    "    cat(\"\\nTotal number of expressed genes: \", length(expressed.genes))\n",
    "    cds = cds[expressed.genes,]\n",
    "    return(cds)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘cowplot’\n",
      "\n",
      "The following object is masked from ‘package:ggplot2’:\n",
      "\n",
      "    ggsave\n",
      "\n",
      "Warning message in file(file, \"rt\"):\n",
      "“cannot open file './data/STAR_cele_gene_count/gene_name_annotate.txt': No such file or directory”"
     ]
    },
    {
     "ename": "ERROR",
     "evalue": "Error in file(file, \"rt\"): cannot open the connection\n",
     "output_type": "error",
     "traceback": [
      "Error in file(file, \"rt\"): cannot open the connection\nTraceback:\n",
      "1. read.csv(df_gene, header = F)",
      "2. read.table(file = file, header = header, sep = sep, quote = quote, \n .     dec = dec, fill = fill, comment.char = comment.char, ...)",
      "3. file(file, \"rt\")"
     ]
    }
   ],
   "source": [
    "# for c.elegans analysis\n",
    "library(monocle)\n",
    "library(stringr)\n",
    "library(dplyr)\n",
    "library(tidyr)\n",
    "library(Matrix.utils)\n",
    "library(Matrix)\n",
    "library(knitr)\n",
    "library(cowplot)\n",
    "options(stringsAsFactors = FALSE)\n",
    "#chosen_color = c('#e41a1c', '#377eb8', '#4daf4a', '#984ea3', '#ff7f00')\n",
    "#chosen_color = c('#e41a1c','#984ea3', '#377eb8', '#ff7f00', '#4daf4a')\n",
    "chosen_color <- c(\"#999999\", \"#E69F00\", \"#56B4E9\", \"#009E73\", \"#F0E442\", \"#0072B2\", \"#D55E00\", \"#CC79A7\")\n",
    "options(jupyter.plot_mimetypes = \"image/svg+xml\")\n",
    "options(repr.plot.width  = 4,\n",
    "        repr.plot.height = 3)\n",
    "\n",
    "### read in the gene count matrix, gene annotat file, cell annotat file, and generate the sparse matrix\n",
    "### Here I write a function that accept a df_gene, df_cell, and a sparse matrix of the gene count, and then\n",
    "# count the gene count in human species ('_ambiguous' '_ambiguous_intron' 'ENSG' 'ENSMUSG' '_no_feature' '_unmapped')\n",
    "\n",
    "species_count = function(df_gene, df_cell, gene_matrix) {\n",
    "    df_gene_sep = df_gene %>% separate(gene_id, c(\"source\", \"gene_id\"), sep = \"0\")\n",
    "    unique_species = unique(df_gene_sep$source)\n",
    "    \n",
    "    # for each element in the unique species, filter out the genes for the gene_matrix and\n",
    "    # sum the count for each cell\n",
    "    # add a new column to the df_cell object\n",
    "    \n",
    "    for(i in 1:length(unique_species)) {\n",
    "        s_name = unique_species[i]\n",
    "        filtered_gene_matrix = gene_matrix[df_gene_sep$source == s_name, ]\n",
    "        s_count = colSums(filtered_gene_matrix)\n",
    "        df_cell[s_name] = s_count\n",
    "        g_name = paste(s_name, \"_gene\", sep=\"\")\n",
    "        g_count = colSums(filtered_gene_matrix > 0)\n",
    "        df_cell[g_name] = g_count\n",
    "    }\n",
    "    return(df_cell)\n",
    "}\n",
    "\n",
    "# Here I write a function, that accept a df_gene, and gene_count matrix, and then return a gene count matrix with\n",
    "# combined exonic and intronic count\n",
    "combine_exon_intron <- function(df_gene, gene_count) {\n",
    "    gene_count_exon = gene_count[df_gene$exon_intron == \"exon\", ]\n",
    "    gene_count_intron = gene_count[df_gene$exon_intron == \"intron\", ]\n",
    "    if(nrow(gene_count_exon) == nrow(gene_count_intron)) {\n",
    "        gene_count_combine = gene_count_exon  + gene_count_intron\n",
    "    }\n",
    "    else {\n",
    "        gene_count_combine = gene_count_exon[-nrow(gene_count_exon), ] + gene_count_intron\n",
    "        gene_count_combine = rbind(gene_count_combine, gene_count_exon[nrow(gene_count_exon), ])\n",
    "    }\n",
    "    \n",
    "    return(gene_count_combine)\n",
    "}\n",
    "\n",
    "# Check the gene count data\n",
    "gene_matrix = \"./data/STAR_cele_gene_count/count.MM\"\n",
    "df_gene = \"./data/STAR_cele_gene_count/gene_name_annotate.txt\"\n",
    "df_cell = \"./data/STAR_cele_gene_count/cell_annotate.txt\"\n",
    "df_report = \"./data/STAR_cele_gene_count/report.MM\"\n",
    "report_annotate = \"./data/STAR_cele_gene_count/report_annotate.txt\"\n",
    "\n",
    "df_gene = read.csv(df_gene, header=F)\n",
    "df_cell = read.csv(df_cell, header=F)\n",
    "gene_matrix = read.csv(gene_matrix, header=F)\n",
    "\n",
    "colnames(df_gene) = c(\"gene_id\", \"gene_type\", \"exon_intron\", \"gene_name\", \"index\")\n",
    "colnames(df_cell) = c(\"cell_name\", \"index\")\n",
    "gene_count = sparseMatrix(i = gene_matrix$V1, j = gene_matrix$V2, x = gene_matrix$V3)\n",
    "df_gene = df_gene[1:nrow(gene_count),]\n",
    "rownames(gene_count) = df_gene$V1\n",
    "\n",
    "# Count the species specific reads\n",
    "df_cell = df_cell[1:max(gene_matrix$V2),]\n",
    "df_cell = species_count(df_gene %>% filter(exon_intron == \"exon\"), df_cell, combine_exon_intron(df_gene, gene_count))\n",
    "rownames(gene_count) = df_gene$gene_name\n",
    "colnames(gene_count) = df_cell$cell_name\n",
    "\n",
    "# Read in the report matrix (including the reads number mapping to exons and introns)\n",
    "reportMM = read.csv(df_report, header=F)\n",
    "df_report = sparseMatrix(i = reportMM$V1, j = reportMM$V2, x = reportMM$V3)\n",
    "df_report = as.matrix(t(df_report))\n",
    "df_report_annotate = read.csv(report_annotate, header=F)\n",
    "colnames(df_report) = df_report_annotate$V2\n",
    "df_report = data.frame(df_report)\n",
    "df_report[\"index\"] = as.numeric(rownames(df_report))\n",
    "\n",
    "#add the report information to cell information\n",
    "df_cell_combine = inner_join(df_cell,df_report, by=\"index\")\n",
    "df_cell_combine[\"all_exon\"] = df_cell_combine$X.Perfect.intersect.exon.match + df_cell_combine$X.Nearest.intersect.exon.match + df_cell_combine$X.Perfect.combine.exon.match + df_cell_combine$X.Nearest.combine.exon.match\n",
    "df_cell_combine[\"all_intron\"] = df_cell_combine$X.Perfect.intersect.gene.match + df_cell_combine$X.Nearest.intersect.gene.match + df_cell_combine$X.Perfect.combine.gene.match + df_cell_combine$X.Nearest.combine.gene.match\n",
    "df_cell_combine[\"all_reads\"] = df_cell_combine$all_exon + df_cell_combine$all_intron + df_cell_combine$X.No.match\n",
    "df_cell_combine[\"unmatched_rate\"] = df_cell_combine$X.No.match / df_cell_combine$all_reads\n",
    "df_cell_combine[\"exon_rate\"] = df_cell_combine$all_exon / df_cell_combine$all_reads\n",
    "df_cell_combine[\"intron_rate\"] = df_cell_combine$all_intron / df_cell_combine$all_reads\n",
    "# check the reads distribution in different RNA types\n",
    "df_gene[\"total_reads\"] = rowSums(gene_count)\n",
    "df_read_distribution = df_gene %>% group_by(gene_type) %>% summarise(read_count = sum(total_reads))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "# plot the human and mouse distribution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#plot the read distribution for the dilute 25 nuclei\n",
    "g1 = (ggplot(data=df_cell, aes(x=human_reads, y=mouse_reads, colour=source))\n",
    "      + geom_point()\n",
    "      + labs(x=\"# of human UMIs\", y=\"# of mouse UMIs\")\n",
    "      + scale_colour_manual(values = c(chosen_color[7], chosen_color[1], chosen_color[6]))\n",
    "      + scale_x_continuous(limits=c(0, 30000))\n",
    "      + scale_y_continuous(limits=c(0, 30000))\n",
    "      + theme(legend.justification=c(1, 0), legend.position=c(1, 0.5)))\n",
    "g1\n",
    "save_plot(\"./pdf/hm_all.pdf\", g1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Monocle analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Construct the cds object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# provide df_cell, df_gene, UMI and construct the cds single cell object for downstream analysis\n",
    "\n",
    "pd = new(\"AnnotatedDataFrame\", data = df_cell)\n",
    "fd = new(\"AnnotatedDataFrame\", data = df_gene)\n",
    "\n",
    "dim(pd)\n",
    "dim(UMI)\n",
    "dim(fd)\n",
    "\n",
    "colnames(UMI) = df_cell$sample\n",
    "row.names(UMI) = df_gene$gene_id\n",
    "row.names(pd) = colnames(UMI)\n",
    "row.names(fd) = row.names(UMI) \n",
    "\n",
    "# for the family, use negbinomial.size() for gene count data, tobit() for TPM/FPKM, gaussianff() for log transformed data\n",
    "cds = newCellDataSet(UMI, \n",
    "                     phenoData = pd, \n",
    "                     featureData = fd, \n",
    "                     expressionFamily = negbinomial.size())\n",
    "\n",
    "save(cds, file = \"./data/cds.RData\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# select the cds with the pData\n",
    "cds = cds[, with(pData(cds), \n",
    "                 !is.na(cell.type) & cell.type == \"a\" & !(Cluster %in% c(5, 6, 8, 13)))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# estimate size and dispersions\n",
    "\n",
    "cds = estimateSizeFactors(cds)\n",
    "cds = estimateDispersions(cds)\n",
    "cds = detectGenes(cds, 0.1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Select genes and cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# select expressed genes using the proportion of cells expressed and size factor\n",
    "cds = cds[, pData(cds)$Size_Factor >= 0.5]\n",
    "fData(cds)$prop.expr = apply(exprs(cds), 1, function(x) sum(x > 0) / length(x))\n",
    "expressed.genes = rownames(subset(fData(cds), prop.expr >= 0.005))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# filter cells with low read count and also high read count\n",
    "pData(cds)$Total_mRNAs <- Matrix::colSums(exprs(cds))\n",
    "cds <- cds[,pData(cds)$Total_mRNAs < 1e6]\n",
    "upper_bound <- 10^(mean(log10(pData(cds)$Total_mRNAs)) + 2*sd(log10(pData(cds)$Total_mRNAs))) \n",
    "lower_bound <- 10^(mean(log10(pData(cds)$Total_mRNAs)) - 2*sd(log10(pData(cds)$Total_mRNAs))) \n",
    "g1 = (qplot(Total_mRNAs, data=pData(cds), geom=\"density\", colour=cell_type) \n",
    "      + geom_vline(xintercept=lower_bound) + geom_vline(xintercept=upper_bound)\n",
    "      + monocle:::monocle_theme_opts())\n",
    "g1\n",
    "\n",
    "cat(upper_bound)\n",
    "cat(\"\\n\")\n",
    "cat(lower_bound)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cds <- cds[,pData(cds)$Total_mRNAs > lower_bound &\n",
    "             pData(cds)$Total_mRNAs < upper_bound]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add cell type to the cells based on gene markers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "gene_1_id <- row.names(subset(fData(cds), gene_short_name == \"gene_1\"))\n",
    "gene_2_id <- row.names(subset(fData(cds), gene_short_name == \"gene_2\"))\n",
    "\n",
    "cth = newCellTypeHierarchy()\n",
    "cth <- addCellType(cth, \"cell_type_1\", classify_func=function(x) {x[gene_1_id,] >= 1})\n",
    "\n",
    "cth <- addCellType(cth, \"cell_type_2\", classify_func=function(x)\n",
    "       {x[gene_1_id,] < 1 & x[gene_2_id,] > 1})\n",
    "\n",
    "cds = classifyCells(cds, cth, 0.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pData(cds.bwm)$ap = with(pData(cds.bwm), ifelse(\n",
    "    tni.3 & !(cwn.1 | egl.20), \"Anterior\", ifelse(\n",
    "    !tni.3 & (cwn.1 | egl.20), \"Posterior\", NA)))\n",
    "\n",
    "pData(cds.bwm)$ap = factor(pData(cds.bwm)$ap, levels = c(\"Anterior\", \"Posterior\"))\n",
    "\n",
    "cds.bwm.ap = cds.bwm[, !is.na(pData(cds.bwm)$ap)]\n",
    "cds.bwm.ap = detectGenes(cds.bwm.ap, 0.1)\n",
    "summary(cds.bwm.ap$ap)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### filter genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# filter the highly expressed genes\n",
    "disp_table = dispersionTable(cds)\n",
    "unsup_clustering_genes = subset(disp_table, mean_expression >= 0.1)\n",
    "cds = setOrderingFilter(cds, unsup_clustering_genes$gene_id)\n",
    "plot_ordering_genes(cds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# filter the highly varied genes\n",
    "disp_table = dispersionTable(cds)\n",
    "ordering_genes <- subset(disp_table,\n",
    "                         mean_expression >= 0.5 &\n",
    "                         dispersion_empirical >= 1 * dispersion_fit)$gene_id\n",
    "cds <- setOrderingFilter(cds, ordering_genes)\n",
    "plot_ordering_genes(cds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# filter the covaried marker genes\n",
    "marker_diff <- markerDiffTable(cds[expressed_genes,],\n",
    "                                 cth,\n",
    "                                 residualModelFormulaStr=\"~Media + num_genes_expressed\",\n",
    "                                 cores=1)\n",
    "\n",
    "candidate_clustering_genes <- row.names(subset(marker_diff, qval < 0.01)) \n",
    "marker_spec <- calculateMarkerSpecificity(HSMM[candidate_clustering_genes,], cth)\n",
    "\n",
    "semisup_clustering_genes <- unique(selectTopMarkers(marker_spec, 500)$gene_id)\n",
    "cds <- setOrderingFilter(cds, semisup_clustering_genes)\n",
    "plot_ordering_genes(cds)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PCA analysis for the genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "plot_pc_variance_explained(cds, return_all = F)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### dimension reduction and cluster the cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# reduce dimension, choose the normalization method and reduction method and dimensions for reduction\n",
    "set.seed(123)\n",
    "cds = reduceDimension(\n",
    "    cds, max_components = 2, num_dim = 40, norm_method = \"log\",\n",
    "    reduction_method = 'tSNE', \n",
    "    residualModelFormulaStr = \"~ experiment + num_genes_expressed\", verbose = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# plot the result after reduce dimension\n",
    "pData(cds)$tsne_1 = reducedDimA(cds)[1,]\n",
    "pData(cds)$tsne_2 = reducedDimA(cds)[2,]\n",
    "\n",
    "g1 = (ggplot(data=pData(cds), aes(x=tsne_1, y=tsne_2, colour = as.factor(pData(cds)$experiment))) \n",
    "      + geom_point()\n",
    "      + scale_color_manual(name=\"Experiment\", values = c(chosen_color[7], chosen_color[6]))\n",
    "      + labs(x=\"tSNE 1\", y=\"tSNE 2\"))\n",
    "g1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# cluster the cells and choose the good rho and delta threshold, and then plot the clusters\n",
    "cds = clusterCells(cds, num_clusters = 2)\n",
    "plot_rho_delta(cds, rho_threshold = 30, delta_threshold = 2)\n",
    "plot_cell_clusters(cds, cell_size = 0.2, color_by = \"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cds = clusterCells_Density_Peak(cds, rho_threshold = 30, delta_threshold = 3, skip_rho_sigma = T)\n",
    "plot_cell_clusters(cds, cell_size = 0.2) + facet_wrap(~CellType)\n",
    "plot_cell_clusters(cds, color = \"Cluster %in% c(1, 4, 7, 10, 11, 14)\", cell_size = 0.333)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "plot = ggplot(pData(cds), aes(x = tsne_1, y = tsne_2, colour = Cluster)) +\n",
    "    geom_point(size = 0.5) +\n",
    "    monocle:::monocle_theme_opts()\n",
    "\n",
    "plot = ggplot(pData(cds.experiment.1), aes(x = tsne_1, y = tsne_2, color = cluster.name)) +\n",
    "    geom_point(size = 0.002) +\n",
    "    xlab(\"\") + ylab(\"\") +\n",
    "    guides(color = guide_legend(\n",
    "        title = \"Cell type\\n(aggregated)\",\n",
    "        override.aes = list(size = 4))) +\n",
    "    theme_void() +\n",
    "    monocle:::monocle_theme_opts() +\n",
    "    theme(legend.title = element_text(size=6),\n",
    "          legend.text = element_text(size=6),\n",
    "          legend.margin = margin(0, -10, 0, 10),\n",
    "          legend.key.width=unit(0.15, \"in\"),\n",
    "          legend.key.height=unit(0.15, \"in\"),\n",
    "          legend.position = \"left\")\n",
    "\n",
    "\n",
    "plot = ggplot(pData(cds.hyp), aes(x = tsne_1, y = tsne_2, color = plot.group, alpha = plot.group)) +\n",
    "    facet_wrap(~ experiment, ncol = 2, scales = \"free\") +\n",
    "    geom_point(size = 0.04) +\n",
    "    scale_color_manual(values =\n",
    "        c(\"firebrick\", \"indianred3\", \"darkorange2\", \"forestgreen\", \"dodgerblue3\", \"grey70\")) +\n",
    "    scale_alpha_manual(values = c(1, 1, 1, 1, 1, 1.0/3.0)) +\n",
    "    guides(color = guide_legend(title = \"Marker gene\", override.aes = list(size = 3)), alpha = F) +\n",
    "    theme_void(base_size = 6) +\n",
    "    monocle:::monocle_theme_opts() +\n",
    "    theme(\n",
    "        legend.position = \"left\",\n",
    "        strip.text = element_text(size = 6),\n",
    "        legend.title = element_text(size = 6),\n",
    "        legend.text = element_text(size = 6),\n",
    "        legend.margin = margin(0, -10, 0, 10),\n",
    "        legend.key.width = unit(0.15, \"in\"),\n",
    "        legend.key.height = unit(0.15, \"in\"))\n",
    "\n",
    "\n",
    "ggsave(\"./pdf/clusters.pdf\", device = cairo_pdf, plot = plot, units = \"in\", width = 7.12, height = 5.7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# recover the tsne location for a subset of cells\n",
    "\n",
    "cds.subset@reducedDimA = cds@reducedDimA[,\n",
    "    with(pData(cds), !is.na(cell.type) & cell.type == \"A\")]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### check the UMIs in different cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "ggplot(pData(cds), aes(x = tsne_1, y = tsne_2, color = ifelse(n.umi >= 200, 1, NA))) +\n",
    "    geom_point(size = 0.1) +\n",
    "    monocle:::monocle_theme_opts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### DE gene analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# check differential expressed genes between different clusters\n",
    "DE.genes = differentialGeneTest(cds[expressed.genes,],\n",
    "                                fullModelFormulaStr = \"~ plate + Cluster\",\n",
    "                                reducedModelFormulaStr = \"~ plate\",\n",
    "                                cores = 4)\n",
    "\n",
    "# check differential expressed genes between semi-labeled cells\n",
    "\n",
    "\n",
    "\n",
    "# check differential expressed genes among the psudo-time\n",
    "diff_test_res <- differentialGeneTest(cds, \n",
    "                                      fullModelFormulaStr=\"~sm.ns(Pseudotime)\")\n",
    "\n",
    "# multi-factor analysis\n",
    "diff_test_res <- differentialGeneTest(cds,\n",
    "                                      fullModelFormulaStr=\"~CellType + Hours\",\n",
    "                                      reducedModelFormulaStr=\"~Hours\")\n",
    "\n",
    "# brached gene differential analysis ? is this different from the branch?\n",
    "diff_test_res <- differentialGeneTest(cds,\n",
    "                                      fullModelFormulaStr=\"~sm.ns(Pseudotime) + Branch + sm.ns(Pseudotime) : Branch\",\n",
    "                                      reducedModelFormulaStr=\"~sm.ns(Pseudotime)\")\n",
    "BEAM_res <- BEAM(lung, branch_point=1, cores = 1)\n",
    "BEAM_res <- BEAM_res[order(BEAM_res$qval),]\n",
    "BEAM_res <- BEAM_res[,c(\"gene_short_name\", \"pval\", \"qval\")]\n",
    "\n",
    "# Alternatively, can call the model separately\n",
    "full_model_fits <- fitModel(cds,  modelFormulaStr=\"~CellType\")\n",
    "reduced_model_fits <- fitModel(cds, modelFormulaStr=\"~1\")\n",
    "diff_test_res <- compareModels(full_model_fits, reduced_model_fits)\n",
    "\n",
    "# find the DE genes\n",
    "ordering_genes = row.names(subset(DE.genes, qval < 0.1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# identify DE genes\n",
    "\n",
    "# here I define a function that accept a DE.gene list and then return the transformed q value for each gene\n",
    "filter_p_value <- function(DE.genes, alpha)\n",
    "    {\n",
    "    library(genefilter)\n",
    "    theta = seq(from = 0, to = 0.8, by = 0.02)\n",
    "    result = filtered_R(alpha = alpha, \n",
    "                        filter = DE.genes$prop.expr, \n",
    "                        test = DE.genes$pval, \n",
    "                        theta = theta, method = \"BH\")\n",
    "    adj_p = filtered_p(filter = DE.genes$prop.expr, \n",
    "                       test = DE.genes$pval, \n",
    "                       theta = theta[which(result == max(result))[1]], method = \"BH\")\n",
    "    adj_p[is.na(adj_p)] = 1\n",
    "    return(adj_p)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### plot DE genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "library(\"pheatmap\")\n",
    "\n",
    "sig.genes = (DE.genes %>% arrange(qval) %>% filter(qval <= 0.05)) %>% head(20)\n",
    "cds_filter = cds[sig.genes$gene_id,]\n",
    "\n",
    "plot.matrix = t(t(exprs(cds_filter)) / pData(cds_filter)$Size_Factor)\n",
    "\n",
    "plot.matrix = t(scale(t(plot.matrix), center = T))\n",
    "\n",
    "row.names(plot.matrix) = fData(cds_filter)$gene_short_name\n",
    "\n",
    "pData(cds_filter) = pData(cds_filter) %>% mutate(Tissue = ifelse(condition == 3, \"Lung\", \"Brain\"))\n",
    "\n",
    "anno = pData(cds_filter) %>% select(Tissue, Cluster)\n",
    "\n",
    "row.names(anno) = colnames(plot.matrix)\n",
    "\n",
    "pheatmap(plot.matrix, show_rownames = T, show_colnames = F, \n",
    "         annotation_col = anno,scale = \"row\", file = \"./pdf/gene_heatmap.pdf\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DE gene and fold change analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# include DE gene fold change in two cell types\n",
    "\n",
    "# DE gene analysis\n",
    "BWM.AP.DE.genes = differentialGeneTest(cds.bwm.ap[expressed.genes,],\n",
    "    fullModelFormulaStr = \"~ ap\")\n",
    "\n",
    "BWM.AP.norm.means = sapply(c(\"Anterior\", \"Posterior\"), function(x) {\n",
    "    rowMeans(norm.expr[, pData(cds)$cell %in% subset(pData(cds.bwm.ap), ap == x)$cell])\n",
    "})\n",
    "\n",
    "rownames(BWM.AP.norm.means) = fData(cds)$symbol\n",
    "\n",
    "# add gene expression fold change between different cell types\n",
    "\n",
    "BWM.AP.DE.genes = inner_join(BWM.AP.DE.genes, data.frame(\n",
    "    symbol = row.names(BWM.AP.norm.means), \n",
    "    anterior.tpm = (BWM.AP.norm.means[, \"Anterior\"] / sum(BWM.AP.norm.means[, \"Anterior\"])) * 1000000,\n",
    "    posterior.tpm = (BWM.AP.norm.means[, \"Posterior\"] / sum(BWM.AP.norm.means[, \"Posterior\"])) * 1000000\n",
    "), by = \"symbol\") %>% mutate(\n",
    "    A.P.ratio = anterior.tpm / (posterior.tpm + 1),\n",
    "    P.A.ratio = posterior.tpm / (anterior.tpm + 1),\n",
    "    log2.fc = ifelse(A.P.ratio > 1, log2(A.P.ratio), -log2(P.A.ratio))) %>% select(\n",
    "    gene_id, symbol, qval, anterior.tpm, posterior.tpm, log2.fc)\n",
    "\n",
    "BWM.AP.DE.genes$sig.group = with(BWM.AP.DE.genes, ifelse(\n",
    "    qval < 0.05 & abs(log2.fc) >= log2(3), \"Significant, FC >= 3\", ifelse(\n",
    "    qval < 0.05, \"Significant, FC < 3\", \"Not significant\")))\n",
    "\n",
    "BWM.AP.DE.genes$sig.group = factor(BWM.AP.DE.genes$sig.group, levels = c(\n",
    "    \"Significant, FC >= 3\", \"Significant, FC < 3\", \"Not significant\"))\n",
    "\n",
    "summary(BWM.AP.DE.genes$sig.group)\n",
    "\n",
    "# plot the volcanol plot\n",
    "plot = ggplot(BWM.AP.DE.genes, aes(x = log2.fc, y = pmin(100, -log10(qval)), color = sig.group)) +\n",
    "    geom_point(size = 0.2) +\n",
    "    scale_color_manual(values = c(\"firebrick\", \"dodgerblue3\", \"black\")) +\n",
    "    xlab(\"log2(anterior BWM / posterior BWM)\") +\n",
    "    ylab(\"-log10(q-value)\") +\n",
    "    guides(color = F) +\n",
    "    theme_bw(base_size = 6) +\n",
    "    monocle:::monocle_theme_opts()\n",
    "\n",
    "show(plot)\n",
    "\n",
    "ggsave(\"plots/Fig_S8b.pdf\", plot = plot,\n",
    "    device = cairo_pdf, units = \"in\", width = 2.2, height = 2.2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### identify cluster specific genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "get.cluster.markers = function(cds) {    \n",
    "    cluster_means = lapply(1:length(unique(pData(cds)$Cluster)),\n",
    "                           function(x) rowMeans(exprs(cds)[, pData(cds)$Cluster == x]))\n",
    "    cluster_means = t(do.call(rbind, cluster_means))\n",
    "    colnames(cluster_means) = 1:length(unique(pData(cds)$Cluster))\n",
    "    specificity_scores = apply(cluster_means, 1, function(x) 1 - sum(x >= 0.1*max(x))/length(x))\n",
    "    cluster_marker_scores = sweep(cluster_means, 1, specificity_scores^3, \"*\")\n",
    "    \n",
    "    cluster_marker_df = melt(cluster_marker_scores)\n",
    "    names(cluster_marker_df) = c(\"gene_id\", \"cluster\", \"score\")\n",
    "    cluster_marker_df = inner_join(cluster_marker_df, fData(cds)[, 1:3], by= \"gene_id\") %>%\n",
    "        group_by(cluster) %>% arrange(-score) %>% mutate(rank = rank(-score)) %>%\n",
    "        filter(rank <= 100) %>% arrange(cluster, rank)\n",
    "        \n",
    "    return(cluster_marker_df)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### set cluster name and quick functions for plotting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "plot.clusters = function(cds, cell_size = 0.2) {\n",
    "    ggplot(pData(cds), aes(x = tsne_1, y = tsne_2, color = cluster.name)) +\n",
    "        geom_point(size = cell_size) + monocle:::monocle_theme_opts() + theme(legend.position = \"left\")\n",
    "}\n",
    "\n",
    "plot.tissues = function(cds, cell_size = 0.2) {\n",
    "    ggplot(pData(cds), aes(x = tsne_1, y = tsne_2, color = tissue)) +\n",
    "        geom_point(size = cell_size) + monocle:::monocle_theme_opts() + theme(legend.position = \"left\")\n",
    "}\n",
    "\n",
    "plot.cell.types = function(cds, cell_size = 0.2) {\n",
    "    ggplot(pData(cds), aes(x = tsne_1, y = tsne_2, color = cell.type)) +\n",
    "        geom_point(size = cell_size) + monocle:::monocle_theme_opts() + theme(legend.position = \"left\")\n",
    "}\n",
    "\n",
    "plot.fine.grained.neuron.types = function(cds, cell_size = 0.2) {\n",
    "    ggplot(pData(cds), aes(x = tsne_1, y = tsne_2, color = fine.grained.neuron.type)) +\n",
    "        geom_point(size = cell_size) + monocle:::monocle_theme_opts() + theme(legend.position = \"left\")\n",
    "}\n",
    "\n",
    "set.cluster.name = function(cds, cluster.name, logical.vec) {\n",
    "    pData(cds)$cluster.name = ifelse(logical.vec, cluster.name, pData(cds)$cluster.name)\n",
    "    return(cds)\n",
    "}\n",
    "\n",
    "set.tissue = function(cds, tissue, logical.vec) {\n",
    "    pData(cds)$tissue = ifelse(logical.vec, tissue, pData(cds)$tissue)\n",
    "    return(cds)\n",
    "}\n",
    "\n",
    "set.cell.type = function(cds, cell.type, logical.vec) {\n",
    "    pData(cds)$cell.type = ifelse(logical.vec, cell.type, pData(cds)$cell.type)\n",
    "    return(cds)\n",
    "}\n",
    "\n",
    "set.fine.grained.neuron.type = function(cds, fine.grained.neuron.type, logical.vec) {\n",
    "    pData(cds)$fine.grained.neuron.type =\n",
    "        ifelse(logical.vec, fine.grained.neuron.type, pData(cds)$fine.grained.neuron.type)\n",
    "    return(cds)\n",
    "}\n",
    "\n",
    "get.gene.id = function(cds, symbol) {\n",
    "    return(as.character(fData(cds)[fData(cds)$symbol == symbol, \"gene\"]))\n",
    "}\n",
    "\n",
    "expresses.gene = function(cds, symbol, thresh = 1) {\n",
    "    return(exprs(cds)[get.gene.id(cds, symbol), ] >= thresh)\n",
    "}\n",
    "\n",
    "pData(cds)$cluster.name = rep(NA, ncol(cds))\n",
    "pData(cds)$tissue = rep(NA, ncol(cds))\n",
    "pData(cds)$cell.type = rep(NA, ncol(cds))\n",
    "\n",
    "\n",
    "cds = set.cell.type(cds, \"Pharyngeal muscle\",\n",
    "    pData(cds)$Cluster == 11 &\n",
    "    expresses.gene(cds, \"myo-2\") + expresses.gene(cds, \"myo-1\") +\n",
    "    expresses.gene(cds, \"myo-5\") + expresses.gene(cds, \"tnt-4\") +\n",
    "    expresses.gene(cds, \"mlc-1\") + expresses.gene(cds, \"mlc-2\") >= 2)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### plot gene expression in the tsne plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# plot and check one marker\n",
    "plot.expr = function (cds, gene, thresh = 1)  {\n",
    "    pData(cds)$tmp = exprs(cds)[gene, ] >= thresh\n",
    "    plot = ggplot(pData(cds), aes(x = tsne_1, y = tsne_2, color = tmp)) + \n",
    "        geom_point(size = 0.2) + scale_color_manual(values = c(\"grey50\", \n",
    "        \"#F8766D\")) + guides(color = guide_legend(title = gene)) + \n",
    "        monocle:::monocle_theme_opts() + theme(legend.position = \"top\")\n",
    "    pData(cds)$tmp = NULL\n",
    "    return(plot)\n",
    "}\n",
    "\n",
    "# plot and check two markers\n",
    "\n",
    "plot.coexpr.2 = function (cds, g1, g2, g3 = NULL, thresh = 1, cell_size = 0.2) \n",
    "{\n",
    "    g1.id = g1\n",
    "    g2.id = g2\n",
    "    pData(cds)$tmp1 = exprs(cds)[g1.id, ] >= thresh\n",
    "    pData(cds)$tmp2 = exprs(cds)[g2.id, ] >= thresh\n",
    "    if (!is.null(g3)) {\n",
    "        g3.id = g3\n",
    "        pData(cds)$tmp3 = exprs(cds)[g3.id, ] >= thresh\n",
    "    }\n",
    "    if (is.null(g3)) {\n",
    "        pData(cds)$coexpr = with(pData(cds), ifelse(tmp1 & tmp2, \n",
    "            paste(g1, g2, sep = \" and \"), ifelse(tmp1, paste(g1, \n",
    "                \"only\"), ifelse(tmp2, paste(g2, \"only\"), NA))))\n",
    "        pData(cds)$coexpr = factor(pData(cds)$coexpr, levels = c(paste(g1, \n",
    "            g2, sep = \" and \"), paste(g1, \"only\"), paste(g2, \n",
    "            \"only\")))\n",
    "        cat(\"# expressing both:\", with(pData(cds), sum(tmp1 & \n",
    "            tmp2)), \"\\n\")\n",
    "    }\n",
    "    else {\n",
    "        pData(cds)$coexpr = factor(with(pData(cds), tmp1 + tmp2 + \n",
    "            tmp3), levels = c(1, 2, 3))\n",
    "        cat(\"# expressing all:\", with(pData(cds), sum(tmp1 & \n",
    "            tmp2 & tmp3)), \"\\n\")\n",
    "    }\n",
    "    plot = ggplot(pData(cds), aes(x = tsne_1, y = tsne_2, color = coexpr)) + \n",
    "        geom_point(size = cell_size) + monocle:::monocle_theme_opts() + \n",
    "        theme(legend.position = \"top\")\n",
    "    pData(cds)$tmp1 = NULL\n",
    "    pData(cds)$tmp2 = NULL\n",
    "    pData(cds)$tmp3 = NULL\n",
    "    pData(cds)$coexpr = NULL\n",
    "    return(plot)\n",
    "}\n",
    "\n",
    "\n",
    "# plot and check multiple markers\n",
    "\n",
    "pData(cds.bwm)$tni.3 = expresses.gene(cds.bwm, \"tni-3\")\n",
    "pData(cds.bwm)$cwn.1 = expresses.gene(cds.bwm, \"cwn-1\")\n",
    "pData(cds.bwm)$egl.20 = expresses.gene(cds.bwm, \"egl-20\")\n",
    "\n",
    "pData(cds.bwm)$marker.gene = with(pData(cds.bwm), ifelse(\n",
    "    tni.3 + cwn.1 + egl.20 > 1, \"Multiple\", ifelse(\n",
    "    tni.3, \"tni-3\", ifelse(\n",
    "    cwn.1, \"cwn-1\", ifelse(\n",
    "    egl.20, \"egl-20\", \"None\")))))\n",
    "\n",
    "pData(cds.bwm)$marker.gene = factor(pData(cds.bwm)$marker.gene, levels = c(\n",
    "    \"tni-3\", \"cwn-1\", \"egl-20\", \"Multiple\", \"None\"))\n",
    "plot = ggplot(pData(cds.bwm), aes(x = tsne_1, y = tsne_2, color = marker.gene, alpha = marker.gene)) +\n",
    "    geom_point(size = 0.1) +\n",
    "    scale_color_manual(values = c(\"firebrick\", \"forestgreen\", \"dodgerblue3\", \"#984EA3\", \"grey70\")) +\n",
    "    scale_alpha_manual(values = c(1, 1, 1, 1, 1.0/3.0)) +\n",
    "    guides(\n",
    "        color = guide_legend(title = \"Marker gene\", override.aes = list(size = 3)),\n",
    "        alpha = F) +\n",
    "    theme_void() +\n",
    "    monocle:::monocle_theme_opts() +\n",
    "    theme(\n",
    "        axis.line = element_blank(),\n",
    "        legend.position = \"left\",\n",
    "        legend.title = element_text(size=6),\n",
    "        legend.text = element_text(size=6),\n",
    "        legend.margin = margin(0, -10, 0, 10),\n",
    "        legend.key.width = unit(0.15, \"in\"),\n",
    "        legend.key.height = unit(0.15, \"in\"))\n",
    "\n",
    "show(plot)\n",
    "\n",
    "ggsave(\"plots/Fig_S8a.pdf\", plot = plot,\n",
    "    device = cairo_pdf, units = \"in\", width = 3, height = 2.5)\n",
    "\n",
    "\n",
    "# plot gene expression panal\n",
    "\n",
    "plot.expr.panels = function(cds, genes, thresh=1, ncol=4, cell_size=0.2) {\n",
    "    df = pData(cds)[, c(\"cell\", \"tsne_1\", \"tsne_2\")]\n",
    "    gene.ids = sapply(genes, function(g) get.gene.id(cds, g))\n",
    "    for (i in 1:length(genes)) {\n",
    "        df[, sub(\"-\", \"_\", genes[i])] = exprs(cds)[gene.ids[i],] > 0\n",
    "    }\n",
    "        \n",
    "    df = melt(df, id.vars=c(\"cell\", \"tsne_1\", \"tsne_2\"), variable.name=\"gene\", value.name=\"expr\")\n",
    "    df$gene = sub(\"_\", \"-\", df$gene)\n",
    "\n",
    "    plot = ggplot(df, aes(x = tsne_1, y = tsne_2, color = ifelse(expr, 1, NA))) +\n",
    "        facet_wrap(~ gene, ncol=ncol) +\n",
    "        geom_point(size = cell_size) +\n",
    "        guides(color = F) +\n",
    "        monocle:::monocle_theme_opts()\n",
    "    return(plot)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### jitter plot for the gene expression for different groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# jitter plot for the gene expression for different groups\n",
    "gene_list <- cds[row.names(subset(fData(cds),\n",
    "                                      gene_short_name %in% c(\"gene_1\", \"gene_2\"))),]\n",
    "plot_genes_jitter(gene_list, grouping=\"Cluster\", color_by = \"CellType\", ncol=2)\n",
    "\n",
    "plot_genes_jitter(cds_subset,\n",
    "                  grouping=\"Hours\", color_by=\"CellType\", plot_trend=TRUE) + facet_wrap( ~ feature_label, scales=\"free_y\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### plot gene change in psudotime"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# plot several genes in psudotime\n",
    "my_genes <- row.names(subset(fData(cds),\n",
    "                             gene_short_name %in% c(\"CDK1\", \"MEF2C\", \"MYH3\")))\n",
    "cds_subset <- cds[my_genes,]\n",
    "\n",
    "\n",
    "plot_genes_in_pseudotime(cds_subset, color_by=\"Hours\")\n",
    "\n",
    "# plot several genes in one plot\n",
    "qork = melt(pData(cds.hyp.2.trajectory) %>%\n",
    "    select(Pseudotime, expr.qua.1, expr.sqt.1, expr.bcl.11, expr.col.12),\n",
    "    id.vars = c(\"Pseudotime\"), variable.name = \"gene\", value.name = \"value\") %>%\n",
    "    group_by(gene) %>% mutate(value = (value - mean(value)) / sd(value)) %>% ungroup() \n",
    "\n",
    "plot = ggplot(qork, aes(x = Pseudotime, y = value, color = gene)) +\n",
    "    geom_smooth(se = F) +\n",
    "    xlab(\"Pseudotime\") +\n",
    "    ylab(\"Expression z-score\") +\n",
    "    scale_color_brewer(palette = \"Set1\", labels = c(\"qua-1\", \"sqt-1\", \"bcl-11\", \"col-12\")) +\n",
    "    guides(color = guide_legend(title = \"Gene\")) +\n",
    "    monocle:::monocle_theme_opts()\n",
    "\n",
    "show(plot)\n",
    "\n",
    "ggsave(\"plots/seam.cell.trajectory.marker.genes.pdf\",\n",
    "    device = cairo_pdf, plot = plot, units = \"in\", width = 4.5, height = 3.5)\n",
    "\n",
    "\n",
    "\n",
    "# plot gene change in different braches\n",
    "my_genes <- row.names(subset(fData(cds),\n",
    "                             gene_short_name %in% c(\"CDK1\", \"MEF2C\", \"MYH3\")))\n",
    "cds_subset <- cds[my_genes,]\n",
    "\n",
    "plot_genes_branched_pseudotime(cds_subset,\n",
    "                               branch_point=1,\n",
    "                               color_by=\"Hours\",\n",
    "                               ncol=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### plot gene cluster heatmap in psudotime"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# plot gene cluster in psudotime\n",
    "cds.tra.heatmap = plot_pseudotime_heatmap(cds[DE_gene_names,],\n",
    "                        num_clusters = 3,\n",
    "                        cores = 1,\n",
    "                        show_rownames = T)\n",
    "\n",
    "\n",
    "# identify gene names in different cluster\n",
    "cds.trajectory.DE.gene.clusters = cutree(cds.tra.heatmap$tree_row, 3)\n",
    "cds.trajectory.DE.genes %>% filter(gene_id %in% names(which(cds.trajectory.DE.gene.clusters == 7))) %>%\n",
    "    arrange(qval) %>% head(10)\n",
    "\n",
    "# plot the gene cluster for different branches in psudotime\n",
    "plot_genes_branched_heatmap(cds[row.names(subset(BEAM_res, qval < 1e-4)),], branch_point = 1,\n",
    "                            num_clusters = 4,\n",
    "                            cores = 1,\n",
    "                            use_gene_short_name = T,\n",
    "                            show_rownames = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### generate cell trajectory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# set the order\n",
    "cds = setOrderingFilter(cds,\n",
    "    ((DE.genes %>% arrange(qval) %>% head(200))$gene_id))\n",
    "\n",
    "# construct the DDR tree\n",
    "\n",
    "system.time({\n",
    "set.seed(42)\n",
    "cds = reduceDimension(cds, max_components = 2,\n",
    "    residualModelFormulaStr = \"~ sm.ns(Size_Factor, df = 3) + plate\",\n",
    "    auto_param_selection = F, ncenter = 420)\n",
    "})\n",
    "\n",
    "\n",
    "# define the root state\n",
    "\n",
    "GM_state <- function(cds){\n",
    "  if (length(unique(pData(cds)$State)) > 1){\n",
    "    T0_counts <- table(pData(cds)$State, pData(cds)$Hours)[,\"0\"]\n",
    "return(as.numeric(names(T0_counts)[which(T0_counts == max(T0_counts))])) }else { return (1) }\n",
    "}\n",
    "\n",
    "# order cells\n",
    "cds = orderCells(cds)\n",
    "cds = orderCells(cds, root_state = GM_state(cds))\n",
    "plot_cell_trajectory(cds, color_by = \"State/cell_type/Pseudotime\")\n",
    "\n",
    "# plot the DDR tree with gene markers\n",
    "# plot the DDR tree with gene markers\n",
    "pData(cds)$DDRTree_1 = cds@reducedDimS[1,]\n",
    "pData(cds)$DDRTree_2 = cds@reducedDimS[2,]\n",
    "\n",
    "plot = ggplot(pData(cds), aes(x = DDRTree_1, y = DDRTree_2, color = plot.group)) +\n",
    "    geom_point(aes(alpha = plot.group %in% c(\"None\", \"Multiple\")), size = 0.5) +\n",
    "    scale_color_manual(values = c(\"#9E0142\", \"goldenrod3\", \"forestgreen\", \"#3288BD\", \"#5E4FA2\", \"grey20\", \"grey80\")) +\n",
    "    scale_alpha_manual(values = c(1.0, 0.3333)) +\n",
    "    guides(color = guide_legend(title = \"Marker gene\", override.aes = list(size=3)), alpha = F) +\n",
    "    monocle:::monocle_theme_opts()\n",
    "\n",
    "show(plot)\n",
    "\n",
    "# cluster cells along the trajectory\n",
    "plot_cell_trajectory(cds.hyp.3.trajectory, cell_size = 0.333, color_by = \"factor(round(Pseudotime/2)*2)\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### semi-supervised ordering with know marker genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# classify different cells \n",
    "\n",
    "gene_1_id <- row.names(subset(fData(cds), gene_short_name == \"gene_1\"))\n",
    "gene_2_id <- row.names(subset(fData(cds), gene_short_name == \"gene_2\"))\n",
    "\n",
    "cth = newCellTypeHierarchy()\n",
    "cth <- addCellType(cth, \"cell_type_1\", classify_func=function(x) {x[gene_1_id,] >= 1})\n",
    "\n",
    "cth <- addCellType(cth, \"cell_type_2\", classify_func=function(x)\n",
    "       {x[gene_1_id,] < 1 & x[gene_2_id,] > 1})\n",
    "\n",
    "cds = classifyCells(cds, cth, 0.1)\n",
    "\n",
    "# filter the covaried marker genes\n",
    "marker_diff <- markerDiffTable(cds[expressed_genes,],\n",
    "                                 cth,\n",
    "                                 residualModelFormulaStr=\"~Media + num_genes_expressed\",\n",
    "                                 cores=1)\n",
    "\n",
    "candidate_clustering_genes <- row.names(subset(marker_diff, qval < 0.01)) \n",
    "marker_spec <- calculateMarkerSpecificity(HSMM[candidate_clustering_genes,], cth)\n",
    "\n",
    "semisup_clustering_genes <- unique(selectTopMarkers(marker_spec, 500)$gene_id)\n",
    "cds <- setOrderingFilter(cds, semisup_clustering_genes)\n",
    "plot_ordering_genes(cds)\n",
    "\n",
    "# dimension reduction and plot cell trajectory in the same way."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "?plot_genes_branched_pseudotime"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### impute the cell type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cds = clusterCells(cds,\n",
    "                  num_clusters = 2,\n",
    "                  frequency_thresh = 0.1,\n",
    "                  cell_type_hierarchy = cth)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "plot_cell_clusters(cds, 1, 2, color = \"CellType\", markers = c(\"gene_1_id\", \"gene_2_id\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### plot the cells based on expressed genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# plot the expression of a single gene\n",
    "plot.expr(cds, \"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# annotate the cells based on expressed genes\n",
    "pData(cds)$plot.group = ifelse(\n",
    "    expresses.gene(cds, \"a\") + expresses.gene(cds, \"b\") +\n",
    "    expresses.gene(cds, \"c\") + expresses.gene(cds, \"d\") +\n",
    "    expresses.gene(cds, \"e\") > 1, \"Multiple\", ifelse(\n",
    "    expresses.gene(cds, \"a\"), \"a\", ifelse(\n",
    "    expresses.gene(cds, \"b\"), \"b\", ifelse(\n",
    "    expresses.gene(cds, \"c\"), \"c\", ifelse(\n",
    "    expresses.gene(cds, \"d\"), \"d\", ifelse(\n",
    "    expresses.gene(cds, \"e\"), \"e\", \"None\"))))))\n",
    "\n",
    "pData(cds)$plot.group = factor(pData(cds)$plot.group, levels =\n",
    "    c(\"a\", \"b\", \"c\", \"d\", \"e\", \"Multiple\", \"None\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# plot the groups\n",
    "plot = ggplot(pData(cds), aes(x = tsne_1, y = tsne_2, colour = plot.group)) +\n",
    "    geom_point(aes(alpha = plot.group %in% c(\"None\", \"Multiple\")), size = 0.5) +\n",
    "    #scale_color_manual(values = c(\"#F8766D\", \"#A3A500\", \"#00BF7D\", \"#00B0F6\", \"#E76BF3\", \"grey20\", \"grey80\")) +\n",
    "    scale_color_manual(values = c(\"#9E0142\", \"goldenrod3\", \"forestgreen\", \"#3288BD\", \"#5E4FA2\", \"grey20\", \"grey80\")) +\n",
    "    scale_alpha_manual(values = c(1.0, 0.3333)) +\n",
    "    guides(color = guide_legend(title = \"Marker gene\", override.aes = list(size=3)), alpha = F) +\n",
    "    monocle:::monocle_theme_opts()\n",
    "\n",
    "show(plot)\n",
    "\n",
    "ggsave(\"pdf/result.pdf\",\n",
    "    device = cairo_pdf, plot = plot, units = \"in\", width = 7.4, height = 5.7)\n",
    "\n",
    "ggsave(\"pdf/result.png\",\n",
    "    plot = plot, units = \"in\", width = 7.4, height = 5.7, dpi = 300)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### set directions and identify differential expressed genes and trajectory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Define the psuedo-space\n",
    "centroids = list()\n",
    "\n",
    "centroids$Cluster.1 = pData(cds)[pData(cds)$Cluster == 1,] %>%\n",
    "    group_by(1) %>% summarize(tsne.1 = median(tsne_1), tsne.2 = median(tsne_2)) %>% select(tsne.1, tsne.2)\n",
    "\n",
    "centroids$Cluster.2 = pData(cds[pData(cds)$Cluster == 2,] %>%\n",
    "    group_by(1) %>% summarize(tsne.1 = median(tsne_1), tsne.2 = median(tsne_2)) %>% select(tsne.1, tsne.2)\n",
    "\n",
    "plot = ggplot(pData(cds), aes(x = tsne_1, y = tsne_2, colour = plot.group)) +\n",
    "    geom_point(aes(alpha = plot.group %in% c(\"None\", \"Multiple\")), size = 0.5) +\n",
    "    geom_point(x = centroids$Cluster.1$tsne.1, y = centroids$Cluster.1$tsne.2, color = \"black\", size = 1.5) +\n",
    "    geom_point(x = centroids$Cluster.2$tsne.1, y = centroids$Cluster.2$tsne.2, color = \"black\", size = 1.5) +\n",
    "    geom_segment(x = centroids$Cluster.1$tsne.1, xend = centroids$Cluster.7$tsne.1,\n",
    "                 y = centroids$Cluster.1$tsne.2, yend = centroids$Cluster.7$tsne.2,\n",
    "                 color = \"black\", size = 0.75) +\n",
    "    scale_color_manual(values = c(\"#9E0142\", \"goldenrod3\", \"forestgreen\", \"#3288BD\", \"#5E4FA2\", \"grey20\", \"grey80\")) +\n",
    "    scale_alpha_manual(values = c(1.0, 0.3333)) +\n",
    "    guides(color = guide_legend(title = \"Marker gene\", override.aes = list(size=3)), alpha = F) +\n",
    "    monocle:::monocle_theme_opts()\n",
    "\n",
    "show(plot)\n",
    "\n",
    "# generate the start and end position\n",
    "centroids = list()\n",
    "\n",
    "centroids$Cluster.1 = pData(cds)[pData(cds)$Cluster == 1,] %>%\n",
    "    group_by(1) %>% summarize(tsne.1 = median(tsne_1), tsne.2 = median(tsne_2)) %>% select(tsne.1, tsne.2)\n",
    "\n",
    "centroids$Cluster.2 = pData(cds[pData(cds)$Cluster == 2,] %>%\n",
    "    group_by(1) %>% summarize(tsne.1 = median(tsne_1), tsne.2 = median(tsne_2)) %>% select(tsne.1, tsne.2)\n",
    "\n",
    "a.to.b.axis = centroids$Cluster.2 - centroids$Cluster.1\n",
    "axis = a.to.b.axis / sqrt(sum(a.to.b.axis^2))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# project the data to the start and end position\n",
    "pData(cds)$a.to.b.axis.projection = mapply(function(x, y) {\n",
    "    sum(c(x, y) * a.to.b.axis)\n",
    "}, pData(cds)$tsne_1, pData(cds)$tsne_2)\n",
    "\n",
    "pData(cds)$a.to.b.axis.projection = with(pData(cds),\n",
    "    (a.to.b.axis.projection - min(a.to.b.axis.projection)) /\n",
    "    (max(a.to.b.axis.projection) - min(a.to.b.axis.projection)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# select expressed genes\n",
    "fData(cds)$prop.expr = apply(exprs(cds), 1, function(x) sum(x > 0) / length(x))\n",
    "expressed.genes = rownames(subset(fData(cds), prop.expr >= 0.005))\n",
    "length(expressed.genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# compare models and differential test (from Jonathan's worm analysis script)\n",
    "\n",
    "my.compareModels = function (full_models, reduced_models) \n",
    "{\n",
    "    stopifnot(length(full_models) == length(reduced_models))\n",
    "    test_res <- mapply(function(x, y) {\n",
    "        if (is.null(x) == FALSE && is.null(y) == FALSE) {\n",
    "            lrt <- VGAM::lrtest(x, y)\n",
    "            pval = lrt@Body[\"Pr(>Chisq)\"][2, ]\n",
    "            McFadden.R2 = 1 - lrt@Body[1, 2] / lrt@Body[2, 2]\n",
    "            family = x@family@vfamily\n",
    "            if (length(family) > 1) \n",
    "                family = family[1]\n",
    "            data.frame(status = \"OK\", family = \"binomialff\", McFadden.R2 = McFadden.R2, pval = pval)\n",
    "        }\n",
    "        else {\n",
    "            data.frame(status = \"FAIL\", family = NA, McFadden.R2 = 0, pval = 1)\n",
    "        }\n",
    "    }, full_models, reduced_models, SIMPLIFY = FALSE, USE.NAMES = TRUE)\n",
    "    test_res <- do.call(rbind.data.frame, test_res)\n",
    "    test_res$qval <- p.adjust(test_res$pval, method = \"BH\")\n",
    "    test_res\n",
    "}\n",
    "\n",
    "my.diff_test_helper = function (x, fullModelFormulaStr, reducedModelFormulaStr, expressionFamily, \n",
    "    weights, disp_func = NULL, verbose = FALSE) \n",
    "{\n",
    "    reducedModelFormulaStr <- paste(\"f_expression\", reducedModelFormulaStr, \n",
    "        sep = \"\")\n",
    "    fullModelFormulaStr <- paste(\"f_expression\", fullModelFormulaStr, \n",
    "        sep = \"\")\n",
    "    \n",
    "    f_expression <- ifelse(x > 0, 1, 0)\n",
    "    \n",
    "    test_res <- tryCatch({\n",
    "        if (verbose) {\n",
    "            full_model_fit <- VGAM::vglm(as.formula(fullModelFormulaStr), \n",
    "              epsilon = 0.1, family = expressionFamily)\n",
    "            reduced_model_fit <- VGAM::vglm(as.formula(reducedModelFormulaStr), \n",
    "              epsilon = 0.1, family = expressionFamily)\n",
    "        }\n",
    "        else {\n",
    "            full_model_fit <- suppressWarnings(VGAM::vglm(as.formula(fullModelFormulaStr), \n",
    "              epsilon = 0.1, family = expressionFamily))\n",
    "            reduced_model_fit <- suppressWarnings(VGAM::vglm(as.formula(reducedModelFormulaStr), \n",
    "              epsilon = 0.1, family = expressionFamily))\n",
    "        }\n",
    "        my.compareModels(list(full_model_fit), list(reduced_model_fit))\n",
    "    }, error = function(e) {\n",
    "        if (verbose) \n",
    "            print(e)\n",
    "        data.frame(status = \"FAIL\", family = \"binomialff\", \n",
    "            McFadden.R2 = 0, pval = 1, qval = 1)\n",
    "    })\n",
    "    test_res\n",
    "}\n",
    "\n",
    "                         \n",
    "my.differentialGeneTest = function (cds, fullModelFormulaStr = \"~sm.ns(Pseudotime, df=3)\", \n",
    "    reducedModelFormulaStr = \"~1\", cores = 1, \n",
    "    verbose = FALSE) \n",
    "{\n",
    "    status <- NA\n",
    "    if (cores > 1) {\n",
    "        diff_test_res <- monocle:::mcesApply(cds, 1, my.diff_test_helper, \n",
    "            c(\"BiocGenerics\", \"VGAM\", \"Matrix\"), cores = cores, \n",
    "            fullModelFormulaStr = fullModelFormulaStr, reducedModelFormulaStr = reducedModelFormulaStr, \n",
    "            expressionFamily = binomialff(),\n",
    "            disp_func = cds@dispFitInfo[[\"blind\"]]$disp_func, \n",
    "            verbose = verbose)\n",
    "        diff_test_res\n",
    "    }\n",
    "    else {\n",
    "        diff_test_res <- monocle:::smartEsApply(cds, 1, my.diff_test_helper, \n",
    "            convert_to_dense = TRUE, fullModelFormulaStr = fullModelFormulaStr, \n",
    "            reducedModelFormulaStr = reducedModelFormulaStr, \n",
    "            expressionFamily = binomialff(),\n",
    "            disp_func = cds@dispFitInfo[[\"blind\"]]$disp_func, \n",
    "            verbose = verbose)\n",
    "        diff_test_res\n",
    "    }\n",
    "    diff_test_res <- do.call(rbind.data.frame, diff_test_res)\n",
    "    diff_test_res$qval <- 1\n",
    "    diff_test_res$qval[which(diff_test_res$status == \"OK\")] <- p.adjust(subset(diff_test_res, \n",
    "        status == \"OK\")[, \"pval\"], method = \"BH\")\n",
    "    diff_test_res <- merge(diff_test_res, fData(cds), by = \"row.names\")\n",
    "    row.names(diff_test_res) <- diff_test_res[, 1]\n",
    "    diff_test_res[, 1] <- NULL\n",
    "    diff_test_res\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# identify the differential expressed genes\n",
    "system.time({\n",
    "    a.b.axis.DE.genes = my.differentialGeneTest(cds[expressed.genes,],\n",
    "        fullModelFormulaStr = \"~ sm.ns(Size_Factor, df=3) + sm.ns(a.to.b.axis.projection, df=3)\",\n",
    "        reducedModelFormulaStr = \"~ sm.ns(Size_Factor, df=3)\")\n",
    "})\n",
    "\n",
    "# alternatively use the differential expressed genes from the monocle pipeline\n",
    "\n",
    "a.b.DE.genes = differentialGeneTest(cds[expressed.genes, !is.na(pData(cds)$c)],\n",
    "    fullModelFormulaStr = \"~ sm.ns(Size_Factor, df = 3) + s(a.to.b.axis.projection) + c\",\n",
    "    reducedModelFormulaStr = \"~ sm.ns(Size_Factor, df = 3) + s(a.to.b.axis.projection)\", \n",
    "                                    cores = 6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# write the DE analysis genes to te file\n",
    "a.b.axis.DE.genes %>% arrange(qval) %>% head(15)\n",
    "sum(a.b.axis.DE.genes$qval < 0.01 & a.b.axis.DE.genes$McFadden.R2 >= 0.05)\n",
    "write.table(format(as.data.frame(a.b.axis.DE.genes %>%\n",
    "    filter(qval < 0.01, McFadden.R2 >= 0.05) %>%\n",
    "    select(symbol, qval, McFadden.R2, num_cells_expressed) %>%\n",
    "    arrange(qval)), digits = 4),\n",
    "    file = \"./a.b.axis.DE\",\n",
    "    sep = \"\\t\", quote = F, row.names = F, col.names = F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# plot the DE analysis genes for selected genes in line plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# plot the DE analysis genes for selected genes\n",
    "a.b.axis.DE.gene.smooth.curves = lapply(subset(a.b.axis.DE.genes,\n",
    "        qval < 0.01 & McFadden.R2 >= 0.05)$gene_id, function(g) {\n",
    "    pData(cds.bwm)$expr = exprs(cds.bwm)[g,]\n",
    "    model = vgam(expr > 0 ~ sm.ns(Size_Factor, df=3) + s(a.to.b.axis.projection), binomialff(), pData(cds))\n",
    "    pData(cds)$expr = NULL\n",
    "    res = data.frame(a.to.b.axis.projection = seq(0, 1, 0.005), Size_Factor = 1, gene_id = g)\n",
    "    res$odds = exp(predict(model, res))[, 1]\n",
    "    res$prob = res$odds / (res$odds + 1)\n",
    "    res\n",
    "})\n",
    "\n",
    "a.b.axis.DE.gene.smooth.curves = do.call(rbind.data.frame, a.b.axis.DE.gene.smooth.curves) %>%\n",
    "    inner_join(fData(cds) %>% select(gene_id, symbol), by = \"gene_id\")\n",
    "\n",
    "head(a.b.axis.DE.gene.smooth.curves)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "tmp = a.b.axis.DE.gene.smooth.curves %>% filter(\n",
    "    symbol %in% c(\"a\", \"b\", \"c\", \"d\", \"e\"))\n",
    "\n",
    "tmp$symbol = factor(tmp$symbol, levels =\n",
    "    c(\"a\", \"b\", \"c\", \"d\", \"e\"))\n",
    "\n",
    "plot = ggplot(tmp, aes(x = a.to.b.axis.projection, y = prob, color = symbol)) +\n",
    "    geom_line(size = 1.2) +\n",
    "    scale_x_continuous(labels = percent) +\n",
    "    scale_y_continuous(labels = percent) +\n",
    "    scale_color_manual(values = c(\"#984EA3\", \"#4DAF4A\", \"#377EB8\", \"#FF7F00\", \"#B79F00\")) +\n",
    "    xlab(\"Distance along a-b axis\") +\n",
    "    ylab(\"P(>0 reads\\ngiven Size Factor = 1)\") +\n",
    "    guides(color = guide_legend(title = \"Marker gene\")) +\n",
    "    monocle:::monocle_theme_opts() +\n",
    "    theme(axis.title.x = element_text(margin = margin(8, 0, 0, 0)),\n",
    "          axis.text.x = element_text(margin = margin(0, 8, 0, 0)))\n",
    "\n",
    "rm(tmp)\n",
    "show(plot)\n",
    "\n",
    "ggsave(\"./pdf/a.to.b.DE.genes.pdf\",\n",
    "    device = cairo_pdf, plot = plot, units = \"in\", width = 3*1.618, height = 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# plot the DE genes in heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "a.b.axis.DE.gene.smooth.curve.mat = acast(a.b.axis.DE.gene.smooth.curves,\n",
    "    symbol ~ a.to.b.axis.projection, value.var = \"prob\")\n",
    "a.b.axis.DE.gene.smooth.curve.mat[1:5, 1:5]\n",
    "\n",
    "a.b.heatmap = pheatmap(t(scale(t(a.b.axis.DE.gene.smooth.curve.mat))),\n",
    "    cluster_cols = F, cutree_rows = 13,\n",
    "    show_rownames = F, show_colnames = F,\n",
    "    treeheight_row = 30)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "?scale"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "a.b.gene.clusters = cutree(a.b.heatmap$tree_row, 13)\n",
    "a.b.heatmap = pheatmap(t(scale(t(a.b.axis.DE.gene.smooth.curve.mat))),\n",
    "    cluster_cols = F, cutree_rows = 13,\n",
    "    annotation_row = data.frame(Cluster = factor(a.b.gene.clusters)),\n",
    "    show_rownames = F, show_colnames = F,\n",
    "    treeheight_row = 30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# check the differential expressed genes in each cluster\n",
    "a.b.axis.DE.genes %>% arrange(-McFadden.R2) %>%\n",
    "    filter(symbol %in% names(a.b.gene.clusters)[which(a.b.gene.clusters == 1)]) %>%\n",
    "    select(McFadden.R2, qval, symbol, num_cells_expressed, prop.expr) %>% head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# check the gene expression fold change between different groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "tmp1 = subset(pData(cds.bwm), expr.gene.1 == 1)$Size_Factor\n",
    "tmp2 = subset(pData(cds.bwm), expr.gene.2 == 1)$Size_Factor\n",
    "\n",
    "a.b.DE.genes = filter(a.b.DE.genes, qval < 0.01)\n",
    "\n",
    "a.b.DE.genes$a.mean = apply(\n",
    "    exprs(cds)[subset(a.b.DE.genes, qval < 0.01)$gene_id, pData(cds)$expr.gene.1 == 1],\n",
    "    1, function(x) { mean(x / tmp1) })\n",
    "\n",
    "a.b.DE.genes$b.mean = apply(\n",
    "    exprs(cds)[subset(a.b.DE.genes, qval < 0.01)$gene_id, pData(cds)$expr.gene.b == 1],\n",
    "    1, function(x) { mean(x / tmp2) })\n",
    "\n",
    "a.b.DE.genes$a.to.b.ratio = round(a.b.DE.genes$a.mean / a.b.DE.genes$b.mean, 2)\n",
    "\n",
    "rm(tmp1)\n",
    "rm(tmp2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# show gene expression jitter plot in different groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "plot = plot_genes_jitter(cds[sapply(c(\"a\", \"b\", \"c\", \"d\"),\n",
    "        function(g) get.gene.id(cds, g)), !is.na(pData(cds)$a)],\n",
    "    grouping = \"a\", color_by = \"a\", plot_trend = T,\n",
    "    cell_size = 0.1, ncol = 2)\n",
    "show(plot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## construct the single cell trajectory "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# set the order\n",
    "cds = setOrderingFilter(cds,\n",
    "    ((DE.genes %>% arrange(qval) %>% head(200))$gene_id))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# set the order\n",
    "cds = setOrderingFilter(cds,\n",
    "    ((DE.genes %>% arrange(qval) %>% head(200))$gene_id))\n",
    "\n",
    "# construct the DDR tree\n",
    "\n",
    "system.time({\n",
    "set.seed(42)\n",
    "cds = reduceDimension(cds, max_components = 2,\n",
    "    residualModelFormulaStr = \"~ sm.ns(Size_Factor, df = 3) + plate\",\n",
    "    auto_param_selection = F, ncenter = 420)\n",
    "})\n",
    "\n",
    "\n",
    "# define the root state\n",
    "\n",
    "GM_state <- function(cds){\n",
    "  if (length(unique(pData(cds)$State)) > 1){\n",
    "    T0_counts <- table(pData(cds)$State, pData(cds)$Hours)[,\"0\"]\n",
    "return(as.numeric(names(T0_counts)[which(T0_counts == max(T0_counts))])) }else { return (1) }\n",
    "}\n",
    "\n",
    "# order cells\n",
    "cds = orderCells(cds)\n",
    "cds = orderCells(cds, root_state = GM_state(cds))\n",
    "plot_cell_trajectory(cds, color_by = \"State/cell_type/Pseudotime\")\n",
    "\n",
    "# plot the DDR tree with gene markers\n",
    "# plot the DDR tree with gene markers\n",
    "pData(cds)$DDRTree_1 = cds@reducedDimS[1,]\n",
    "pData(cds)$DDRTree_2 = cds@reducedDimS[2,]\n",
    "\n",
    "plot = ggplot(pData(cds), aes(x = DDRTree_1, y = DDRTree_2, color = plot.group)) +\n",
    "    geom_point(aes(alpha = plot.group %in% c(\"None\", \"Multiple\")), size = 0.5) +\n",
    "    scale_color_manual(values = c(\"#9E0142\", \"goldenrod3\", \"forestgreen\", \"#3288BD\", \"#5E4FA2\", \"grey20\", \"grey80\")) +\n",
    "    scale_alpha_manual(values = c(1.0, 0.3333)) +\n",
    "    guides(color = guide_legend(title = \"Marker gene\", override.aes = list(size=3)), alpha = F) +\n",
    "    monocle:::monocle_theme_opts()\n",
    "\n",
    "show(plot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# define the root state\n",
    "\n",
    "GM_state <- function(cds){\n",
    "  if (length(unique(pData(cds)$State)) > 1){\n",
    "    T0_counts <- table(pData(cds)$State, pData(cds)$Hours)[,\"0\"]\n",
    "return(as.numeric(names(T0_counts)[which(T0_counts == max(T0_counts))])) }else { return (1) }\n",
    "}\n",
    "\n",
    "HSMM_myo <- orderCells(HSMM_myo, root_state=GM_state(HSMM_myo))\n",
    "plot_cell_trajectory(HSMM_myo, color_by=\"Pseudotime\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generate smooth line for gene expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "model = vgam(value > 0 ~ sm.ns(Size_factor, df=3 ) + s(distance), binomialff(), pData(cds))\n",
    "res = data.frame(direction = seq(0, 1, 0.005), Size_factor = 1, gene_id = g)\n",
    "res$odds = exp(predict(model, res))[,1]\n",
    "res$prob = res$odds / (res$odds + 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generate heatmap using pheatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "A.P.heatmap = pheatmap(t(scale(t(A.P.axis.DE.gene.smooth.curve.mat))),\n",
    "    cluster_cols = F, cutree_rows = 13,\n",
    "    show_rownames = F, show_colnames = F,\n",
    "    treeheight_row = 30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "A.P.gene.clusters = cutree(A.P.heatmap$tree_row, 13)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pheatmap(t(scale(t(data))),\n",
    "    cluster_cols = F, cutree_rows = 13,\n",
    "    annotation_row = data.frame(Cluster = factor(A.P.gene.clusters)),\n",
    "    show_rownames = F, show_colnames = F,\n",
    "    treeheight_row = 30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### use deviance for model selection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "m1 = vgam(expr > 0 ~ s(n.umi) + s(tsne_1) + s(tsne_2), family = binomialff(), data = pData(cds))\n",
    "m0 = vgam(expr > 0 ~ s(n.umi), family = binomialff(), data = pData(cds)\n",
    "return(1 - deviance(m1) / deviance(m0)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### binary matrix transformation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "binary.matrices = Matrix(ifelse(as.matrix(exprs(cds)) > 0, 1, 0), sparse=T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### identify the genes covaried with tsne"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "expr.tsne.r2 = function(cds, gene) {\n",
    "    pData(cds)$expr = exprs(cds)[gene,]\n",
    "    \n",
    "    m1 = vgam(expr > 0 ~ s(n.umi) + s(tsne_1) + s(tsne_2), family = binomialff(), data = pData(cds))\n",
    "    m0 = vgam(expr > 0 ~ s(n.umi), family = binomialff(), data = pData(cds))\n",
    "    \n",
    "    pData(cds)$expr = NULL\n",
    "    return(1 - deviance(m1) / deviance(m0))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### shuffle a binary matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# http://www.nature.com/articles/ncomms5114\n",
    "curve.ball = function(m) {\n",
    "    RC = dim(m)\n",
    "    R = RC[1]\n",
    "    C = RC[2]\n",
    "    hp = list()\n",
    "    for (row in 1:dim(m)[1]) {\n",
    "        hp[[row]] = which(m[row,] == 1)\n",
    "    }\n",
    "\n",
    "    l_hp = length(hp)\n",
    "    for (rep in 1:(5*l_hp)) {\n",
    "        AB = sample(1:l_hp, 2)\n",
    "        a = hp[[AB[1]]]\n",
    "        b = hp[[AB[2]]]\n",
    "        ab = intersect(a,b)\n",
    "        l_ab = length(ab)\n",
    "        l_a = length(a)\n",
    "        l_b = length(b)\n",
    "        if ((l_ab %in% c(l_a,l_b)) == F) {\n",
    "            tot = setdiff(c(a,b), ab)\n",
    "            l_tot = length(tot)\n",
    "            tot = sample(tot, l_tot, replace = FALSE, prob = NULL)\n",
    "            L = l_a - l_ab\n",
    "            hp[[AB[1]]] = c(ab,tot[1:L])\n",
    "            hp[[AB[2]]] = c(ab,tot[(L+1):l_tot])\n",
    "        }\n",
    "    }\n",
    "\n",
    "    rm = matrix(0, R, C)\n",
    "    for (row in 1:R) {\n",
    "        rm[row, hp[[row]]] = 1\n",
    "    }\n",
    "\n",
    "    rownames(rm) = rownames(m)\n",
    "    colnames(rm) = colnames(m)\n",
    "\n",
    "    Matrix(rm, sparse=T)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### gene covariance network analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "subset.cds = list()\n",
    "binary.matrices = list()\n",
    "expressed.genes = list()\n",
    "glmnet.input.mat = list()\n",
    "\n",
    "# subset the data set\n",
    "subset.cds$bwm = cds[, with(pData(cds), !is.na(cell.type) & cell.type == \"Body wall muscle\")]\n",
    "tmp = exprs(subset.cds$bwm)\n",
    "rownames(tmp) = fData(subset.cds$bwm)$symbol\n",
    "\n",
    "subset.cds$bwm = newCellDataSet(\n",
    "    tmp,\n",
    "    phenoData = new(\"AnnotatedDataFrame\", pData(subset.cds$bwm)),\n",
    "    featureData = new(\"AnnotatedDataFrame\",\n",
    "        data.frame(fData(subset.cds$bwm), row.names = fData(subset.cds$bwm)$symbol)),\n",
    "    expressionFamily = negbinomial.size())\n",
    "\n",
    "rm(tmp)\n",
    "\n",
    "# transform the data to binary data set\n",
    "binary.matrices$bwm = Matrix(ifelse(as.matrix(exprs(subset.cds$bwm)) > 0, 1, 0), sparse=T)\n",
    "\n",
    "# filter genes\n",
    "fData(subset.cds$bwm)$prop.expr = apply(as.matrix(binary.matrices$bwm), 1, function(x) { sum(x) / length(x) })\n",
    "\n",
    "subset.cds$bwm = estimateSizeFactors(subset.cds$bwm)\n",
    "subset.cds$bwm = estimateDispersions(subset.cds$bwm)\n",
    "subset.cds$bwm = detectGenes(subset.cds$bwm, 0.1)\n",
    "expressed.genes$bwm = subset(fData(subset.cds$bwm), num_cells_expressed >= 10 & prop.expr >= 0.005)$symbol\n",
    "\n",
    "# select genes covaried with tsne\n",
    "expr.tsne.r2 = function(cds, gene) {\n",
    "    pData(cds)$expr = exprs(cds)[gene,]\n",
    "    \n",
    "    m1 = vgam(expr > 0 ~ s(n.umi) + s(tsne_1) + s(tsne_2), family = binomialff(), data = pData(cds))\n",
    "    m0 = vgam(expr > 0 ~ s(n.umi), family = binomialff(), data = pData(cds))\n",
    "    \n",
    "    pData(cds)$expr = NULL\n",
    "    return(1 - deviance(m1) / deviance(m0))\n",
    "}\n",
    "cluster = makeCluster(8)\n",
    "clusterCall(cluster, function() library(monocle))\n",
    "clusterExport(cluster, c(\"subset.cds\", \"expressed.genes\", \"expr.tsne.r2\"))\n",
    "system.time({\n",
    "\n",
    "fData(subset.cds$bwm)$tsne.r2 = parSapply(cluster, fData(subset.cds$bwm)$symbol, function(gene) {\n",
    "    ifelse(gene %in% expressed.genes$bwm, expr.tsne.r2(subset.cds$bwm, gene), NA)\n",
    "})\n",
    "    \n",
    "})\n",
    "\n",
    "# transform the data to binary data\n",
    "glmnet.input.mat$bwm.5p = t(\n",
    "    binary.matrices$bwm[with(fData(subset.cds$bwm), !is.na(tsne.r2) & tsne.r2 >= 0.05),])\n",
    "    \n",
    "\n",
    "# generate shuffed data\n",
    "curve.ball = function(m) {\n",
    "    RC = dim(m)\n",
    "    R = RC[1]\n",
    "    C = RC[2]\n",
    "    hp = list()\n",
    "    for (row in 1:dim(m)[1]) {\n",
    "        hp[[row]] = which(m[row,] == 1)\n",
    "    }\n",
    "\n",
    "    l_hp = length(hp)\n",
    "    for (rep in 1:(5*l_hp)) {\n",
    "        AB = sample(1:l_hp, 2)\n",
    "        a = hp[[AB[1]]]\n",
    "        b = hp[[AB[2]]]\n",
    "        ab = intersect(a,b)\n",
    "        l_ab = length(ab)\n",
    "        l_a = length(a)\n",
    "        l_b = length(b)\n",
    "        if ((l_ab %in% c(l_a,l_b)) == F) {\n",
    "            tot = setdiff(c(a,b), ab)\n",
    "            l_tot = length(tot)\n",
    "            tot = sample(tot, l_tot, replace = FALSE, prob = NULL)\n",
    "            L = l_a - l_ab\n",
    "            hp[[AB[1]]] = c(ab,tot[1:L])\n",
    "            hp[[AB[2]]] = c(ab,tot[(L+1):l_tot])\n",
    "        }\n",
    "    }\n",
    "\n",
    "    rm = matrix(0, R, C)\n",
    "    for (row in 1:R) {\n",
    "        rm[row, hp[[row]]] = 1\n",
    "    }\n",
    "\n",
    "    rownames(rm) = rownames(m)\n",
    "    colnames(rm) = colnames(m)\n",
    "\n",
    "    Matrix(rm, sparse=T)\n",
    "}\n",
    "\n",
    "lambda.seq = list()\n",
    "glmnet.input.mat.shuf = list()\n",
    "glmnet.shuf.res = list()\n",
    "est.fdr = list()\n",
    "selected.lambda.idx = list()\n",
    "glmnet.res = list()\n",
    "coexpr.networks = list()\n",
    "coexpr.igraph = list()\n",
    "\n",
    "clusterExport(cluster, c(\"glmnet.input.mat\", \"curve.ball\"))\n",
    "glmnet.input.mat.shuf$bwm.5p = parLapply(cluster, 1:10, function(i) {\n",
    "    curve.ball(glmnet.input.mat$bwm.5p)\n",
    "})\n",
    "\n",
    "lambda.seq$bwm.5p = 10^seq(-1, -3, -0.02)\n",
    "\n",
    "clusterCall(cluster, function() library(glmnet))\n",
    "\n",
    "clusterExport(cluster, c(\"glmnet.input.mat.shuf\", \"lambda.seq\"))\n",
    "\n",
    "for (shuf.idx in 1:10) {\n",
    "    clusterExport(cluster, c(\"shuf.idx\"))\n",
    "    \n",
    "    glmnet.shuf.res$bwm.5p[[shuf.idx]] = parLapply(cluster, 1:ncol(glmnet.input.mat$bwm.5p), function(i) {\n",
    "        coef(glmnet(\n",
    "            glmnet.input.mat.shuf$bwm.5p[[shuf.idx]][, -i],\n",
    "            glmnet.input.mat.shuf$bwm.5p[[shuf.idx]][, i],\n",
    "            alpha = 1.0, lambda = lambda.seq$bwm.5p))\n",
    "    })\n",
    "    \n",
    "    message(shuf.idx)\n",
    "}\n",
    "\n",
    "# calculate the number of edges for the shuffed sample\n",
    "    \n",
    "shuf.n.edges = function(original.mat, glmnet.shuf.res, shuf.idx, lambda.idx) {\n",
    "    network = data.frame(from = character(), to = character())\n",
    "    \n",
    "    for(i in 1:ncol(original.mat)) {\n",
    "        if (length(which(glmnet.shuf.res[[shuf.idx]][[i]][-1, lambda.idx] != 0)) > 0) {\n",
    "            network = rbind(network, data.frame(\n",
    "                from = names(which(glmnet.shuf.res[[shuf.idx]][[i]][-1, lambda.idx] != 0)),\n",
    "                to = colnames(original.mat)[i]))\n",
    "        }\n",
    "    }\n",
    "    \n",
    "    network$from = as.character(network$from)\n",
    "    network$to = as.character(network$to)\n",
    "    \n",
    "    network = rbind(network, mutate(network, tmp = from, from = to, to = tmp) %>% select(from, to)) %>%\n",
    "        filter(from < to) %>% group_by(from, to) %>% summarize(bidirectional = n() > 1)\n",
    "    \n",
    "    nrow(network)\n",
    "}\n",
    "    \n",
    "\n",
    "shuf.n.edges(glmnet.input.mat$bwm.10p, glmnet.shuf.res$bwm.10p, 1, 160)\n",
    "\n",
    "# calculate the number of possible edges\n",
    "n.possible.edges = function(original.mat) {\n",
    "    (ncol(original.mat) * (ncol(original.mat) - 1)) / 2\n",
    "}\n",
    "\n",
    "# calculate the FDR rate using the shuffed sample and select the best lambda\n",
    "clusterCall(cluster, function() library(dplyr))\n",
    "    \n",
    "clusterExport(cluster, c(\"glmnet.input.mat\", \"glmnet.shuf.res\", \"shuf.n.edges\", \"n.possible.edges\"))\n",
    "\n",
    "est.fdr$bwm.5p = parSapply(cluster, 1:length(lambda.seq$bwm.5p), function(lambda.idx) {\n",
    "    mean(sapply(1:10, function(shuf.idx) {\n",
    "        shuf.n.edges(glmnet.input.mat$bwm.5p, glmnet.shuf.res$bwm.5p, shuf.idx, lambda.idx) /\n",
    "        n.possible.edges(glmnet.input.mat$bwm.5p)\n",
    "    }))\n",
    "})\n",
    "\n",
    "est.fdr$bwm.5p = data.frame(\n",
    "    lambda = lambda.seq$bwm.5p,\n",
    "    est.fdr = est.fdr$bwm.5p)\n",
    "\n",
    "ggplot(est.fdr$bwm.5p, aes(x = log10(lambda), y = est.fdr)) + geom_line() + monocle:::monocle_theme_opts()\n",
    "\n",
    "\n",
    "selected.lambda.idx$bwm.5p = tail(est.fdr$bwm.5p %>% mutate(rn = row_number()) %>% filter(est.fdr < 0.05), 1)$rn\n",
    "    \n",
    "    \n",
    "    \n",
    "# use the best lambda for calculating the gene networks\n",
    "\n",
    "glmnet.res$bwm.5p = parLapply(cluster, 1:ncol(glmnet.input.mat$bwm.5p), function(i) {\n",
    "    coef(glmnet(\n",
    "        glmnet.input.mat$bwm.5p[, -i], glmnet.input.mat$bwm.5p[, i],\n",
    "        alpha = 1.0, lambda = lambda.seq$bwm.5p))\n",
    "})\n",
    "    \n",
    "make.network.from.glmnet.res = function(input.mat, glmnet.res, lambda.idx) {\n",
    "    network = data.frame(from = character(), to = character(), weight = numeric())\n",
    "    \n",
    "    for(i in 1:ncol(input.mat)) {\n",
    "        if (length(which(glmnet.res[[i]][-1, lambda.idx] != 0)) > 0) {\n",
    "            network = rbind(network, data.frame(\n",
    "                from = names(which(glmnet.res[[i]][-1, lambda.idx] > 0)),\n",
    "                to = colnames(input.mat)[i],\n",
    "                weight = glmnet.res[[i]][-1, lambda.idx][glmnet.res[[i]][-1, lambda.idx] > 0]))\n",
    "        }\n",
    "    }\n",
    "    \n",
    "    network$from = as.character(network$from)\n",
    "    network$to = as.character(network$to)\n",
    "    \n",
    "    network = rbind(network, mutate(network, tmp = from, from = to, to = tmp) %>% select(from, to, weight)) %>%\n",
    "        filter(from < to) %>% group_by(from, to) %>% summarize(weight = max(weight), bidirectional = n() > 1)\n",
    "    \n",
    "    network\n",
    "}\n",
    "\n",
    "coexpr.networks$bwm.5p = make.network.from.glmnet.res(\n",
    "    glmnet.input.mat$bwm.5p, glmnet.res$bwm.5p, selected.lambda.idx$bwm.5p)\n",
    "\n",
    "\n",
    "# plot the network\n",
    "plot.network = function(g, file, seed, width=8.0, height=8.0,\n",
    "                        label.cex = 0.75, edge.width.scaling=20.0, vertex.size = 12,\n",
    "                        vertex.color=\"#D4DCFF\") {\n",
    "    E(g)$color = ifelse(E(g)$bidirectional, \"black\", rgb(0.666, 0.666, 0.666, 0.5))\n",
    "    V(g)$color = vertex.color\n",
    "    V(g)$label.cex = label.cex\n",
    "    V(g)$name = sub(\"(?<=[a-z])[.]\", \"-\", V(g)$name, perl=T)\n",
    "\n",
    "    cairo_pdf(file, width = width, height = height)\n",
    "    set.seed(seed)\n",
    "    g.layout = layout_nicely(g)\n",
    "    par(mar=c(0,0,0,0))\n",
    "    plot(g,\n",
    "        edge.width = edge.width.scaling * E(g)$weight,\n",
    "        vertex.size = vertex.size,\n",
    "        vertex.label.family = \"sans\",\n",
    "        vertex.label.color = \"black\",\n",
    "        edge.color = E(g)$color,\n",
    "        layout = g.layout)\n",
    "    dev.off()\n",
    "\n",
    "    return(list(\n",
    "        g = g, g.layout = g.layout,\n",
    "        file=file, width = width, height = height,\n",
    "        vertex.size = vertex.size,\n",
    "        edge.width.scaling = edge.width.scaling))\n",
    "}\n",
    "\n",
    "coexpr.plots$bwm.10p = plot.network(coexpr.igraph$bwm.10p, file = \"plots/BWM.coexpr.pdf\", seed = 42)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gene spacial differential analysis 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# define the clusters of two clusters\n",
    "centroids = list()\n",
    "centroids$bwm.tni.3 = pData(subset.cds$bwm)[exprs(subset.cds$bwm)[\"tni-3\",] > 0,] %>%\n",
    "    group_by(1) %>% summarize(tsne.1 = median(tsne_1), tsne.2 = median(tsne_2)) %>% select(tsne.1, tsne.2)\n",
    "\n",
    "centroids$bwm.egl.20 = pData(subset.cds$bwm)[exprs(subset.cds$bwm)[\"egl-20\",] > 0,] %>%\n",
    "    group_by(1) %>% summarize(tsne.1 = median(tsne_1), tsne.2 = median(tsne_2)) %>% select(tsne.1, tsne.2)\n",
    "\n",
    "# plot the marker genes\n",
    "\n",
    "pData(subset.cds$bwm)$expr.tni.3 = exprs(subset.cds$bwm)[\"tni-3\",] > 0\n",
    "pData(subset.cds$bwm)$expr.sfrp.1 = exprs(subset.cds$bwm)[\"sfrp-1\",] > 0\n",
    "pData(subset.cds$bwm)$expr.cwn.1 = exprs(subset.cds$bwm)[\"cwn-1\",] > 0\n",
    "pData(subset.cds$bwm)$expr.egl.20 = exprs(subset.cds$bwm)[\"egl-20\",] > 0\n",
    "\n",
    "pData(subset.cds$bwm)$plot.expr = with(pData(subset.cds$bwm), ifelse(\n",
    "    expr.tni.3 + expr.sfrp.1 + expr.cwn.1 + expr.egl.20 > 1, \"Multiple\", ifelse(\n",
    "        expr.tni.3, \"tni-3\", ifelse(\n",
    "            expr.sfrp.1, \"sfrp-1\", ifelse(\n",
    "                expr.cwn.1, \"cwn-1\", ifelse(\n",
    "                    expr.egl.20, \"egl-20\", \"None\"))))))\n",
    "\n",
    "pData(subset.cds$bwm)$plot.expr = factor(pData(subset.cds$bwm)$plot.expr, levels = c(\n",
    "    \"tni-3\", \"sfrp-1\", \"cwn-1\", \"egl-20\", \"Multiple\", \"None\"))\n",
    "plot = ggplot(pData(subset.cds$bwm), aes(x = tsne_1, y = tsne_2)) +\n",
    "    geom_point(aes(color = plot.expr), size = 0.25) +\n",
    "    scale_color_manual(values = c(\"#0FA3B1\", \"#8ABF69\", \"#FFBA49\", \"#EF5B5B\", \"darkorchid1\", \"grey50\")) +\n",
    "    geom_point(\n",
    "        x = centroids$bwm.tni.3$tsne.1, y = centroids$bwm.tni.3$tsne.2,\n",
    "        size = 2.0, color = \"#0FA3B1\") +\n",
    "    geom_point(\n",
    "        x = centroids$bwm.egl.20$tsne.1, y = centroids$bwm.egl.20$tsne.2,\n",
    "        size = 2.0, color = \"#EF5B5B\") +\n",
    "    geom_segment(\n",
    "        x = centroids$bwm.tni.3$tsne.1, xend = centroids$bwm.egl.20$tsne.1,\n",
    "        y = centroids$bwm.tni.3$tsne.2, yend = centroids$bwm.egl.20$tsne.2,\n",
    "        color = \"black\", size = 0.5) +\n",
    "    guides(color = guide_legend(title = \"Marker gene\")) +\n",
    "    monocle:::monocle_theme_opts()\n",
    "\n",
    "\n",
    "# define the axis of the centroid and project other cells into the axis\n",
    "\n",
    "BWM.posterior.to.anterior.axis = centroids$bwm.tni.3 - centroids$bwm.egl.20\n",
    "BWM.posterior.to.anterior.axis = BWM.posterior.to.anterior.axis / sqrt(sum(BWM.posterior.to.anterior.axis^2))\n",
    "\n",
    "pData(subset.cds$bwm)$post.ant.axis.projection = mapply(function(x, y) {\n",
    "    sum(c(x, y) * BWM.posterior.to.anterior.axis)\n",
    "}, pData(subset.cds$bwm)$tsne_1, pData(subset.cds$bwm)$tsne_2)\n",
    "\n",
    "pData(subset.cds$bwm)$post.ant.axis.projection =\n",
    "    (pData(subset.cds$bwm)$post.ant.axis.projection - min(pData(subset.cds$bwm)$post.ant.axis.projection)) /\n",
    "    (max(pData(subset.cds$bwm)$post.ant.axis.projection) - min(pData(subset.cds$bwm)$post.ant.axis.projection))\n",
    "\n",
    "\n",
    "\n",
    "BWM.gene.post.ant.dist = data.frame(gene = character(), post.ant.projection = numeric())\n",
    "\n",
    "for (gene in sub(\"(?<=[a-z])[.]\", \"-\",\n",
    "        unique(c(coexpr.networks$bwm.10p$from, coexpr.networks$bwm.10p$to)), perl=T)) {\n",
    "    BWM.gene.post.ant.dist = rbind(BWM.gene.post.ant.dist, data.frame(\n",
    "        gene = gene,\n",
    "        post.ant.projection = pData(subset.cds$bwm)[exprs(subset.cds$bwm)[gene,] > 0,]$post.ant.axis.projection))\n",
    "}\n",
    "\n",
    "tmp = setdiff(\n",
    "    unique(BWM.gene.post.ant.dist$gene),\n",
    "    subset(fData(subset.cds$bwm), prop.expr > 0.5)$symbol)\n",
    "\n",
    "BWM.post.ant.quant.mat = t(sapply(tmp, function(this.gene) {\n",
    "    quantile(filter(BWM.gene.post.ant.dist, gene == this.gene)$post.ant.projection, seq(0, 1, 0.005))\n",
    "}))\n",
    "\n",
    "rownames(BWM.post.ant.quant.mat) = tmp\n",
    "rm(tmp)\n",
    "\n",
    "\n",
    "# plot the heatmap of genes\n",
    "\n",
    "pdf(\"plots/BWM.anterior.posterior.heatmap.pdf\", onefile = F, width = 3, height = 3*1.618)\n",
    "BWM.ant.post.hm = pheatmap(BWM.post.ant.quant.mat,\n",
    "    cluster_cols = F, cutree_rows = 6,\n",
    "    border_color = NA, show_rownames = F, show_colnames = F)\n",
    "\n",
    "BWM.ant.post.clusters = cutree(BWM.ant.post.hm$tree_row, 6)\n",
    "\n",
    "pdf(\"plots/BWM.anterior.posterior.heatmap.pdf\", onefile = F, width = 5, height = 5)\n",
    "BWM.ant.post.hm = pheatmap(BWM.post.ant.quant.mat,\n",
    "    cluster_cols = F, cutree_rows = 6,\n",
    "    annotation_row = data.frame(\n",
    "        Cluster = factor(unname(BWM.ant.post.clusters), labels = c(\n",
    "            \"Posterior\", \"Anterior\", \"Head and tail\", \"Head\", \"Dispersed\", \"Mid-to-posterior\")),\n",
    "        row.names = row.names(BWM.post.ant.quant.mat)),\n",
    "    border_color = NA, show_rownames = F, show_colnames = F)\n",
    "dev.off()\n",
    "\n",
    "BWM.ant.post.gene.lists = list()\n",
    "BWM.ant.post.gene.lists$posterior = names(BWM.ant.post.clusters)[which(BWM.ant.post.clusters == 1)]\n",
    "BWM.ant.post.gene.lists$anterior = names(BWM.ant.post.clusters)[which(BWM.ant.post.clusters %in% c(2, 4))]\n",
    "BWM.ant.post.gene.lists$head = names(BWM.ant.post.clusters)[which(BWM.ant.post.clusters == 4)]\n",
    "\n",
    "coexpr.igraph$bwm.10p = graph.data.frame(coexpr.networks$bwm.10p %>% filter(weight >= 0.05), directed=F)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### gene spacial differential analysis 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# define centroid\n",
    "centroids = list()\n",
    "\n",
    "centroids$Cluster.1 = pData(cds.bwm)[pData(cds.bwm)$Cluster == 1,] %>%\n",
    "    group_by(1) %>% summarize(tsne.1 = median(tsne_1), tsne.2 = median(tsne_2)) %>% select(tsne.1, tsne.2)\n",
    "\n",
    "centroids$Cluster.7 = pData(cds.bwm)[pData(cds.bwm)$Cluster == 7,] %>%\n",
    "    group_by(1) %>% summarize(tsne.1 = median(tsne_1), tsne.2 = median(tsne_2)) %>% select(tsne.1, tsne.2)\n",
    "\n",
    "plot = ggplot(pData(cds.bwm), aes(x = tsne_1, y = tsne_2, colour = plot.group)) +\n",
    "    geom_point(aes(alpha = plot.group %in% c(\"None\", \"Multiple\")), size = 0.5) +\n",
    "    geom_point(x = centroids$Cluster.1$tsne.1, y = centroids$Cluster.1$tsne.2, color = \"black\", size = 1.5) +\n",
    "    geom_point(x = centroids$Cluster.7$tsne.1, y = centroids$Cluster.7$tsne.2, color = \"black\", size = 1.5) +\n",
    "    geom_segment(x = centroids$Cluster.1$tsne.1, xend = centroids$Cluster.7$tsne.1,\n",
    "                 y = centroids$Cluster.1$tsne.2, yend = centroids$Cluster.7$tsne.2,\n",
    "                 color = \"black\", size = 0.75) +\n",
    "    scale_color_manual(values = c(\"#9E0142\", \"goldenrod3\", \"forestgreen\", \"#3288BD\", \"#5E4FA2\", \"grey20\", \"grey80\")) +\n",
    "    scale_alpha_manual(values = c(1.0, 0.3333)) +\n",
    "    guides(color = guide_legend(title = \"Marker gene\", override.aes = list(size=3)), alpha = F) +\n",
    "    monocle:::monocle_theme_opts()\n",
    "\n",
    "show(plot)\n",
    "\n",
    "ggsave(\"plots/yo.check.out.these.Wnt.gradients.2.pdf\",\n",
    "    device = cairo_pdf, plot = plot, units = \"in\", width = 7.4, height = 5.7)\n",
    "\n",
    "\n",
    "# define axis and project data\n",
    "\n",
    "A.to.P.axis = centroids$Cluster.7 - centroids$Cluster.1\n",
    "A.to.P.axis = A.to.P.axis / sqrt(sum(A.to.P.axis^2))\n",
    "\n",
    "pData(cds.bwm)$A.to.P.axis.projection = mapply(function(x, y) {\n",
    "    sum(c(x, y) * A.to.P.axis)\n",
    "}, pData(cds.bwm)$tsne_1, pData(cds.bwm)$tsne_2)\n",
    "\n",
    "pData(cds.bwm)$A.to.P.axis.projection = with(pData(cds.bwm),\n",
    "    (A.to.P.axis.projection - min(A.to.P.axis.projection)) /\n",
    "    (max(A.to.P.axis.projection) - min(A.to.P.axis.projection)))\n",
    "\n",
    "fData(cds.bwm)$prop.expr = apply(exprs(cds.bwm), 1, function(x) sum(x > 0) / length(x))\n",
    "expressed.genes = rownames(subset(fData(cds.bwm), prop.expr >= 0.005))\n",
    "length(expressed.genes)\n",
    "\n",
    "# differential gene test along the axis\n",
    "system.time({\n",
    "    A.P.axis.DE.genes = my.differentialGeneTest(cds.bwm[expressed.genes,],\n",
    "        fullModelFormulaStr = \"~ sm.ns(Size_Factor, df=3) + sm.ns(A.to.P.axis.projection, df=3)\",\n",
    "        reducedModelFormulaStr = \"~ sm.ns(Size_Factor, df=3)\")\n",
    "})\n",
    "\n",
    "# write the table into file\n",
    "\n",
    "write.table(format(as.data.frame(A.P.axis.DE.genes %>%\n",
    "    filter(qval < 0.01, McFadden.R2 >= 0.05) %>%\n",
    "    select(symbol, qval, McFadden.R2, num_cells_expressed) %>%\n",
    "    arrange(qval)), digits = 4),\n",
    "    file = \"/net/trapnell/vol1/jspacker/wsc/gene-sets/BWM.A.P.axis\",\n",
    "    sep = \"\\t\", quote = F, row.names = F, col.names = F)\n",
    "    \n",
    "# generate and plot the smoothed gene expression for differentially expressed genes\n",
    "\n",
    "A.P.axis.DE.gene.smooth.curves = lapply(subset(A.P.axis.DE.genes,\n",
    "        qval < 0.01 & McFadden.R2 >= 0.05)$gene_id, function(g) {\n",
    "    pData(cds.bwm)$expr = exprs(cds.bwm)[g,]\n",
    "    model = vgam(expr > 0 ~ sm.ns(Size_Factor, df=3) + s(A.to.P.axis.projection), binomialff(), pData(cds.bwm))\n",
    "    pData(cds.bwm)$expr = NULL\n",
    "    res = data.frame(A.to.P.axis.projection = seq(0, 1, 0.005), Size_Factor = 1, gene_id = g)\n",
    "    res$odds = exp(predict(model, res))[, 1]\n",
    "    res$prob = res$odds / (res$odds + 1)\n",
    "    res\n",
    "})\n",
    "    \n",
    "A.P.axis.DE.gene.smooth.curves = do.call(rbind.data.frame, A.P.axis.DE.gene.smooth.curves) %>%\n",
    "    inner_join(fData(cds.bwm) %>% select(gene_id, symbol), by = \"gene_id\")\n",
    "\n",
    "head(A.P.axis.DE.gene.smooth.curves)\n",
    "\n",
    "tmp = A.P.axis.DE.gene.smooth.curves %>% filter(\n",
    "    symbol %in% c(\"tni-3\", \"sfrp-1\", \"cwn-2\", \"cwn-1\", \"egl-20\"))\n",
    "\n",
    "tmp$symbol = factor(tmp$symbol, levels =\n",
    "    c(\"tni-3\", \"sfrp-1\", \"cwn-2\", \"cwn-1\", \"egl-20\"))\n",
    "\n",
    "plot = ggplot(tmp, aes(x = A.to.P.axis.projection, y = prob, color = symbol)) +\n",
    "    geom_line(size = 1.2) +\n",
    "    scale_x_continuous(labels = percent) +\n",
    "    scale_y_continuous(labels = percent) +\n",
    "    scale_color_manual(values = c(\"#984EA3\", \"#4DAF4A\", \"#377EB8\", \"#FF7F00\", \"#B79F00\")) +\n",
    "    xlab(\"Distance along anterior-to-posterior axis\") +\n",
    "    ylab(\"P(>0 reads\\ngiven Size Factor = 1)\") +\n",
    "    guides(color = guide_legend(title = \"Marker gene\")) +\n",
    "    monocle:::monocle_theme_opts() +\n",
    "    theme(axis.title.x = element_text(margin = margin(8, 0, 0, 0)),\n",
    "          axis.text.x = element_text(margin = margin(0, 8, 0, 0)))\n",
    "\n",
    "rm(tmp)\n",
    "show(plot)\n",
    "\n",
    "ggsave(\"plots/BWM.A-P.marker.genes.smooth.curves.pdf\",\n",
    "    device = cairo_pdf, plot = plot, units = \"in\", width = 3*1.618, height = 2)\n",
    "    \n",
    "\n",
    "# generate the expression matrix for the DE genes\n",
    "A.P.axis.DE.gene.smooth.curve.mat = acast(A.P.axis.DE.gene.smooth.curves,\n",
    "    symbol ~ A.to.P.axis.projection, value.var = \"prob\")\n",
    "\n",
    "pdf(\"plots/BWM.A.P.heatmap.pdf\", width = 4, height = 4, onefile = F)\n",
    "\n",
    "A.P.heatmap = pheatmap(t(scale(t(A.P.axis.DE.gene.smooth.curve.mat))),\n",
    "    cluster_cols = F, cutree_rows = 13,\n",
    "    show_rownames = F, show_colnames = F,\n",
    "    treeheight_row = 30)\n",
    "\n",
    "dev.off()\n",
    "\n",
    "show(A.P.heatmap)\n",
    "    \n",
    "A.P.gene.clusters = cutree(A.P.heatmap$tree_row, 13)\n",
    "\n",
    "pdf(\"plots/BWM.A.P.heatmap.pdf\", width = 4, height = 4, onefile = F)\n",
    "\n",
    "A.P.heatmap = pheatmap(t(scale(t(A.P.axis.DE.gene.smooth.curve.mat))),\n",
    "    cluster_cols = F, cutree_rows = 13,\n",
    "    annotation_row = data.frame(Cluster = factor(A.P.gene.clusters)),\n",
    "    show_rownames = F, show_colnames = F,\n",
    "    treeheight_row = 30)\n",
    "\n",
    "dev.off()\n",
    "    \n",
    "\n",
    "# define the type of genes based on their position\n",
    "    \n",
    "fData(cds.bwm)$head = ifelse(fData(cds.bwm)$symbol %in%\n",
    "    names(A.P.gene.clusters)[which(A.P.gene.clusters %in% c(1))], 1, 0)\n",
    "fData(cds.bwm)$anterior = ifelse(fData(cds.bwm)$symbol %in%\n",
    "    names(A.P.gene.clusters)[which(A.P.gene.clusters %in% c(1, 5, 7, 8))], 1, 0)\n",
    "fData(cds.bwm)$posterior = ifelse(fData(cds.bwm)$symbol %in%\n",
    "    names(A.P.gene.clusters)[which(A.P.gene.clusters %in% c(2, 4))], 1, 0)\n",
    "fData(cds.bwm)$tail = ifelse(fData(cds.bwm)$symbol %in%\n",
    "    names(A.P.gene.clusters)[which(A.P.gene.clusters %in% c(4))], 1, 0)\n",
    "\n",
    "sum(fData(cds.bwm)$head)\n",
    "sum(fData(cds.bwm)$anterior)\n",
    "sum(fData(cds.bwm)$posterior)\n",
    "sum(fData(cds.bwm)$tail)\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### correlation analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "plot = ggplot(tmp.df, aes(x = L2.tpm, y = value)) +\n",
    "    facet_wrap(~ zork, ncol = 1) +\n",
    "    geom_point(size = 0.01, alpha = 0.01333) +\n",
    "    geom_abline(color = \"#E41A1C\", size = 0.5) +\n",
    "    geom_smooth(method = \"lm\", color = \"#377EB8\", size = 0.5) +\n",
    "    scale_x_log10(labels = comma, breaks = c(1, 100, 10000)) +\n",
    "    scale_y_log10(labels = comma, breaks = c(1, 100, 10000)) +\n",
    "    coord_fixed() +\n",
    "    ylab(\"sci-RNA-seq TPM\") +\n",
    "    xlab(\"Bulk whole-organism\\nRNA-seq TPM\") +\n",
    "    theme_bw(base_size = 6) +\n",
    "    monocle:::monocle_theme_opts() +\n",
    "    theme(axis.title.x = element_text(margin = margin(5, 0, 0, 0)),\n",
    "          axis.title.y = element_text(margin = margin(0, 2, 0, 0)),\n",
    "          strip.text.x = element_text(size = 6))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### bar plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "plot = ggplot(tmp.df, aes(x = cell.type, y = proportion, fill = group)) +\n",
    "    geom_bar(stat = \"identity\", position = \"dodge\") +\n",
    "    #geom_text(aes(\n",
    "    #    label = prettyNum(umi.total, big.mark = \",\"), y = proportion + 0.01,\n",
    "    #    group = group, alpha = group), position = position_dodge(1.0), size = 1.2) +\n",
    "    scale_y_continuous(breaks = seq(0, 1, 0.05), labels = percent) +\n",
    "    scale_fill_manual(\n",
    "        labels = c(\"In individual worm\", \"Identifiable in\\nsci-RNA-seq data\"),\n",
    "        values = brewer.pal(3, \"Set1\")[1:2]) +\n",
    "    #scale_alpha_manual(values = c(1, 0)) +\n",
    "    xlab(\"\") + ylab(\"\") +\n",
    "    coord_flip() +\n",
    "    guides(fill = F, alpha = F) +\n",
    "    theme_bw(base_size = 7) +\n",
    "    monocle:::monocle_theme_opts() +\n",
    "    theme(\n",
    "        axis.title.x = element_text(margin = margin(-5, 0, 0, 0)),\n",
    "        axis.title.y = element_text(angle = 0, vjust = 0.6, margin = margin(0, 5, 0, 0)),\n",
    "        axis.text.x = element_text(\n",
    "            angle = 90, hjust = 1.0, vjust = 0.5, margin = margin(3, 0, 0, 0),\n",
    "            color = \"black\"),\n",
    "        axis.text.y = element_text(color = \"black\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### plot heatmap for gene expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "mat = cell.type.norm.means[, colnames(cell.type.norm.means) != \"Dopaminergic neurons\"]\n",
    "rownames(mat) = fData(cds)$symbol\n",
    "mat = mat[apply(mat, 1, max) >= 0.05,]\n",
    "mat = log2(mat+1)\n",
    "mat = t(scale(t(mat)))\n",
    "mat = ifelse(mat < -2.0, -2.0, mat)\n",
    "mat = ifelse(mat > 2.0, 2.0, mat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "png(\"plots/main.figure.heatmap.png\", units=\"in\", res=300, width=3.0, height=4.05)\n",
    "pheatmap(t(mat), show_colnames = F, clustering_method = \"ward.D2\",\n",
    "    fontsize = 6, treeheight_col = 12, treeheight_row = 12)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Identify cell type specific genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# identify cell type specific size factor\n",
    "pData(cds)$cell.type.specific.size.factor =\n",
    "    (pData(cds) %>% group_by(cell.type) %>% mutate(x = n.umi / median(n.umi)))$x\n",
    "\n",
    "# remove cells with <0.25 size factor\n",
    "pData(cds) %>% group_by(cell.type) %>%\n",
    "    summarize(\n",
    "        n = n(),\n",
    "        prop.excluded = sum(cell.type.specific.size.factor < 0.25) / n(),\n",
    "        n.remaining = sum(cell.type.specific.size.factor >= 0.25)) %>%\n",
    "    arrange(-prop.excluded)\n",
    "\n",
    "cds = cds[, pData(cds)$cell.type.specific.size.factor >= 0.25]\n",
    "\n",
    "# normalize the expression with size factor\n",
    "norm.expr = sweep(exprs(cds), 2, pData(cds)$Size_Factor, \"/\")\n",
    "\n",
    "# tissue normalized expression\n",
    "\n",
    "tissue.norm.means = sapply(setdiff(unique(pData(cds)$tissue), NA), function(x) {\n",
    "    rowMeans(norm.expr[, with(pData(cds), !is.na(tissue) & tissue == x)])\n",
    "})\n",
    "\n",
    "rownames(tissue.norm.means) = fData(cds)$symbol\n",
    "\n",
    "tissue.tpm = sweep(tissue.norm.means, 2, apply(tissue.norm.means, 2, sum), \"/\") * 1000000\n",
    "\n",
    "tissue.tpm = tissue.tpm[, tissues]\n",
    "\n",
    "# identify the gene expresion in two max tissues\n",
    "\n",
    "top.2.tissues.per.gene = do.call(rbind, apply(tissue.tpm, 1, function(x) {\n",
    "    o = order(x, decreasing=T)\n",
    "    y = x[o]\n",
    "    return(list(\n",
    "        max.tissue = tissues[o[1]],\n",
    "        max.tissue.expr = unname(y[1]),\n",
    "        tissue.2 = tissues[o[2]],\n",
    "        tissue.2.expr = unname(y[2])))\n",
    "}))\n",
    "\n",
    "top.2.tissues.per.gene = data.frame(matrix(unlist(\n",
    "    top.2.tissues.per.gene), nrow=nrow(top.2.tissues.per.gene)),\n",
    "    row.names = row.names(top.2.tissues.per.gene), stringsAsFactors=F)\n",
    "\n",
    "names(top.2.tissues.per.gene) = c(\"max.tissue\", \"max.tissue.expr\", \"tissue.2\", \"tissue.2.expr\")\n",
    "top.2.tissues.per.gene$max.tissue.expr = as.numeric(top.2.tissues.per.gene$max.tissue.expr)\n",
    "top.2.tissues.per.gene$tissue.2.expr = as.numeric(top.2.tissues.per.gene$tissue.2.expr)\n",
    "\n",
    "top.2.tissues.per.gene$gene = rownames(top.2.tissues.per.gene)\n",
    "rownames(top.2.tissues.per.gene) = NULL\n",
    "top.2.tissues.per.gene = top.2.tissues.per.gene[, c(5, 1:4)]\n",
    "top.2.tissues.per.gene = mutate(top.2.tissues.per.gene,\n",
    "    ratio = max.tissue.expr / (tissue.2.expr + 1))\n",
    "\n",
    "\n",
    "# tissue specific gene test\n",
    "tissue.specificity.test.res = data.frame(symbol = character(), pval = numeric())\n",
    "\n",
    "# should take about an hour for all ~20k genes\n",
    "for (i in rownames(unique(top.2.tissues.per.gene[, c(\"max.tissue\", \"tissue.2\")]))) {\n",
    "    i = as.integer(i)\n",
    "    this.max.tissue = top.2.tissues.per.gene[i, \"max.tissue\"]\n",
    "    this.tissue.2 = top.2.tissues.per.gene[i, \"tissue.2\"]\n",
    "    \n",
    "    tmp.df = filter(top.2.tissues.per.gene, max.tissue == this.max.tissue, tissue.2 == this.tissue.2)\n",
    "    tmp.cds = cds[fData(cds)$symbol %in% tmp.df$gene, with(pData(cds),\n",
    "        !is.na(tissue) & tissue %in% c(this.max.tissue, this.tissue.2))]\n",
    "        \n",
    "    tmp.test.res = differentialGeneTest(tmp.cds,\n",
    "        fullModelFormulaStr = \"~ tissue\")\n",
    "        \n",
    "    tissue.specificity.test.res = rbind(tissue.specificity.test.res, tmp.test.res[, c(\"symbol\", \"pval\")])\n",
    "}\n",
    "\n",
    "# identify DE genes with ratio >= 5 and q valude < 0.05\n",
    "top.2.tissues.per.gene = inner_join(top.2.tissues.per.gene, tissue.specificity.test.res, by = c(\"gene\" = \"symbol\"))\n",
    "top.2.tissues.per.gene$qval = p.adjust(top.2.tissues.per.gene$pval, method = \"fdr\")\n",
    "\n",
    "tissue.specific.genes = filter(top.2.tissues.per.gene, ratio >= 5.0, qval < 0.05) %>%\n",
    "    filter(gene %in% filter(fData(cds), num_cells_expressed >= 10)$symbol) %>%\n",
    "    arrange(-sqrt(max.tissue.expr * ratio))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### regularized linear regression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "suppressPackageStartupMessages({\n",
    "    library(RColorBrewer)\n",
    "    library(dplyr)\n",
    "    library(ggplot2)\n",
    "    library(glasso)\n",
    "    library(glmnet)\n",
    "    library(igraph)\n",
    "    library(monocle)\n",
    "    library(pheatmap)\n",
    "    library(qgraph)\n",
    "    library(reshape2)\n",
    "    library(scales)\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# filter the peaks with distance to TSS less than 2000, and distance to the second nearest gene\n",
    "# is over 1.5 fold, and the overlap fraction less than 0.2, and generate the matrix with each gene\n",
    "# and the adjacent transcription factors\n",
    "\n",
    "df = filter(peaks, distance.to.TSS < 2000, distance.ratio >= 1.5, overlap.frac < 0.2)\n",
    "df$dummy = 1\n",
    "nrow(df %>% filter(tf != \"DUMMY\")) / nrow(peaks %>% filter(tf != \"DUMMY\"))\n",
    "\n",
    "df = dcast(df, gene ~ tf, value.var = \"overlap.frac\", fill = 0,\n",
    "    fun.aggregate = function(x) (0.2 - min(0.2, x)))\n",
    "        \n",
    "# transform the gene expression matrix\n",
    "mat = cell.type.norm.means[, colnames(cell.type.norm.means) != \"Dopaminergic neurons\"]\n",
    "rownames(mat) = fData(cds.experiment.1)$symbol\n",
    "mat = mat[apply(mat, 1, max) >= 0.05,]\n",
    "mat = log2(mat+1)\n",
    "mat = t(scale(t(mat)))\n",
    "\n",
    "all.heatmap.genes = rownames(mat)\n",
    "all.ts.genes = unique(tissue.specific.genes$gene)\n",
    "        \n",
    "        \n",
    "tissue.tpm.melt = melt(tissue.tpm)\n",
    "names(tissue.tpm.melt) = c(\"gene\", \"tissue\", \"tpm\")\n",
    "tissue.tpm.melt$gene = as.character(tissue.tpm.melt$gene)\n",
    "    \n",
    "# identify the TFs expressed in each tissue\n",
    "tissue.expr.TFs = list()\n",
    "        \n",
    "TF.to.symbol = filter(TF.to.symbol, TF %in% colnames(df))\n",
    "nrow(TF.to.symbol)\n",
    "\n",
    "for (this.tissue in tissues) {\n",
    "    tissue.expr.TFs[[this.tissue]] = (tissue.tpm.melt %>%\n",
    "        filter(tissue == this.tissue, tpm >= 10.0) %>%\n",
    "        inner_join(TF.to.symbol, by = c(\"gene\" = \"symbol\")) %>%\n",
    "        arrange(-tpm))$TF\n",
    "}\n",
    "\n",
    "# apply the linear regression to calculate the correlation of TFs with gene expression\n",
    "ct.expr.models = list()\n",
    "\n",
    "for(cell.type in cell.types) {\n",
    "    message(cell.type)\n",
    "    \n",
    "    y = log2(cell.type.tpm[all.heatmap.genes, cell.type] + 1)\n",
    "    \n",
    "    tmp.df = df[df$gene %in% all.heatmap.genes,]\n",
    "    X = as.matrix(tmp.df[, cell.type.expr.TFs[[cell.type]]])\n",
    "    rownames(X) = tmp.df$gene\n",
    "    X = X[all.heatmap.genes, apply(X, 2, function(x) sum(x > 0) >= 50)]\n",
    "    message(ncol(X))\n",
    "        \n",
    "    ct.expr.models[[cell.type]] = cv.glmnet(X, y, family = \"gaussian\", alpha = 0.5)\n",
    "}\n",
    "\n",
    "*****************\n",
    "for (tissue in tissues) {\n",
    "    message(tissue)\n",
    "\n",
    "    X = as.matrix(df[df$gene %in% all.ts.genes, tissue.expr.TFs[[tissue]]])\n",
    "    X = X[, apply(X, 2, function(x) sum(x > 0)) >= 20]\n",
    "    message(ncol(X))\n",
    "\n",
    "    y = ifelse(df[df$gene %in% all.ts.genes,]$gene %in% ts.gene.lists[[tissue]], 1, 0)\n",
    "\n",
    "    ts.models[[tissue]] = cv.glmnet(X, y, family = \"binomial\", alpha = 0.5)\n",
    "\n",
    "    X = as.matrix(df[df$gene %in% all.heatmap.genes, tissue.expr.TFs[[tissue]]])\n",
    "    X = X[, apply(X, 2, function(x) sum(x > 0)) >= 20]\n",
    "    message(ncol(X))\n",
    "\n",
    "    y = ifelse(df[df$gene %in% all.heatmap.genes,]$gene %in% ts.gene.lists[[tissue]], 1, 0)\n",
    "\n",
    "    ts.models.all.heatmap.genes[[tissue]] = cv.glmnet(X, y, family = \"binomial\", alpha = 0.5)\n",
    "}\n",
    "                  \n",
    "model.coef.df = do.call(rbind, lapply(cell.types, function(ct) { data.frame(\n",
    "    cell.type = ct,\n",
    "    tf = names(coef(ct.expr.models[[ct]], s = \"lambda.1se\")[-1, 1]),\n",
    "    value = coef(ct.expr.models[[ct]], s = \"lambda.1se\")[-1, 1]) }))\n",
    "\n",
    "ts.model.coef = do.call(rbind.data.frame, lapply(ls(ts.models), function(tissue) {\n",
    "    data.frame(\n",
    "        tissue = tissue,\n",
    "        tf = names(coef(ts.models[[tissue]], s = \"lambda.1se\")[-1, 1]),\n",
    "        beta = coef(ts.models[[tissue]], s = \"lambda.1se\")[-1, 1],\n",
    "        stringsAsFactors = F) %>% filter(beta != 0)\n",
    "}))\n",
    "\n",
    "ts.model.coef.all.heatmap.genes = do.call(rbind.data.frame,\n",
    "                                          lapply(ls(ts.models.all.heatmap.genes), function(tissue) {\n",
    "    data.frame(\n",
    "        tissue = tissue,\n",
    "        tf = names(coef(ts.models.all.heatmap.genes[[tissue]], s = \"lambda.1se\")[-1, 1]),\n",
    "        beta = coef(ts.models.all.heatmap.genes[[tissue]], s = \"lambda.1se\")[-1, 1],\n",
    "        stringsAsFactors = F) %>% filter(beta != 0)\n",
    "}))\n",
    "\n",
    "head(ts.model.coef)\n",
    "head(ts.model.coef.all.heatmap.genes)\n",
    "**************\n",
    "                  \n",
    "\n",
    "model.coef.mat = acast(model.coef.df, tf ~ cell.type, value.var = \"value\", fill = -999)\n",
    "model.coef.mat = model.coef.mat[apply(model.coef.mat, 1, max) >= 5,]\n",
    "\n",
    "zork = model.coef.mat\n",
    "zork = ifelse(zork < 0 & zork != -999, 0, zork)\n",
    "zork = ifelse(zork == -999, -0.1, zork)\n",
    "zork = ifelse(zork > 10, 10, zork)\n",
    "\n",
    "# NOTE: with this color bar, 0 coefficient gets grey too, not just not expressed\n",
    "\n",
    "pal = colorRampPalette(rev(brewer.pal(n = 7, name = \"RdYlBu\")))(250)\n",
    "pal[1] = \"grey60\"\n",
    "pal = pal[c(1, 71:250)]\n",
    "                                   \n",
    "rownames(zork) = sub(\"(?<=[A-Z])[.]\", \"-\", rownames(zork), perl = T)\n",
    "rownames(zork) = sub(\"_\", \" \", rownames(zork))\n",
    "rownames(zork) = sub(\"EM\", \"(EM)\", rownames(zork))\n",
    "rownames(zork) = sub(\"PE\", \"(PE)\", rownames(zork))\n",
    "rownames(zork)\n",
    "                                   \n",
    "                                   \n",
    "rownames(zork) = ifelse(rownames(zork) == \"F37D6.2 (PE)\", \"ROW-1 (PE)\", rownames(zork))\n",
    "                                   \n",
    "pdf(\"plots/ChIP.model.coef.pdf\", width = 4, height = 6, onefile = F)\n",
    "pheatmap(zork, scale = \"none\", clustering_method = \"ward.D2\",\n",
    "    treeheight_row = 16, treeheight_col = 16, fontsize = 6, color = pal)\n",
    "dev.off()\n",
    "\n",
    "# learn about the r square\n",
    "get.model.R2 = function(model, min=F) {\n",
    "    if (min)\n",
    "        model$glmnet.fit$dev.ratio[which(model$lambda == model$lambda.min)]\n",
    "    else\n",
    "        model$glmnet.fit$dev.ratio[which(model$lambda == model$lambda.1se)]\n",
    "}\n",
    "        \n",
    "summary(sapply(ct.expr.models, get.model.R2), decreasing = T)\n",
    "sort(sapply(ct.expr.models, get.model.R2), decreasing = T)\n",
    "        \n",
    "        \n",
    "# calculate the enrichment of genes associated with specific TF\n",
    "        \n",
    "genes.with.peak = list()\n",
    "genes.without.peak = list()\n",
    "\n",
    "for (TF in all.TFs) {\n",
    "    if (sum(df[df$gene %in% all.heatmap.genes, TF]) > 0) {\n",
    "        genes.with.peak[[TF]] = df[df$gene %in% all.heatmap.genes & df[, TF] > 0,]$gene\n",
    "        genes.without.peak[[TF]] = df[df$gene %in% all.heatmap.genes & df[, TF] == 0,]$gene\n",
    "    }\n",
    "}\n",
    "        \n",
    "genes.highest.expressed.in = list()\n",
    "\n",
    "for (ct in cell.types) {\n",
    "    genes.highest.expressed.in[[ct]] =\n",
    "        intersect(\n",
    "            (top.2.cell.types.per.gene %>% filter(max.cell.type == ct))$gene,\n",
    "            all.heatmap.genes)\n",
    "}\n",
    "        \n",
    "zork = data.frame(\n",
    "    TF = c(\"HLH.1_EM\", \"HLH.8_EM\",\n",
    "           \"PHA.4_PE\", \"PHA.4_PE\",\n",
    "           \"FKH.8_PE\", \"UNC.86_PE\", \"UNC.55_PE\",\n",
    "           \"NHR.25_PE\", \"NHR.25_PE\",\n",
    "           \"ELT.2_PE\", \"ZTF.7_PE\",\n",
    "           \"XND.1_PE\", \"F49E8.2_PE\"),\n",
    "    cell.type = c(\n",
    "        \"Body wall muscle\", \"Intestinal/rectal muscle\",\n",
    "        \"Pharyngeal muscle\", \"Pharyngeal epithelia\",\n",
    "        \"Ciliated sensory neurons\", \"Touch receptor neurons\", \"GABAergic neurons\",\n",
    "        \"Non-seam hypodermis\", \"Socket cells\",\n",
    "        \"Intestine\", \"Intestine\",\n",
    "        \"Germline\", \"Germline\"),\n",
    "    label = c(\n",
    "        \"HLH-1 (EM)\\nBody wall muscle\",\n",
    "        \"HLH-8 (EM)\\nIntestinal/rectal muscle\",\n",
    "        \"PHA-4 (PE)\\nPharyngeal muscle\",\n",
    "        \"PHA-4 (PE)\\nPharyngeal epithelia\",\n",
    "        \"FKH-8 (PE)\\nCiliated sensory neurons\",\n",
    "        \"UNC-86 (PE)\\nTouch receptor neurons\",\n",
    "        \"UNC-55 (PE)\\nGABAergic neurons\",\n",
    "        \"NHR-25 (PE)\\nNon-seam hypodermis\",\n",
    "        \"NHR-25 (PE)\\nSocket cells\",\n",
    "        \"ELT-2 (PE)\\nIntestine\",\n",
    "        \"ZTF-7 (PE)\\nIntestine\",\n",
    "        \"XND-1 (PE)\\nGermline\",\n",
    "        \"F49E8.2 (PE)\\nGermline\"), stringsAsFactors = F)\n",
    "    \n",
    "zork$prop.with.peak = mapply(function(TF, ct) {\n",
    "    length(intersect(genes.with.peak[[TF]], genes.highest.expressed.in[[ct]])) / length(genes.with.peak[[TF]])\n",
    "}, zork$TF, zork$cell.type)\n",
    "    \n",
    "zork$prop.without.peak = mapply(function(TF, ct) {\n",
    "    length(intersect(genes.without.peak[[TF]], genes.highest.expressed.in[[ct]])) / length(genes.without.peak[[TF]])\n",
    "}, zork$TF, zork$cell.type)\n",
    "    \n",
    "zork$ratio = zork$prop.with.peak / zork$prop.without.peak\n",
    "    \n",
    "zork\n",
    "        \n",
    "tmp.df = melt(zork, id.vars = c(\"TF\", \"cell.type\", \"label\", \"ratio\"),\n",
    "    variable.name = \"variable\", value.name = \"value\")\n",
    "\n",
    "tmp.df$label = factor(tmp.df$label, levels = rev(unique(tmp.df$label)))\n",
    "tmp.df$variable = ifelse(tmp.df$variable == \"prop.with.peak\",\n",
    "    \">0 non-HOT peaks\\nfor TF in promoter\",\n",
    "    \"0 non-HOT peaks\\nfor TF in promoter\")\n",
    "tmp.df$variable = factor(tmp.df$variable, levels = rev(unique(tmp.df$variable)))\n",
    "\n",
    "plot = ggplot(tmp.df, aes(x = label, y = value, fill = variable)) +\n",
    "    geom_bar(stat = \"identity\", position = \"dodge\") +\n",
    "    geom_text(aes(y = value + 0.08, label = round(ratio, 2), group = variable, alpha = variable),\n",
    "        position = position_dodge(1.0), size = 2.0) +\n",
    "    coord_flip() +\n",
    "    scale_y_continuous(limits = c(0, 0.7), breaks = seq(0, 0.8, 0.2), labels = percent) +\n",
    "    scale_fill_manual(values = c(\"dodgerblue3\", \"firebrick\")) +\n",
    "    scale_alpha_manual(values = c(0, 1)) +\n",
    "    xlab(\"\") + ylab(\"Proportion of genes for which the\\ncell type of highest expression is\\nthe cell type associated with the TF\") +\n",
    "    guides(fill = guide_legend(title = \"Gene set\"), alpha = F) +\n",
    "    theme_bw(base_size = 6) +\n",
    "    monocle:::monocle_theme_opts() +\n",
    "    theme(\n",
    "        axis.title.x = element_text(margin = margin(5, 0, 0, 0)),\n",
    "        legend.key.size = unit(0.4, \"cm\"),\n",
    "        legend.margin = margin(0, 0, 0, 0))\n",
    "\n",
    "show(plot)\n",
    "\n",
    "ggsave(\"plots/Fig_S12.pdf\",\n",
    "    device = cairo_pdf, units = \"in\", width = 3.0, height = 4.5)\n",
    "        \n",
    "## calculate the coefficient for the factors\n",
    "        \n",
    "plot.df = ts.model.coef %>% filter(beta >= 3)\n",
    "\n",
    "plot.df$tissue = factor(plot.df$tissue, levels = c(\n",
    "    \"Body wall muscle\", \"Pharynx\", \"Hypodermis\", \"Neurons\", \"Intestine\", \"Gonad\"))\n",
    "\n",
    "plot.df = arrange(plot.df, tissue, -beta)\n",
    "plot.df$tf = sub(\"_\", \" \", sub(\"(?<=[A-Z])[.]\", \"-\", plot.df$tf, perl = T))\n",
    "plot.df$tf = factor(plot.df$tf, levels = rev(unique(plot.df$tf)))\n",
    "\n",
    "ggplot(plot.df, aes(x = tf, y = beta, fill = tissue)) +\n",
    "    geom_bar(stat = \"identity\", position = \"dodge\") +\n",
    "    coord_flip() +\n",
    "    xlab(\"Transcription factor\") +\n",
    "    ylab(\"Elastic net regression coefficient\") +\n",
    "    scale_y_continuous(breaks = seq(-5, 15, 5)) +\n",
    "    scale_fill_manual(values = c(\n",
    "        \"#E41A1C\", \"#377EB8\",  \"#4DAF4A\", \"#984EA3\", \"#E0D90B\", \"#FF7F00\")) +\n",
    "    guides(fill = guide_legend(title = \"Tissue\")) +\n",
    "    theme_bw(base_size = 6) +\n",
    "    monocle:::monocle_theme_opts() +\n",
    "    theme(axis.line = element_line(),\n",
    "          axis.title.x = element_text(margin = margin(4, 0, 0, 0)),\n",
    "          axis.title.y = element_text(margin = margin(0, 4, 0, 0)),\n",
    "          legend.key.size = unit(0.4, \"cm\"),\n",
    "          legend.margin = margin(0, 0, 0, -15))\n",
    "\n",
    "ggsave(\"plots/TF.ChIP.enriched.in.tissue-specific.gene.promoters.pdf\",\n",
    "    device = cairo_pdf, units = \"in\", width = 2.5, height = 5.0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Glasso regression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# generate the input matrix for glasso\n",
    "\n",
    "cluster.df = dcast(PE.peak.clusters.melt,\n",
    "    cluster.id + gene + distance.to.TSS + distance.ratio ~ TF,\n",
    "    value.var = \"dummy\", fun.aggregate = max, fill = 0)\n",
    "\n",
    "cluster.df.TFs = intersect(PE.TFs, unique(do.call(c, cell.type.expr.TFs)))\n",
    "cluster.df.TFs = intersect(sub(\"_PE\", \"\", cluster.df.TFs), colnames(cluster.df))\n",
    "cluster.df.TFs.incl.TS = sort(c(cluster.df.TFs, colnames(cluster.df)[grepl(\":\", colnames(cluster.df))]))\n",
    "\n",
    "length(cluster.df.TFs)\n",
    "length(cluster.df.TFs.incl.TS)\n",
    "\n",
    "cluster.df$n.TFs = apply(cluster.df[, cluster.df.TFs], 1, function(x) sum(x))\n",
    "\n",
    "nrow(cluster.df)\n",
    "cluster.df = cluster.df[, c(\n",
    "    \"cluster.id\", \"gene\", \"n.TFs\", \"distance.to.TSS\", \"distance.ratio\", cluster.df.TFs.incl.TS)] %>%\n",
    "    filter(n.TFs < 0.1 * length(PE.TFs), distance.to.TSS < 2000, distance.ratio >= 1.5)\n",
    "\n",
    "nrow(cluster.df)\n",
    "head(cluster.df)\n",
    "    \n",
    "input.matrices.for.glasso = list()\n",
    "    \n",
    "mat = as.matrix(cluster.df[, cluster.df.TFs.incl.TS])\n",
    "mat = mat[, apply(mat, 2, sum) >= 20]\n",
    "\n",
    "input.matrices.for.glasso[[\"ChIP clusters\"]] = mat\n",
    "rm(mat)\n",
    "    \n",
    "\n",
    "mat = as.matrix(cluster.df[cluster.df$gene %in% all.ts.genes, cluster.df.TFs.incl.TS])\n",
    "mat = mat[, apply(mat, 2, sum) >= 20]\n",
    "\n",
    "input.matrices.for.glasso[[\"ChIP clusters in TS gene promoters\"]] = mat\n",
    "rm(mat)\n",
    "\n",
    "# generate tissue specific input matrix for tissues\n",
    "\n",
    "valid.TFs = cluster.df.TFs.incl.TS\n",
    "valid.TFs = valid.TFs[grepl(\":\", valid.TFs) |\n",
    "                      paste(valid.TFs, \"PE\", sep=\"_\") %in% tissue.expr.TFs[[\"Hypodermis\"]]]\n",
    "length(valid.TFs)\n",
    "\n",
    "mat = as.matrix(cluster.df[cluster.df$gene %in% ts.gene.lists[[\"Hypodermis\"]], valid.TFs])\n",
    "mat = mat[, apply(mat, 2, sum) >= 20]\n",
    "\n",
    "input.matrices.for.glasso[[\"ChIP clusters in hypodermis enriched gene promoters\"]] = mat\n",
    "rm(list = c(\"mat\", \"valid.TFs\"))\n",
    "    \n",
    "# shuff the matrix\n",
    "    \n",
    "# http://www.nature.com/articles/ncomms5114\n",
    "curve.ball = function(m) {\n",
    "    RC = dim(m)\n",
    "    R = RC[1]\n",
    "    C = RC[2]\n",
    "    hp = list()\n",
    "    for (row in 1:dim(m)[1]) {\n",
    "        hp[[row]] = which(m[row,] == 1)\n",
    "    }\n",
    "\n",
    "    l_hp = length(hp)\n",
    "    for (rep in 1:(5*l_hp)) {\n",
    "        AB = sample(1:l_hp, 2)\n",
    "        a = hp[[AB[1]]]\n",
    "        b = hp[[AB[2]]]\n",
    "        ab = intersect(a,b)\n",
    "        l_ab = length(ab)\n",
    "        l_a = length(a)\n",
    "        l_b = length(b)\n",
    "        if ((l_ab %in% c(l_a,l_b)) == F) {\n",
    "            tot = setdiff(c(a,b), ab)\n",
    "            l_tot = length(tot)\n",
    "            tot = sample(tot, l_tot, replace = FALSE, prob = NULL)\n",
    "            L = l_a - l_ab\n",
    "            hp[[AB[1]]] = c(ab,tot[1:L])\n",
    "            hp[[AB[2]]] = c(ab,tot[(L+1):l_tot])\n",
    "        }\n",
    "    }\n",
    "\n",
    "    rm = matrix(0, R, C)\n",
    "    for (row in 1:R) {\n",
    "        rm[row, hp[[row]]] = 1\n",
    "    }\n",
    "\n",
    "    rownames(rm) = rownames(m)\n",
    "    colnames(rm) = colnames(m)\n",
    "\n",
    "    Matrix(rm, sparse=T)\n",
    "}\n",
    "    \n",
    "shuffled.matrices.for.glasso = list()\n",
    "\n",
    "for (i in 1:30) {\n",
    "    message(i)\n",
    "    shuffled.matrices.for.glasso[[i]] = list()\n",
    "    \n",
    "    for (name in names(input.matrices.for.glasso)) {\n",
    "        shuffled.matrices.for.glasso[[i]][[name]] = curve.ball(input.matrices.for.glasso[[name]])\n",
    "    }\n",
    "}\n",
    "\n",
    "# from glasso to network\n",
    "\n",
    "glasso.res.to.network.df = function(glasso.res, v) {\n",
    "    pcor = glasso.res$wi /\n",
    "        (-sqrt(diag(glasso.res$wi) %*% t(diag(glasso.res$wi))))\n",
    "    row.names(pcor) = row.names(v)\n",
    "    colnames(pcor) = colnames(v)\n",
    "    pcor.melt = melt(pcor)\n",
    "    names(pcor.melt) = c(\"from\", \"to\", \"pcor\")\n",
    "    \n",
    "    rv = solve(glasso.res$wi)\n",
    "    row.names(rv) = row.names(v)\n",
    "    colnames(rv) = colnames(v)\n",
    "    rc = cov2cor(rv)\n",
    "    rc.melt = melt(rc)\n",
    "    names(rc.melt) = c(\"from\", \"to\", \"reg.cor\")\n",
    "    \n",
    "    network.df = pcor.melt %>%\n",
    "        inner_join(rc.melt, by=c(\"from\", \"to\")) %>%\n",
    "        filter(pcor > 0, reg.cor > 0) %>%\n",
    "        mutate(weight = pcor) %>%\n",
    "        arrange(-weight)\n",
    "    \n",
    "    return(network.df)\n",
    "}\n",
    "    \n",
    "# calculate the number of edges for different parameter for shuffled sample\n",
    "    \n",
    "rho.list = c(seq(0.01, 0.2, 0.002))\n",
    "\n",
    "nnz.edges.in.shuffled.matrices = list()\n",
    "for (name in ls(input.matrices.for.glasso)) {\n",
    "    nnz.edges.in.shuffled.matrices[[name]] = list()\n",
    "    for (j in 1:length(rho.list)) {\n",
    "        nnz.edges.in.shuffled.matrices[[name]][[j]] = integer()\n",
    "    }\n",
    "}\n",
    "\n",
    "for (i in 1:30) {\n",
    "    message(i)\n",
    "    for (name in ls(input.matrices.for.glasso)) {\n",
    "        v = cov(as.matrix(shuffled.matrices.for.glasso[[i]][[name]]))\n",
    "        glasso.res = list()\n",
    "\n",
    "        for (j in 1:length(rho.list)) {\n",
    "            if (j == 1) {\n",
    "                glasso.res[[j]] = glasso(v, rho = rho.list[j])\n",
    "            } else {\n",
    "                glasso.res[[j]] = glasso(v, rho = rho.list[j], start = \"warm\",\n",
    "                    w.init = glasso.res[[j-1]]$w, wi.init = glasso.res[[j-1]]$wi)\n",
    "            }\n",
    "\n",
    "            network.df = glasso.res.to.network.df(glasso.res[[j]], v)\n",
    "            nnz.edges.in.shuffled.matrices[[name]][[j]] = c(\n",
    "                nnz.edges.in.shuffled.matrices[[name]][[j]], nrow(network.df))\n",
    "        }\n",
    "    }\n",
    "}\n",
    "    \n",
    "\n",
    "# calculate the number of edges in the real sample with different parameters\n",
    "glasso.res = list()\n",
    "for (name in ls(input.matrices.for.glasso)) {\n",
    "    glasso.res[[name]] = list()\n",
    "}\n",
    "\n",
    "for (name in ls(input.matrices.for.glasso)) {\n",
    "    message(name)\n",
    "    v = cov(as.matrix(input.matrices.for.glasso[[name]]))\n",
    "    \n",
    "    for (j in 1:length(rho.list)) {\n",
    "        if (j == 1) {\n",
    "            glasso.res[[name]][[j]] = glasso(v, rho = rho.list[j])\n",
    "        } else {\n",
    "            glasso.res[[name]][[j]] = glasso(v, rho = rho.list[j], start = \"warm\",\n",
    "                w.init = glasso.res[[name]][[j-1]]$w, wi.init = glasso.res[[name]][[j-1]]$wi)\n",
    "        }\n",
    "\n",
    "        glasso.res[[name]][[j]]$network.df = glasso.res.to.network.df(glasso.res[[name]][[j]], v)\n",
    "        glasso.res[[name]][[j]]$est.fdr =\n",
    "            mean(nnz.edges.in.shuffled.matrices[[name]][[j]]) /\n",
    "            nrow(glasso.res[[name]][[j]]$network.df)\n",
    "    }\n",
    "}\n",
    "\n",
    "# select the parameter \n",
    "selected.rho.indices = lapply(glasso.res, function(x) {\n",
    "    which(sapply(x, function(y) y$est.fdr) < 0.05)[1]\n",
    "})\n",
    "          \n",
    "sapply(selected.rho.indices, function(x) rho.list[x])\n",
    "    \n",
    "# generate the network\n",
    "glasso.networks = list()\n",
    "\n",
    "for (name in ls(glasso.res)) {\n",
    "    glasso.networks[[name]] = glasso.res[[name]][[selected.rho.indices[[name]]]]$network.df\n",
    "    glasso.networks[[name]]$from = as.character(glasso.networks[[name]]$from)\n",
    "    glasso.networks[[name]]$to = as.character(glasso.networks[[name]]$to)\n",
    "}\n",
    "    \n",
    "make.network = function(network.df, pcor.thresh = 0.0) {\n",
    "    df = network.df %>% filter(pcor > pcor.thresh)\n",
    "    df2 = df\n",
    "    df2$tmp = df2$from\n",
    "    df2$from = df2$to\n",
    "    df2$to = df2$tmp\n",
    "    df2$tmp = NULL\n",
    "    df = rbind(df, df2)\n",
    "    df = df %>% group_by(from, to) %>% summarize(n = n(), weight = max(pcor)) %>% ungroup()\n",
    "\n",
    "    g = graph.data.frame(df %>% filter(from < to) %>% arrange(-weight), directed=F)\n",
    "    return(g)\n",
    "}\n",
    "    \n",
    "glasso.igraph = list()\n",
    "\n",
    "for (name in ls(glasso.networks)) {\n",
    "    glasso.igraph[[name]] = make.network(glasso.networks[[name]], 0.01)\n",
    "}\n",
    "    \n",
    "# plot the network\n",
    "    \n",
    "plot.network = function(g, file, seed = 42, width = 4.0, height = 4.0,\n",
    "                        vertex.size = 28, vertex.color = \"#D4DCFF\", edge.width.scaling = 100.0,\n",
    "                        label.cex = 0.7, small.label.cex = 0.6, small.label.vertices = NULL,\n",
    "                        name.map = NULL, repulse.rad.exp = 3.0, fr.n.iter = 500) {\n",
    "\n",
    "    V(g)$color = vertex.color\n",
    "    V(g)$label.cex = label.cex\n",
    "    E(g)$color = \"black\"\n",
    "    \n",
    "    if (!is.null(small.label.vertices)) {\n",
    "        V(g)$label.cex = sapply(V(g)$name, function(name) {\n",
    "            ifelse(name %in% small.label.vertices, small.label.cex, label.cex)\n",
    "        })\n",
    "    }\n",
    "    \n",
    "    tmp.vec = V(g)$name\n",
    "    names(tmp.vec) = tmp.vec\n",
    "    full.name.map = lapply(tmp.vec, function(x) x)\n",
    "    if (!is.null(name.map)) {\n",
    "        for (key in names(name.map))\n",
    "            full.name.map[[key]] = name.map[[key]]\n",
    "    }\n",
    "\n",
    "    V(g)$name = sapply(V(g)$name, function(term) full.name.map[[term]])\n",
    "\n",
    "    cairo_pdf(file, width = width, height = height)\n",
    "    set.seed(seed)\n",
    "\n",
    "    g.layout = qgraph.layout.fruchtermanreingold(\n",
    "        get.edgelist(g, names=F), weights = E(g)$weight, vcount = vcount(g),\n",
    "        repulse.rad = vcount(g)^repulse.rad.exp, niter = fr.n.iter)\n",
    "\n",
    "    par(mar=c(0,0,0,0))\n",
    "    plot(g,\n",
    "        edge.width = edge.width.scaling * E(g)$weight,\n",
    "        vertex.size = vertex.size,\n",
    "        vertex.label.family = \"sans\",\n",
    "        vertex.label.color = \"black\",\n",
    "        edge.color = E(g)$color,\n",
    "        layout = g.layout)\n",
    "    dev.off()\n",
    "}\n",
    "                       \n",
    "\n",
    "name.map = list(\n",
    "    \"LIN.9\" = \"LIN-9\",\n",
    "    \"LIN.54\" = \"LIN-54\",\n",
    "    \"EFL.1\" = \"EFL-1\",\n",
    "    \"XND.1\" = \"XND-1\",\n",
    "    \"Germline:EFL.1\" = \"EFL-1\\n(GS)\",\n",
    "    \"NHR.20\" = \"NHR-20\",\n",
    "    \"SNU.23\" = \"SNU-23\",\n",
    "    \"NHR.25\" = \"NHR-25\",\n",
    "    \"BLMP.1\" = \"BLMP-1\",\n",
    "    \"NHR.23\" = \"NHR-23\",\n",
    "    \"PHA.4\" = \"PHA-4\",\n",
    "    \"DAF.16\" = \"DAF-16\",\n",
    "    \"DVE.1\" = \"DVE-1\",\n",
    "    \"NHR.80\" = \"NHR-80\",\n",
    "    \"PQM.1\" = \"PQM-1\",\n",
    "    \"NHR.28\" = \"NHR-28\",\n",
    "    \"ELT.2\" = \"ELT-2\",\n",
    "    \"FOS.1\" = \"FOS-1\",\n",
    "    \"PQM.1\" = \"PQM-1\",\n",
    "    \"NHR.28\" = \"NHR-28\",\n",
    "    \"SMA.9\" = \"SMA-9\",\n",
    "    \"NHR.71\" = \"NHR-71\")\n",
    "\n",
    "plot.network(glasso.igraph[[\"ChIP clusters\"]],\n",
    "    file = \"plots/Fig_S13d.pdf\", width = 3.0, height = 3.0,\n",
    "    vertex.size = 32 * 3.0/2.5 * 0.8, edge.width.scaling = 60.0, name.map = name.map,\n",
    "    label.cex = 0.55, small.label.vertices = c(\"T02C12.2\"), small.label.cex = 0.5)\n",
    "    \n",
    "                       \n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.3.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
