{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting ./scATAC_pipe.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile ./scATAC_pipe.sh\n",
    "\n",
    "read1=\"../../nobackup/170109/fastq/Undetermined_S0_R1_001.fastq.gz\"\n",
    "read2=\"../../nobackup/170109/fastq/Undetermined_S0_R2_001.fastq.gz\"\n",
    "output_folder=\"../../nobackup/170109/scATAC_output/\"\n",
    "P7_index=\"./P7_index.txt\"\n",
    "P5_index=\"./P5_index.txt\"\n",
    "barcode_file=\"/net/shendure/vol1/home/cao1025/analysis_script/ATAC_RNA_coassay_pipe/N7_N5_barcode_txt\"\n",
    "core=20\n",
    "cutoff=500\n",
    "\n",
    "script_folder=\"/net/shendure/vol1/home/cao1025/analysis_script/ATAC_RNA_coassay_pipe/scATAC_seq/\"\n",
    "python=\"/net/shendure/vol1/home/cao1025/anaconda2/bin/python2.7\"\n",
    "\n",
    "\n",
    "# read in the read1 and read2, and P5 barcode, P7 barcode and then split the read1 and read2 based on the P5 barcode\n",
    "# and P7 barcode, and output the splitted reads to the output_folder/fastq files\n",
    "fastq_folder=$output_folder/fastq\n",
    "mkdir -p $fastq_folder\n",
    "$python $script_folder/ATAC_split_fastq.py $read1 $read2 $fastq_folder $P7_index $P5_index\n",
    "gzip $fastq_folder/*.fastq\n",
    "\n",
    "# Trim the reads\n",
    "echo\n",
    "echo \"Start trimming the files...\"\n",
    "echo $(date)\n",
    "module load python/2.7.3\n",
    "module load cutadapt/1.8.3\n",
    "module load trim_galore/0.4.1\n",
    "mkdir $output_folder/trimmed_fastq\n",
    "trimmed_fastq=$output_folder/trimmed_fastq\n",
    "sample_ID=$output_folder/fastq/sample_ID.txt\n",
    "mkdir -p $trimmed_fastq\n",
    "for sample in $(cat $sample_ID); do echo trimming $sample; sem -j+$core trim_galore $fastq_folder/$sample*R1*.gz $fastq_folder/$sample*R2*.gz --paired -a CTGTCTCTTATA -a2 CTGTCTCTTATA --three_prime_clip_R1 1 --three_prime_clip_R2 1 -o $trimmed_fastq; done\n",
    "sem --semaphoretimeout 1200\n",
    "echo \"All trimmed file generated.\"\n",
    "module unload python/2.7.3\n",
    "\n",
    "\n",
    "# align the reads with bowtie2\n",
    "# this script take the input folder, sample ID, output folder and index as input, then it runs the bowtie2 with local alignment, and align single end read to the index\n",
    "input_folder=$trimmed_fastq\n",
    "sample_ID=$output_folder/fastq/sample_ID.txt\n",
    "bowtie_folder=$output_folder/bowtie2_align\n",
    "mkdir -p $bowtie_folder\n",
    "index=\"/net/shendure/vol10/nobackup/shared/alignments/bowtie2-2.0.2/human_g1k_hs37d5/hs37d5\"\n",
    "echo\n",
    "echo \"Start the alignment using bowtie2\"\n",
    "echo input folder: $input_folder\n",
    "echo sample ID file: $sample_ID\n",
    "echo output folder: $bowtie_folder\n",
    "echo index: $index\n",
    "module load bowtie2/2.2.3\n",
    "mkdir -p $output_folder\n",
    "for sample in $(cat $sample_ID); do echo aligning $sample; bowtie2 -p $core --local -X 2000 -x $index -1 $input_folder/$sample*R1*.gz -2 $input_folder/$sample*R2*.gz -S $bowtie_folder/$sample.sam 2>>$bowtie_folder/report.txt; done\n",
    "echo \"all alignment done\"\n",
    "\n",
    "\n",
    "# This function accept a input bowtie2  folder, a sample ID file and a output folder, and then filter the reads: remove the mitochondrial reads, samtools -F4 -q 30, sort the file and use picard to remove the dupllicate; and output the mapped reads into the mapped folder of the output folder, output the remove duplicates reads into the rm_dup folder of the output folder\n",
    "input_folder=$bowtie_folder\n",
    "sample_ID=$output_folder/fastq/sample_ID.txt\n",
    "filtered_folder=$output_folder/filtered_bam\n",
    "# make the output folder\n",
    "mkdir -p $filtered_folder\n",
    "mkdir $filtered_folder/mapped_bam\n",
    "mkdir $filtered_folder/rmdup_bam\n",
    "# filter the files and generate the mapped bam file\n",
    "for sample in $(cat $sample_ID); do echo \"generating mapped bam file\" $sample; sem -j+$core samtools view -h -F 4 -q 30 $input_folder/$sample.sam |awk '$3 != \"MT\" && $3 != \"chrM\"' -|samtools view -bh -|samtools sort -|samtools view -bh>$filtered_folder/mapped_bam/$sample.bam; done\n",
    "sem --semaphoretimeout 1200\n",
    "echo All mapped file generated~\n",
    "# remove duplicates using Picard\n",
    "echo Remove duplicates\n",
    "for sample in $(cat $sample_ID); do echo removing duplicates $sample; sem -j+$core java -Xmx1G -jar /net/gs/vol3/software/modules-sw/picard/1.105/Linux/RHEL6/x86_64/MarkDuplicates.jar INPUT=$filtered_folder/mapped_bam/$sample.bam OUTPUT=$filtered_folder/rmdup_bam/$sample.bam REMOVE_DUPLICATES=true ASSUME_SORTED=True METRICS_FILE=/net/shendure/vol1/home/cao1025/sge_logs/CurrSC.metrics.txt VALIDATION_STRINGENCY=LENIENT; done\n",
    "sem --semaphoretimeout 1200\n",
    "echo All remove duplicated files generated~ \n",
    "\n",
    "# Transform the bam files to sam files\n",
    "bam_folder=$filtered_folder/rmdup_bam\n",
    "sample_list=$output_folder/fastq/sample_ID.txt\n",
    "splitted_folder=$output_folder/splitted_sam\n",
    "# first convert the bam files into sam files\n",
    "echo convert bam files to sam files...\n",
    "for sample in $(cat $sample_list); do echo converting $sample; sem -j+$core samtools view -h $bam_folder/$sample.bam>$bam_folder/$sample.sam; done\n",
    "sem --semaphoretimeout 1200\n",
    "echo All bam files are converted to sam files.\n",
    "\n",
    "echo Start splitting the sam files based on the barcode...\n",
    "mkdir -p $splitted_folder\n",
    "bash $script_folder/samfile_split_permuted.sh $bam_folder $sample_list $splitted_folder $barcode_file $cutoff $core\n",
    "\n",
    "echo \"Start calculating the human and mouse reads number...\"\n",
    "bash $script_folder/report_human_mouse_fraction.sh $splitted_folder $output_folder/barcode_samples.txt $output_folder/report/human_mouse_read_number\n",
    "\n",
    "echo \"Calculate the fragment size distribution....\"\n",
    "bash $script_folder/frag_size.sh $splitted_folder $output_folder/barcode_samples.txt $output_folder/report/frag_size $core\n",
    "echo \"All fragment size calculation is done.\"\n",
    "\n",
    "\n",
    "# calculate the reads number\n",
    "# this script accept the input parental folder, create a report folder and a read_number sub folder, and for each sample, \n",
    "# calculate the read number in fastq, trimmed_fastq, UMI_attached_R2, STAR_alignment, filtered_sam, samfile. after filter\n",
    "# barcode\n",
    "parental_folder=$output_folder\n",
    "sample_ID=$output_folder/fastq/sample_ID.txt\n",
    "fastq_folder=$parental_folder/fastq\n",
    "trimmed_folder=$parental_folder/trimmed_fastq\n",
    "alignment=$bowtie_folder\n",
    "filtered_sam=$filtered_folder/mapped_bam\n",
    "rm_dup_sam=$filtered_folder/rmdup_bam\n",
    "#split_sam=$parental_folder/splited_sam\n",
    "report_folder=$parental_folder/report/read_number\n",
    "echo\n",
    "echo \"Start calculating the reads number...\"\n",
    "#make the report folder\n",
    "mkdir -p $report_folder\n",
    "#calculate the read number and output the read number into the report folder\n",
    "echo sample,total reads,after trimming,all mapped reads, uniquely aligned reads,After remove duplicates>$report_folder/read_number.csv\n",
    "for sample in $(cat $sample_ID); do echo calculating $sample; echo $sample,$(expr $(zcat $fastq_folder/$sample*R2*.gz|wc -l) / 4),$(expr $(zcat $trimmed_folder/$sample*R2*.gz|wc -l) / 4), $(samtools view -F 4 -q 30 $alignment/$sample.sam|wc -l),$(samtools view $filtered_sam/$sample.bam|wc -l),$(samtools view $rm_dup_sam/$sample.bam|wc -l)>>$report_folder/read_number.csv; done\n",
    "echo \"Read number calculation is done.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting ./ATAC_split_fastq.py\n"
     ]
    }
   ],
   "source": [
    "%%writefile ./ATAC_split_fastq.py\n",
    "# python ATAC_split_fastq.py read1 read2 output_folder P7_index P5_index\n",
    "'''\n",
    "This script accept a read1, read2 file, a barcode file P5 and P7 file, output\n",
    "file folder, then it read in the read1 and read2 file, split the read1 and read2 based on\n",
    "the combination of the P5 and P7 barcode, and attach the adaptor barcode to the read1 and read2 \n",
    "sequence\n",
    "'''\n",
    "\n",
    "'''\n",
    "First, I am going to create the files with the P5-P7 index combination,\n",
    "\n",
    "For each read1, I will go through line by line, for line with the index line,\n",
    "extract the index, and file which this index belong to, and then write the read1 and read2 both into the \n",
    "output file.\n",
    "'''\n",
    "import subprocess\n",
    "import sys\n",
    "from Levenshtein import distance\n",
    "import gzip\n",
    "\n",
    "def extract_barcode(index):\n",
    "    '''\n",
    "    This function accept a index sequence, and return the extracted P7+P5 barcode\n",
    "    in the index, and the barcode in the Tn5 adaptor\n",
    "    '''\n",
    "    barcodes = index.split('+')\n",
    "    P7 = barcodes[0][8:]\n",
    "    P5 = barcodes[1][0:10]\n",
    "    N7 = barcodes[0][0:8]\n",
    "    N5 = barcodes[1][10:]\n",
    "    \n",
    "    P_index = P7 + '.' + P5\n",
    "    N_index = N7 + '.' + N5\n",
    "    \n",
    "    return (P_index, N_index)\n",
    "\n",
    "'''\n",
    "def P7_P5_list(P7_file, P5_file):\n",
    "    \n",
    "    #this function accept a text including P7 index and a text file including a \n",
    "    #P5 index and then generate a list for all the P7.P5 barcode combinations\n",
    "    \n",
    "    P7 = open(P7_file)\n",
    "    P5 = open(P5_file)\n",
    "    index1 = P7.readlines()\n",
    "    index2 = P5.readlines()\n",
    "    index1 = map(lambda x: x.strip(), index1)\n",
    "    index2 = map(lambda x: x.strip(), index2)\n",
    "    all_indexes = []\n",
    "    for i1 in index1:\n",
    "        for i2 in index2:\n",
    "            all_indexes.append(i1 + '.' + i2)\n",
    "    \n",
    "    P7.close()\n",
    "    P5.close()\n",
    "    return all_indexes\n",
    "'''\n",
    "\n",
    "def P7_P5_list(P7_file, P5_file):\n",
    "    '''\n",
    "    this function accept a text including P7 index and a text file including a \n",
    "    P5 index and then generate a list for all the P7.P5 barcode combinations\n",
    "    '''\n",
    "    P7 = open(P7_file)\n",
    "    P5 = open(P5_file)\n",
    "    index1 = P7.readlines()\n",
    "    index2 = P5.readlines()\n",
    "    index1 = map(lambda x: x.strip(), index1)\n",
    "    index2 = map(lambda x: x.strip(), index2)\n",
    "    all_indexes = []\n",
    "    for i in range(len(index1)):\n",
    "            all_indexes.append(index1[i] + '.' + index2[i])\n",
    "    \n",
    "    P7.close()\n",
    "    P5.close()\n",
    "    return all_indexes\n",
    "\n",
    "\n",
    "def find_P7_P5_barcode(index, P7_P5_barcodes):\n",
    "    '''\n",
    "    this function accept a P7_P5 barcode in the read, and a list of P7_P5 barcodes provided,\n",
    "    and then find the barcode in the P7_P5 barcode list that match the index with distance <2,\n",
    "    then return the found barcode; if no barcode is found, then return -1\n",
    "    '''\n",
    "    result = -1\n",
    "\n",
    "    for barcode in P7_P5_barcodes:\n",
    "        diff = distance(index, barcode)\n",
    "        if (diff <= 1):\n",
    "            result = barcode\n",
    "            break\n",
    "    \n",
    "    return result\n",
    "    \n",
    "def split_fastq(read1, read2, output_folder, P7_P5_barcodes): \n",
    "    '''\n",
    "    This function accept a read1 file, a read2 file, a list of P7-P5 barcodes, create the output files,\n",
    "    and then go through each line of the read1 and read2 file,\n",
    "    and then check the barcode in the read1, compare it with the P7_P5 barcodes,\n",
    "    if the distance < 2, then output the reads into the output files\n",
    "    '''\n",
    "    \n",
    "    # generate the output files\n",
    "    output_read1 = {}\n",
    "    output_read2 = {}\n",
    "    sample_ID = open(output_folder + \"/sample_ID.txt\", 'w')\n",
    "    barcode_count = {}\n",
    "    for barcode in P7_P5_barcodes:\n",
    "        output_R1 = output_folder + '/' + barcode + '.R1.fastq'\n",
    "        output_R2 = output_folder + '/' + barcode + '.R2.fastq'\n",
    "        output_read1[barcode] = open(output_R1, 'w')\n",
    "        output_read2[barcode] = open(output_R2, 'w')\n",
    "        sample_ID.write(barcode + '\\n')\n",
    "        barcode_count[barcode] = 0\n",
    "        \n",
    "    sample_ID.close()\n",
    "    # open the read1 and read2 file, and then for each read, check the barcode and output the matched read into the\n",
    "    # output file\n",
    "    \n",
    "    f1 = gzip.open(read1, 'rb')\n",
    "    f2 = gzip.open(read2, 'rb')\n",
    "    line_n = 0\n",
    "    total_read=0\n",
    "    \n",
    "    while(True):\n",
    "        line1 = f1.readline()\n",
    "        line2 = f2.readline()\n",
    "        line_n += 1\n",
    "        if (not line1):\n",
    "            break\n",
    "        if (line_n % 4 == 1):\n",
    "            index = line1.strip().split(':')[-1]\n",
    "            #print \"index: \", index\n",
    "            index = extract_barcode(index)\n",
    "            P_index = index[0]\n",
    "            N_index = index[1]\n",
    "            P_index = find_P7_P5_barcode(P_index, P7_P5_barcodes)\n",
    "            total_read+=1\n",
    "            if (P_index != -1):\n",
    "                output_R1 = output_read1[P_index]\n",
    "                output_R2 = output_read2[P_index]\n",
    "                barcode_count[P_index] += 1\n",
    "                line1 = '@' + N_index + ',' + line1[1:]\n",
    "                line2 = '@' + N_index + ',' + line2[1:]\n",
    "                output_R1.write(line1)\n",
    "                output_R2.write(line2)\n",
    "                \n",
    "                # output the other lines into the files\n",
    "                line1 = f1.readline()\n",
    "                line2 = f2.readline()\n",
    "                output_R1.write(line1)\n",
    "                output_R2.write(line2)\n",
    "                line_n += 1\n",
    "                \n",
    "                line1 = f1.readline()\n",
    "                line2 = f2.readline()\n",
    "                output_R1.write(line1)\n",
    "                output_R2.write(line2)\n",
    "                line_n += 1\n",
    "                \n",
    "                line1 = f1.readline()\n",
    "                line2 = f2.readline()\n",
    "                output_R1.write(line1)\n",
    "                output_R2.write(line2)\n",
    "                line_n += 1\n",
    "                \n",
    "            else:\n",
    "                line1 = f1.readline()\n",
    "                line2 = f2.readline()\n",
    "                line_n += 1\n",
    "                \n",
    "                line1 = f1.readline()\n",
    "                line2 = f2.readline()\n",
    "                line_n += 1\n",
    "                \n",
    "                line1 = f1.readline()\n",
    "                line2 = f2.readline()\n",
    "                line_n += 1\n",
    "    # close the files\n",
    "    f1.close()\n",
    "    f2.close()\n",
    "    print \"Total read count: \", total_read\n",
    "    asigned_read = 0\n",
    "    for barcode in P7_P5_barcodes:\n",
    "        output_read1[barcode].close()\n",
    "        output_read2[barcode].close()\n",
    "        print \"Read_count: \", barcode, barcode_count[barcode]\n",
    "        asigned_read += barcode_count[barcode]\n",
    "    print \"Total asigned read: \", asigned_read\n",
    "\n",
    "def split_fastq_main(read1, read2, output_folder, P7_file, P5_file):\n",
    "    '''\n",
    "    this is the main function that accept the read1 file, the read2 file, the output folder,\n",
    "    the files including P7 barcode, P5 barcode and then split the \n",
    "    the read1 and read2 files based on the P5 and P7 combinations\n",
    "    '''\n",
    "    print \"Start splitting the fastq files...\"\n",
    "    P7_P5_barcodes = P7_P5_list(P7_file, P5_file)\n",
    "    print \"Number of output files: \", len(P7_P5_barcodes)\n",
    "    \n",
    "    split_fastq(read1, read2, output_folder, P7_P5_barcodes)\n",
    "    print \"fastq files splitted~\"\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    read1 = sys.argv[1]\n",
    "    read2 = sys.argv[2]\n",
    "    output_folder = sys.argv[3]\n",
    "    P7_file = sys.argv[4]\n",
    "    P5_file = sys.argv[5]\n",
    "    split_fastq_main(read1, read2, output_folder, P7_file, P5_file)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# %load ./bowtie2_PE.sh\n",
    "#!/bin/bash\n",
    "\n",
    "# this script take the input folder, sample ID, output folder and index as input, then it runs the bowtie2 with local alignment, and align single end read to the index\n",
    "input_folder=$1\n",
    "sample_ID=$2\n",
    "output_folder=$3\n",
    "index=$4\n",
    "\n",
    "echo\n",
    "echo \"Start the alignment using bowtie2\"\n",
    "echo input folder: $input_folder\n",
    "echo sample ID file: $sample_ID\n",
    "echo output folder: $output_folder\n",
    "echo index: $index\n",
    "module load bowtie2/latest\n",
    "mkdir -p $output_folder\n",
    "for sample in $(cat $sample_ID); do echo aligning $sample; bowtie2 -p 10 --local -X 2000 -x $index -1 $input_folder/$sample*R1*.gz -2 $input_folder/$sample*R2*.gz -S $output_folder/$sample.sam 2>>$output_folder/report.txt; done\n",
    "echo \"all alignment done\"\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# %load /net/shendure/vol1/home/cao1025/analysis_script/ATAC_RNA_coassay_pipe/scATAC_seq/bamfile_split_permuted.sh\n",
    "#! /bin/bash\n",
    "\n",
    "# this script accept  a input folder for the bam file, a sample ID file, and a output folder, a barcode file, and a cutoff value, and then transform all bam files in the bam folder into sam files, and then use the samfile_split_permuted.sh to split the sam files\n",
    "\n",
    "bam_folder=$1\n",
    "sample_list=$2\n",
    "output_folder=$3\n",
    "barcode_file=$4\n",
    "cutoff=$5\n",
    "\n",
    "# first convert the bam files into sam files\n",
    "echo convert bam files to sam files...\n",
    "for sample in $(cat $sample_list); do echo converting $sample; samtools view -h $bam_folder/$sample.bam>$bam_folder/$sample.sam; done\n",
    "echo All bam files are converted to sam files.\n",
    "\n",
    "echo Start split the sam files based on the barcode...\n",
    "/net/shendure/vol1/home/cao1025/analysis_script/ATAC_RNA_coassay_pipe/samfile_split_permuted.sh $bam_folder $sample_list $output_folder $barcode_file $cutoff\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting /net/shendure/vol1/home/cao1025/analysis_script/ATAC_RNA_coassay_pipe/scATAC_seq/samfile_split_permuted.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile /net/shendure/vol1/home/cao1025/analysis_script/ATAC_RNA_coassay_pipe/scATAC_seq/samfile_split_permuted.sh\n",
    "#!/bin/bash\n",
    "\n",
    "# this script accept a sam file folder, a sample list, a output folder, a barcode file, then it will run the sam_split.py on each\n",
    "# sam file and output the splited sam file to the output folder\n",
    "\n",
    "sam_folder=$1\n",
    "sample_list=$2\n",
    "output_folder=$3\n",
    "barcode_file=$4\n",
    "cutoff=$5\n",
    "core=$6\n",
    "\n",
    "#define the location of the python\n",
    "python_path=\"/net/shendure/vol1/home/cao1025/anaconda2/bin/python2.7\"\n",
    "script_path=\"/net/shendure/vol1/home/cao1025/analysis_script/ATAC_RNA_coassay_pipe/scATAC_seq/\"\n",
    "\n",
    "echo\n",
    "echo \"Start splitting the sam file...\"\n",
    "echo samfile folder: $sam_folder\n",
    "echo sample list: $sample_list\n",
    "echo ouput folder: $output_folder\n",
    "echo barcode file: $barcode_file\n",
    "echo cutoff value: $cutoff\n",
    "\n",
    "mkdir $output_folder\n",
    "for sample in $(cat $sample_list); do echo Now splitting $sample; sem -j+$core $python_path $script_path/sam_split_permuted.py $sam_folder/$sample.sam $barcode_file $output_folder $cutoff; done\n",
    "sem --wait\n",
    "cat $output_folder/*sample_list.txt>$output_folder/All_samples.txt\n",
    "cp $output_folder/All_samples.txt $output_folder/../barcode_samples.txt\n",
    "\n",
    "# output the report the report/barcode_read_distribution folder\n",
    "mkdir -p $output_folder/../report/barcode_read_distribution\n",
    "mv $output_folder/*.txt $output_folder/../report/barcode_read_distribution/\n",
    "mv $output_folder/*.png $output_folder/../report/barcode_read_distribution/\n",
    "\n",
    "echo\n",
    "echo \"All sam file splitted.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# %load /net/shendure/vol1/home/cao1025/analysis_script/ATAC_RNA_coassay_pipe/scATAC_seq/sam_split_permuted.py\n",
    "import sys\n",
    "import matplotlib as mpl\n",
    "mpl.use('Agg')\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "import numpy as np\n",
    "from Levenshtein import distance\n",
    "\n",
    "def find_P7_P5_barcode(index, P7_P5_barcodes):\n",
    "    '''\n",
    "    this function accept a P7_P5 barcode in the read, and a list of P7_P5 barcodes provided,\n",
    "    and then find the barcode in the P7_P5 barcode list that match the index with distance <2,\n",
    "    then return the found barcode; if no barcode is found, then return -1\n",
    "    '''\n",
    "    result = -1\n",
    "\n",
    "    for barcode in P7_P5_barcodes:\n",
    "        diff = distance(index, barcode)\n",
    "        if (diff <= 1):\n",
    "            result = barcode\n",
    "            break\n",
    "    \n",
    "    return result\n",
    "\n",
    "def samfile_barcode_count(sam_file, barcode_file):\n",
    "    \n",
    "    #generate the barcode list and barcode dictionary\n",
    "    barcodes = open(barcode_file)\n",
    "    barcode_ls = []\n",
    "    barcode_dic = {}\n",
    "    for line in barcodes:\n",
    "        barcode = line.strip()\n",
    "        barcode_ls.append(barcode)\n",
    "        barcode_dic[barcode] = 0\n",
    "    barcodes.close()\n",
    "    #read the sam file, and count the number per barcode\n",
    "    sam = open(sam_file)\n",
    "    for line in sam:\n",
    "        if (line[0] == '@'):\n",
    "            continue\n",
    "        else:\n",
    "            name = (((line.split('\\t'))[0]).split(','))\n",
    "            barcode = name[0]\n",
    "            barcode = find_P7_P5_barcode(barcode, barcode_ls)\n",
    "            if (barcode != -1):\n",
    "                barcode_dic[barcode] += 1\n",
    "    sam.close()\n",
    "    return barcode_dic\n",
    "\n",
    "def permute_samples(sam_file, barcode_count, barcode_list, output_folder):\n",
    "    '''\n",
    "    this function accept a sam file, a barcode count dictionary, a barcode list and a output folder.\n",
    "    then for each barcode in the barcode list, it find the reads number n associated with the barcode, \n",
    "    and sample n reads from the samfile and then output the sampled reads to the \n",
    "    '''\n",
    "    \n",
    "    # Generate a list of output file\n",
    "    file_name = (sam_file.split('/')[-1]).split('.')\n",
    "    file_name = file_name[0] + '.' + file_name[1]\n",
    "    output_files = {}\n",
    "    for barcode in barcode_list:\n",
    "        output_file = output_folder + '/' + file_name + '.' + barcode + '.permuted.sam'\n",
    "        output_files[barcode] = open(output_file, 'w')\n",
    "    \n",
    "    # output the header into each file and count the line number in the header and total reads number\n",
    "    header_number = 0\n",
    "    input_file = open(sam_file)\n",
    "    all_lines = input_file.readlines()\n",
    "    for line in all_lines:\n",
    "        if (line[0] == '@'):\n",
    "            header_number += 1\n",
    "            for barcode in barcode_list:\n",
    "                output_files[barcode].write(line)\n",
    "        else:\n",
    "            break\n",
    "    \n",
    "    # for each barcode, generate the permuted line array\n",
    "    # for each barcode, output the permuted lines to the output file\n",
    "    permuted_lines = np.random.permutation(all_lines[header_number:])\n",
    "    first_line = 0\n",
    "    for barcode in barcode_list:\n",
    "        end_line = first_line + barcode_count[barcode]\n",
    "        output_lines = list(permuted_lines[first_line:end_line])\n",
    "        for output_line in output_lines:\n",
    "            output_files[barcode].write(output_line)\n",
    "        first_line = end_line\n",
    "    \n",
    "    # close the output file and sam file\n",
    "    for barcode in barcode_list:\n",
    "        output_files[barcode].close()\n",
    "    input_file.close()\n",
    "    \n",
    "def split_samfile(sam_file, barcode_file, output_folder, cutoff):\n",
    "    '''\n",
    "    this script accept a sam file, a barcode file, a output_file, a cutoff value,\n",
    "    then it will call the samfile_barcode_count function and get the total read count per barcode,\n",
    "    then it use the cutoff value to filter the barcode,\n",
    "    and generate the output samfile for single cells, generate the sample_ID.txt in the output folder,\n",
    "    generate the reads distribution in the output folder/read_distribution_barcode;\n",
    "    '''\n",
    "    \n",
    "    # generate the count per barcode\n",
    "    barcode_count = samfile_barcode_count(sam_file, barcode_file)\n",
    "    \n",
    "    # plot the barcode reads distribution and save the result to the ouput folder\n",
    "    plot_name = (sam_file.split('/')[-1]).split('.')\n",
    "    plot_name = plot_name[0] + '.' + plot_name[1]\n",
    "    fig = plt.figure()\n",
    "    plt.hist(barcode_count.values(), bins=100)\n",
    "    plt.ylabel('frequency')\n",
    "    plt.xlabel('Number of unique reads')\n",
    "    fig_output = output_folder + '/' + plot_name + '.png'\n",
    "    \n",
    "    fig.savefig(fig_output)\n",
    "\n",
    "    #also output the barcode number and distribution to the output folder\n",
    "    read_dist = open(output_folder + '/' + plot_name + '.txt', 'w')\n",
    "    for barcode in barcode_count:\n",
    "        line = barcode + ', %d\\n' %(barcode_count[barcode])\n",
    "        read_dist.write(line)\n",
    "    read_dist.close()\n",
    "    \n",
    "    #filter the barcode based on the cutoff value\n",
    "    barcode_filtered = []\n",
    "    for barcode in barcode_count:\n",
    "        if barcode_count[barcode] >= cutoff:\n",
    "            barcode_filtered.append(barcode)\n",
    "    print barcode_filtered\n",
    "     \n",
    "    # for the barcode in the barcode filter list, generate permuted sample in the output folder\n",
    "    print \"Generatign permuted sequences...\"\n",
    "    permute_samples(sam_file, barcode_count, barcode_filtered, output_folder)\n",
    "    \n",
    "    #generate the output sam file and sample_list file\n",
    "    sample_list_file = open(output_folder + '/' + plot_name + '.' + 'sample_list.txt', 'w')\n",
    "    output_files = {}\n",
    "    for barcode in barcode_filtered:\n",
    "        output_file = output_folder + '/' + plot_name + '.' + barcode + '.sam'\n",
    "        output_files[barcode] = open(output_file, 'w')\n",
    "        sample_list_file.write(plot_name + '.' + barcode + '\\n')\n",
    "    \n",
    "    # output the each read to the output sam file\n",
    "    sam = open(sam_file)\n",
    "    for line in sam:\n",
    "        if (line[0] == '@'):\n",
    "            for barcode in barcode_filtered:\n",
    "                output_files[barcode].write(line)\n",
    "        else:\n",
    "            barcode = (((line.split('\\t'))[0]).split(','))[0]\n",
    "            barcode = find_P7_P5_barcode(barcode, barcode_filtered)\n",
    "            if (barcode != -1):\n",
    "                output_files[barcode].write(line)\n",
    "    \n",
    "    #close the files:\n",
    "    sample_list_file.close()\n",
    "    sam.close()\n",
    "    for barcode in barcode_filtered:\n",
    "        output_files[barcode].close()\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    sam_file = sys.argv[1]\n",
    "    barcode_file = sys.argv[2]\n",
    "    output_folder = sys.argv[3]\n",
    "    cutoff = int(sys.argv[4])\n",
    "    split_samfile(sam_file, barcode_file, output_folder, cutoff)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# %load /net/shendure/vol1/home/cao1025/analysis_script/ATAC_RNA_coassay_pipe/scATAC_seq/sam_split_permuted.py\n",
    "import sys\n",
    "import matplotlib as mpl\n",
    "mpl.use('Agg')\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "import numpy as np\n",
    "from Levenshtein import distance\n",
    "\n",
    "def find_P7_P5_barcode(index, P7_P5_barcodes):\n",
    "    '''\n",
    "    this function accept a P7_P5 barcode in the read, and a list of P7_P5 barcodes provided,\n",
    "    and then find the barcode in the P7_P5 barcode list that match the index with distance <2,\n",
    "    then return the found barcode; if no barcode is found, then return -1\n",
    "    '''\n",
    "    result = -1\n",
    "\n",
    "    for barcode in P7_P5_barcodes:\n",
    "        diff = distance(index, barcode)\n",
    "        if (diff <= 1):\n",
    "            result = barcode\n",
    "            break\n",
    "    \n",
    "    return result\n",
    "\n",
    "def samfile_barcode_count(sam_file, barcode_file):\n",
    "    \n",
    "    #generate the barcode list and barcode dictionary\n",
    "    barcodes = open(barcode_file)\n",
    "    barcode_ls = []\n",
    "    barcode_dic = {}\n",
    "    for line in barcodes:\n",
    "        barcode = line.strip()\n",
    "        barcode_ls.append(barcode)\n",
    "        barcode_dic[barcode] = 0\n",
    "    barcodes.close()\n",
    "    #read the sam file, and count the number per barcode\n",
    "    sam = open(sam_file)\n",
    "    for line in sam:\n",
    "        if (line[0] == '@'):\n",
    "            continue\n",
    "        else:\n",
    "            name = (((line.split('\\t'))[0]).split(','))\n",
    "            barcode = name[0]\n",
    "            barcode = find_P7_P5_barcode(barcode, barcode_ls)\n",
    "            if (barcode != -1):\n",
    "                barcode_dic[barcode] += 1\n",
    "    sam.close()\n",
    "    return barcode_dic\n",
    "\n",
    "def permute_samples(sam_file, barcode_count, barcode_list, output_folder):\n",
    "    '''\n",
    "    this function accept a sam file, a barcode count dictionary, a barcode list and a output folder.\n",
    "    then for each barcode in the barcode list, it find the reads number n associated with the barcode, \n",
    "    and sample n reads from the samfile and then output the sampled reads to the \n",
    "    '''\n",
    "    \n",
    "    # Generate a list of output file\n",
    "    file_name = (sam_file.split('/')[-1]).split('.')\n",
    "    file_name = file_name[0] + '.' + file_name[1]\n",
    "    output_files = {}\n",
    "    for barcode in barcode_list:\n",
    "        output_file = output_folder + '/' + file_name + '.' + barcode + '.permuted.sam'\n",
    "        output_files[barcode] = open(output_file, 'w')\n",
    "    \n",
    "    # output the header into each file and count the line number in the header and total reads number\n",
    "    header_number = 0\n",
    "    input_file = open(sam_file)\n",
    "    all_lines = input_file.readlines()\n",
    "    for line in all_lines:\n",
    "        if (line[0] == '@'):\n",
    "            header_number += 1\n",
    "            for barcode in barcode_list:\n",
    "                output_files[barcode].write(line)\n",
    "        else:\n",
    "            break\n",
    "    \n",
    "    # for each barcode, generate the permuted line array\n",
    "    # for each barcode, output the permuted lines to the output file\n",
    "    permuted_lines = np.random.permutation(all_lines[header_number:])\n",
    "    first_line = 0\n",
    "    for barcode in barcode_list:\n",
    "        end_line = first_line + barcode_count[barcode]\n",
    "        output_lines = list(permuted_lines[first_line:end_line])\n",
    "        for output_line in output_lines:\n",
    "            output_files[barcode].write(output_line)\n",
    "        first_line = end_line\n",
    "    \n",
    "    # close the output file and sam file\n",
    "    for barcode in barcode_list:\n",
    "        output_files[barcode].close()\n",
    "    input_file.close()\n",
    "    \n",
    "def split_samfile(sam_file, barcode_file, output_folder, cutoff):\n",
    "    '''\n",
    "    this script accept a sam file, a barcode file, a output_file, a cutoff value,\n",
    "    then it will call the samfile_barcode_count function and get the total read count per barcode,\n",
    "    then it use the cutoff value to filter the barcode,\n",
    "    and generate the output samfile for single cells, generate the sample_ID.txt in the output folder,\n",
    "    generate the reads distribution in the output folder/read_distribution_barcode;\n",
    "    '''\n",
    "    \n",
    "    # generate the count per barcode\n",
    "    barcode_count = samfile_barcode_count(sam_file, barcode_file)\n",
    "    \n",
    "    # plot the barcode reads distribution and save the result to the ouput folder\n",
    "    plot_name = (sam_file.split('/')[-1]).split('.')\n",
    "    plot_name = plot_name[0] + '.' + plot_name[1]\n",
    "    fig = plt.figure()\n",
    "    plt.hist(barcode_count.values(), bins=100)\n",
    "    plt.ylabel('frequency')\n",
    "    plt.xlabel('Number of unique reads')\n",
    "    fig_output = output_folder + '/' + plot_name + '.png'\n",
    "    \n",
    "    fig.savefig(fig_output)\n",
    "\n",
    "    #also output the barcode number and distribution to the output folder\n",
    "    read_dist = open(output_folder + '/' + plot_name + '.txt', 'w')\n",
    "    for barcode in barcode_count:\n",
    "        line = barcode + ', %d\\n' %(barcode_count[barcode])\n",
    "        read_dist.write(line)\n",
    "    read_dist.close()\n",
    "    \n",
    "    #filter the barcode based on the cutoff value\n",
    "    barcode_filtered = []\n",
    "    for barcode in barcode_count:\n",
    "        if barcode_count[barcode] >= cutoff:\n",
    "            barcode_filtered.append(barcode)\n",
    "    print barcode_filtered\n",
    "     \n",
    "    # for the barcode in the barcode filter list, generate permuted sample in the output folder\n",
    "    print \"Generatign permuted sequences...\"\n",
    "    permute_samples(sam_file, barcode_count, barcode_filtered, output_folder)\n",
    "    \n",
    "    #generate the output sam file and sample_list file\n",
    "    sample_list_file = open(output_folder + '/' + plot_name + '.' + 'sample_list.txt', 'w')\n",
    "    output_files = {}\n",
    "    for barcode in barcode_filtered:\n",
    "        output_file = output_folder + '/' + plot_name + '.' + barcode + '.sam'\n",
    "        output_files[barcode] = open(output_file, 'w')\n",
    "        sample_list_file.write(plot_name + '.' + barcode + '\\n')\n",
    "    \n",
    "    # output the each read to the output sam file\n",
    "    sam = open(sam_file)\n",
    "    for line in sam:\n",
    "        if (line[0] == '@'):\n",
    "            for barcode in barcode_filtered:\n",
    "                output_files[barcode].write(line)\n",
    "        else:\n",
    "            barcode = (((line.split('\\t'))[0]).split(','))[0]\n",
    "            barcode = find_P7_P5_barcode(barcode, barcode_filtered)\n",
    "            if (barcode != -1):\n",
    "                output_files[barcode].write(line)\n",
    "    \n",
    "    #close the files:\n",
    "    sample_list_file.close()\n",
    "    sam.close()\n",
    "    for barcode in barcode_filtered:\n",
    "        output_files[barcode].close()\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    sam_file = sys.argv[1]\n",
    "    barcode_file = sys.argv[2]\n",
    "    output_folder = sys.argv[3]\n",
    "    cutoff = int(sys.argv[4])\n",
    "    split_samfile(sam_file, barcode_file, output_folder, cutoff)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting /net/shendure/vol1/home/cao1025/analysis_script/ATAC_RNA_coassay_pipe/scATAC_seq/sam_split_permuted.py\n"
     ]
    }
   ],
   "source": [
    "%%writefile /net/shendure/vol1/home/cao1025/analysis_script/ATAC_RNA_coassay_pipe/scATAC_seq/sam_split_permuted.py\n",
    "import sys\n",
    "import matplotlib as mpl\n",
    "mpl.use('Agg')\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "import numpy as np\n",
    "from Levenshtein import distance\n",
    "\n",
    "def find_P7_P5_barcode(index, P7_P5_barcodes):\n",
    "    '''\n",
    "    this function accept a P7_P5 barcode in the read, and a list of P7_P5 barcodes provided,\n",
    "    and then find the barcode in the P7_P5 barcode list that match the index with distance <2,\n",
    "    then return the found barcode; if no barcode is found, then return -1\n",
    "    '''\n",
    "    result = -1\n",
    "\n",
    "    for barcode in P7_P5_barcodes:\n",
    "        diff = distance(index, barcode)\n",
    "        if (diff <= 1):\n",
    "            result = barcode\n",
    "            break\n",
    "    \n",
    "    return result\n",
    "\n",
    "def samfile_barcode_count(sam_file, barcode_file):\n",
    "    \n",
    "    #generate the barcode list and barcode dictionary\n",
    "    barcodes = open(barcode_file)\n",
    "    barcode_ls = []\n",
    "    barcode_dic = {}\n",
    "    for line in barcodes:\n",
    "        barcode = line.strip()\n",
    "        barcode_ls.append(barcode)\n",
    "        barcode_dic[barcode] = 0\n",
    "    barcodes.close()\n",
    "    #read the sam file, and count the number per barcode\n",
    "    sam = open(sam_file)\n",
    "    for line in sam:\n",
    "        if (line[0] == '@'):\n",
    "            continue\n",
    "        else:\n",
    "            name = (((line.split('\\t'))[0]).split(','))\n",
    "            barcode = name[0]\n",
    "            barcode = find_P7_P5_barcode(barcode, barcode_ls)\n",
    "            if (barcode != -1):\n",
    "                barcode_dic[barcode] += 1\n",
    "    sam.close()\n",
    "    return barcode_dic\n",
    "\n",
    "def permute_samples(sam_file, barcode_count, barcode_list, output_folder):\n",
    "    '''\n",
    "    this function accept a sam file, a barcode count dictionary, a barcode list and a output folder.\n",
    "    then for each barcode in the barcode list, it find the reads number n associated with the barcode, \n",
    "    and sample n reads from the samfile and then output the sampled reads to the \n",
    "    '''\n",
    "    \n",
    "    # Generate a list of output file\n",
    "    file_name = (sam_file.split('/')[-1]).split('.')\n",
    "    file_name = file_name[0] + '.' + file_name[1]\n",
    "    output_files = {}\n",
    "    for barcode in barcode_list:\n",
    "        output_file = output_folder + '/' + file_name + '.' + barcode + '.permuted.sam'\n",
    "        output_files[barcode] = open(output_file, 'w')\n",
    "    \n",
    "    # output the header into each file and count the line number in the header and total reads number\n",
    "    header_number = 0\n",
    "    input_file = open(sam_file)\n",
    "    all_lines = input_file.readlines()\n",
    "    for line in all_lines:\n",
    "        if (line[0] == '@'):\n",
    "            header_number += 1\n",
    "            for barcode in barcode_list:\n",
    "                output_files[barcode].write(line)\n",
    "        else:\n",
    "            break\n",
    "    \n",
    "    # for each barcode, generate the permuted line array\n",
    "    # for each barcode, output the permuted lines to the output file\n",
    "    \n",
    "    \n",
    "    permuted_lines = np.random.permutation(all_lines[header_number:])\n",
    "    first_line = 0\n",
    "    for barcode in barcode_list:\n",
    "        end_line = first_line + barcode_count[barcode]\n",
    "        output_lines = list(permuted_lines[first_line:end_line])\n",
    "        for output_line in output_lines:\n",
    "            output_files[barcode].write(output_line)\n",
    "        first_line = end_line\n",
    "    \n",
    "    \n",
    "    # close the output file and sam file\n",
    "    for barcode in barcode_list:\n",
    "        output_files[barcode].close()\n",
    "    input_file.close()\n",
    "    \n",
    "def split_samfile(sam_file, barcode_file, output_folder, cutoff):\n",
    "    '''\n",
    "    this script accept a sam file, a barcode file, a output_file, a cutoff value,\n",
    "    then it will call the samfile_barcode_count function and get the total read count per barcode,\n",
    "    then it use the cutoff value to filter the barcode,\n",
    "    and generate the output samfile for single cells, generate the sample_ID.txt in the output folder,\n",
    "    generate the reads distribution in the output folder/read_distribution_barcode;\n",
    "    '''\n",
    "    \n",
    "    # generate the count per barcode\n",
    "    barcode_count = samfile_barcode_count(sam_file, barcode_file)\n",
    "    \n",
    "    \n",
    "    # plot the barcode reads distribution and save the result to the ouput folder\n",
    "    plot_name = (sam_file.split('/')[-1]).split('.')\n",
    "    plot_name = plot_name[0] + '.' + plot_name[1]\n",
    "    '''\n",
    "    fig = plt.figure()\n",
    "    plt.hist(barcode_count.values(), bins=100)\n",
    "    plt.ylabel('frequency')\n",
    "    plt.xlabel('Number of unique reads')\n",
    "    fig_output = output_folder + '/' + plot_name + '.png'\n",
    "    \n",
    "    fig.savefig(fig_output)\n",
    "    '''\n",
    "\n",
    "    #also output the barcode number and distribution to the output folder\n",
    "    read_dist = open(output_folder + '/' + plot_name + '.txt', 'w')\n",
    "    for barcode in barcode_count:\n",
    "        line = barcode + ', %d\\n' %(barcode_count[barcode])\n",
    "        read_dist.write(line)\n",
    "    read_dist.close()\n",
    "    \n",
    "    #filter the barcode based on the cutoff value\n",
    "    barcode_filtered = []\n",
    "    for barcode in barcode_count:\n",
    "        if barcode_count[barcode] >= cutoff:\n",
    "            barcode_filtered.append(barcode)\n",
    "    print barcode_filtered\n",
    "    \n",
    "    '''\n",
    "    # for the barcode in the barcode filter list, generate permuted sample in the output folder\n",
    "    print \"Generatign permuted sequences...\"\n",
    "    permute_samples(sam_file, barcode_count, barcode_filtered, output_folder)\n",
    "    '''\n",
    "    \n",
    "    #generate the output sam file and sample_list file\n",
    "    sample_list_file = open(output_folder + '/' + plot_name + '.' + 'sample_list.txt', 'w')\n",
    "    output_files = {}\n",
    "    for barcode in barcode_filtered:\n",
    "        output_file = output_folder + '/' + plot_name + '.' + barcode + '.sam'\n",
    "        output_files[barcode] = open(output_file, 'w')\n",
    "        sample_list_file.write(plot_name + '.' + barcode + '\\n')\n",
    "    \n",
    "    # output the each read to the output sam file\n",
    "    sam = open(sam_file)\n",
    "    for line in sam:\n",
    "        if (line[0] == '@'):\n",
    "            for barcode in barcode_filtered:\n",
    "                output_files[barcode].write(line)\n",
    "        else:\n",
    "            barcode = (((line.split('\\t'))[0]).split(','))[0]\n",
    "            barcode = find_P7_P5_barcode(barcode, barcode_filtered)\n",
    "            if (barcode != -1):\n",
    "                output_files[barcode].write(line)\n",
    "    \n",
    "    #close the files:\n",
    "    sample_list_file.close()\n",
    "    sam.close()\n",
    "    for barcode in barcode_filtered:\n",
    "        output_files[barcode].close()\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    sam_file = sys.argv[1]\n",
    "    barcode_file = sys.argv[2]\n",
    "    output_folder = sys.argv[3]\n",
    "    cutoff = int(sys.argv[4])\n",
    "    split_samfile(sam_file, barcode_file, output_folder, cutoff)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# %load /net/shendure/vol1/home/cao1025/analysis_script/ATAC_RNA_coassay_pipe/scATAC_seq/samfile_split_permuted.sh\n",
    "#!/bin/bash\n",
    "\n",
    "# this script accept a sam file folder, a sample list, a output folder, a barcode file, then it will run the sam_split.py on each\n",
    "# sam file and output the splited sam file to the output folder\n",
    "\n",
    "sam_folder=$1\n",
    "sample_list=$2\n",
    "output_folder=$3\n",
    "barcode_file=$4\n",
    "cutoff=$5\n",
    "core=$6\n",
    "\n",
    "#define the location of the python\n",
    "python_path=\"/net/shendure/vol1/home/cao1025/anaconda2/bin/python2.7\"\n",
    "script_path=\"/net/shendure/vol1/home/cao1025/analysis_script/ATAC_RNA_coassay_pipe/scATAC_seq/\"\n",
    "\n",
    "echo\n",
    "echo \"Start splitting the sam file...\"\n",
    "echo samfile folder: $sam_folder\n",
    "echo sample list: $sample_list\n",
    "echo ouput folder: $output_folder\n",
    "echo barcode file: $barcode_file\n",
    "echo cutoff value: $cutoff\n",
    "\n",
    "mkdir $output_folder\n",
    "for sample in $(cat $sample_list); do echo Now splitting $sample; sem -j+$core $python_path $script_path/sam_split_permuted.py $sam_folder/$sample.sam $barcode_file $output_folder $cutoff; done\n",
    "sem --wait\n",
    "cat $output_folder/*sample_list.txt>$output_folder/All_samples.txt\n",
    "cp $output_folder/All_samples.txt $output_folder/../barcode_samples.txt\n",
    "\n",
    "# output the report the report/barcode_read_distribution folder\n",
    "mkdir -p $output_folder/../report/barcode_read_distribution\n",
    "mv $output_folder/*.txt $output_folder/../report/barcode_read_distribution/\n",
    "mv $output_folder/*.png $output_folder/../report/barcode_read_distribution/\n",
    "\n",
    "echo\n",
    "echo \"All sam file splitted.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting /net/shendure/vol1/home/cao1025/analysis_script/ATAC_RNA_coassay_pipe/scATAC_seq/report_human_mouse_fraction.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile /net/shendure/vol1/home/cao1025/analysis_script/ATAC_RNA_coassay_pipe/scATAC_seq/report_human_mouse_fraction.sh\n",
    "#! /bin/bash\n",
    "\n",
    "# this script accept a input folder, a sample ID file, and a output folder, then count the read number in mouse and human read into the report folder\n",
    "input_folder=$1\n",
    "sample_ID=$2\n",
    "output_folder=$3\n",
    "\n",
    "echo \n",
    "echo \"Start calculating the mouse and human fraction...\"\n",
    "mkdir -p $output_folder\n",
    "echo sample,human_reads,mouse_reads>$output_folder/human_mouse_fraction.txt\n",
    "echo sample,human_reads,mouse_reads>$output_folder/human_mouse_fraction_permuted.txt\n",
    "for sample in $(cat $sample_ID); do echo Processing $sample; echo $sample,$(samtools view $input_folder/$sample.sam|grep 'chr' -v|wc -l),$(samtools view $input_folder/$sample.sam|grep 'chr'|wc -l)>>$output_folder/human_mouse_fraction.txt; done\n",
    "\n",
    "#for sample in $(cat $sample_ID); do echo Processing permuted $sample; echo $sample.permuted,$(samtools view $input_folder/$sample.permuted.sam|grep 'chr' -v|wc -l),$(samtools view $input_folder/$sample.permuted.sam|grep 'chr'|wc -l)>>$output_folder/human_mouse_fraction_permuted.txt; done\n",
    "echo \"Calculation done.\"\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# %load /net/shendure/vol1/home/cao1025/analysis_script/ATAC_RNA_coassay_pipe/scATAC_seq/bam_fragment_size.py\n",
    "\"\"\"\n",
    "Created on Mon Aug 10 09:07:28 2015\n",
    "\n",
    "@author: Junyue\n",
    "\"\"\"\n",
    "\n",
    "'''\n",
    "python bam_fragment_size.py input_folder sampleID output_folder\n",
    "\n",
    "this script accept a input folder which have all the bam file and sample ID,\n",
    "then calculate all the fragment size and output the file to the output folder\n",
    "\n",
    "'''\n",
    "import sys\n",
    "import subprocess\n",
    "\n",
    "def bam_fragment_size(input_folder, sampleID, output_folder):\n",
    "    #create the output folder\n",
    "    make_folder = 'mkdir -p ' + output_folder\n",
    "    err = subprocess.call(make_folder, shell=True)\n",
    "    \n",
    "    #transform bam file to sam file and calculate the fragment size for each file\n",
    "    # in the input folder according to the sample ID\n",
    "    f1 = open(sampleID, 'r')\n",
    "    for i in f1:\n",
    "        sample = i.strip()\n",
    "        input_file = input_folder + '/' + sample + '.bam'\n",
    "        sam_file = input_folder + '/sam_file/' + sample + '.sam'\n",
    "        frag_size_file = output_folder + '/' + sample + '_frag_size.tsv'\n",
    "        \n",
    "        print '''\n",
    "        ***********calculate the fragment size******************\n",
    "        sample: %s\n",
    "        input_file: %s\n",
    "        output_file: %s''' %(sample, input_file, frag_size_file)\n",
    "        \n",
    "        #make a folder for the samfile\n",
    "        err = subprocess.call('mkdir -p ' + input_folder + '/sam_file', shell = True)\n",
    "        #transform the file to sam file\n",
    "        bamtosam = 'samtools view ' + input_file + ' >' + sam_file\n",
    "        err = subprocess.call(bamtosam, shell=True)\n",
    "        if err != 0:\n",
    "            raise IOError(\"!!!!!!!!bam file cannot be transformed to samfile!!!!!!!!\")\n",
    "        \n",
    "        #calculate the fragment size of the sam file\n",
    "        frag_size_cal = 'awk -v pat=' + \"'\" + sample + \"'\" + ''' '{if ($9 > 0) {print $9,pat}}' ''' + sam_file \\\n",
    "        + ' >' + frag_size_file\n",
    "        \n",
    "        print 'command for fragment size calculation: %s' %(frag_size_cal)\n",
    "        \n",
    "        err = subprocess.call(frag_size_cal, shell=True)\n",
    "        \n",
    "        if err != 0:\n",
    "            raise IOError(\"!!!!!!!!!!fragment size calculation met error!!!!!!!!!\")\n",
    "        else:\n",
    "            print \"~~~~~~~~~fragment size calculation for %s is done~~~~~~~~~~\" %(sample)\n",
    "    \n",
    "    print \"~~~~~~~~~~~~~~~~~all fragment size calculation is done ~~~~~~~~~~~~~\"\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    input_folder = sys.argv[1]\n",
    "    sampleID = sys.argv[2]\n",
    "    output_folder = sys.argv[3]\n",
    "    bam_fragment_size(input_folder, sampleID, output_folder)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting frag_size.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile frag_size.sh\n",
    "\n",
    "input_folder=$1\n",
    "sample_ID=$2\n",
    "output_folder=$3\n",
    "core=$4\n",
    "\n",
    "mkdir -p $output_folder\n",
    "for sample in $(cat $sample_ID); do echo Processing $sample; sem -j+$core samtools view $input_folder/$sample.sam | awk -v pat=$sample '{if ($9 > 0) {print $9,pat}}' - >$output_folder/$sample.csv; done\n",
    "sem --semaphoretimeout 600"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# %load ./dnase_intersection.sh\n",
    "#!/bin/bash\n",
    "\n",
    "# this script takes a input folder including all the bed files, a sample ID file, and a output foler, a exon1 file, a exon2 file,\n",
    "#  and then it calculate the intersected read number for each bed file\n",
    "\n",
    "input_folder=$1\n",
    "sample_ID=$2\n",
    "output_folder=$3\n",
    "exon_bed=~/reference/bed_reference/hg19/dnase_hotspot/rmchr_dnase_hotspot.bed\n",
    "gene_bed=~/reference/bed_reference/mm10/dnas_hot_spot/dnase_3T3_L1/3T3_L1.dnase.peak.bed\n",
    "\n",
    "echo\n",
    "echo \"Start read distribution calculation..\"\n",
    "mkdir $output_folder\n",
    "echo sample,total_reads,293T,3T3 >$output_folder/read_distribution.csv\n",
    "for sample in $(cat $sample_ID); do echo Calculation for $sample; echo $sample,$(cat $input_folder/$sample.bed|wc -l),$(bedtools intersect -a $input_folder/$sample.bed -b $exon_bed -u|wc -l),$(bedtools intersect -a $input_folder/$sample.bed -b $gene_bed -u|wc -l)>>$output_folder/read_distribution.csv;done\n",
    "\n",
    "echo \"Read distribution calculation finished~\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# %load ../../ATACseq_pipe/report_reads_num.sh\n",
    "#!/bin/bash/\n",
    "\n",
    "# this script accept the input parental folder, create a report folder and a read_number sub folder, and for each sample, \n",
    "# calculate the read number in fastq, trimmed_fastq, UMI_attached_R2, STAR_alignment, filtered_sam, samfile. after filter\n",
    "# barcode\n",
    "\n",
    "parental_folder=$1\n",
    "sample_ID=$2\n",
    "fastq_folder=$parental_folder/fastq\n",
    "trimmed_folder=$parental_folder/trimmed_fastq\n",
    "alignment=$parental_folder/alignment_bowtie2\n",
    "filtered_sam=$parental_folder/filtered_sam/mapped_bam\n",
    "rm_dup_sam=$parental_folder/filtered_sam/rmdup_bam\n",
    "#split_sam=$parental_folder/splited_sam\n",
    "report_folder=$parental_folder/report/read_number\n",
    "\n",
    "echo\n",
    "echo \"Start calculating the reads number...\"\n",
    "#make the report folder\n",
    "mkdir -p $report_folder\n",
    "#calculate the read number and output the read number into the report folder\n",
    "\n",
    "echo sample,total reads,after trimming,all mapped reads, uniquely aligned reads,After remove duplicates>$report_folder/read_number.csv\n",
    "\n",
    "for sample in $(cat $sample_ID); do echo calculating $sample; echo $sample,$(expr $(zcat $fastq_folder/$sample*R2*.gz|wc -l) / 4),$(expr $(zcat $trimmed_folder/$sample*R2*.gz|wc -l) / 4), $(samtools view -F 4 -q 30 $alignment/$sample.sam|wc -l),$(samtools view $filtered_sam/$sample.bam|wc -l),$(samtools view $rm_dup_sam/$sample.bam|wc -l)>>$report_folder/read_number.csv; done\n",
    "echo \"Read number calculation is done.\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
